<!DOCTYPE html>
<html lang="" xml:lang="">
<head>

  <meta charset="utf-8" />
  <meta http-equiv="X-UA-Compatible" content="IE=edge" />
  <title>Chapter 4 Annotation diagnostics | Assigning cell types with SingleR</title>
  <meta name="description" content="The SingleR book. Because sometimes, a vignette just isn’t enough." />
  <meta name="generator" content="bookdown 0.19 and GitBook 2.6.7" />

  <meta property="og:title" content="Chapter 4 Annotation diagnostics | Assigning cell types with SingleR" />
  <meta property="og:type" content="book" />
  
  
  <meta property="og:description" content="The SingleR book. Because sometimes, a vignette just isn’t enough." />
  

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="Chapter 4 Annotation diagnostics | Assigning cell types with SingleR" />
  
  <meta name="twitter:description" content="The SingleR book. Because sometimes, a vignette just isn’t enough." />
  



<meta name="date" content="2020-05-24" />

  <meta name="viewport" content="width=device-width, initial-scale=1" />
  <meta name="apple-mobile-web-app-capable" content="yes" />
  <meta name="apple-mobile-web-app-status-bar-style" content="black" />
  
  
<link rel="prev" href="using-single-cell-references.html"/>
<link rel="next" href="using-multiple-references.html"/>
<script src="libs/jquery-2.2.3/jquery.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-table.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-clipboard.css" rel="stylesheet" />











<style type="text/css">
div.sourceCode { overflow-x: auto; }
table.sourceCode, tr.sourceCode, td.lineNumbers, td.sourceCode {
  margin: 0; padding: 0; vertical-align: baseline; border: none; }
table.sourceCode { width: 100%; line-height: 100%; background-color: #f8f8f8; }
td.lineNumbers { text-align: right; padding-right: 4px; padding-left: 4px; color: #aaaaaa; border-right: 1px solid #aaaaaa; }
td.sourceCode { padding-left: 5px; }
pre, code { background-color: #f8f8f8; }
code > span.kw { color: #204a87; font-weight: bold; } /* Keyword */
code > span.dt { color: #204a87; } /* DataType */
code > span.dv { color: #0000cf; } /* DecVal */
code > span.bn { color: #0000cf; } /* BaseN */
code > span.fl { color: #0000cf; } /* Float */
code > span.ch { color: #4e9a06; } /* Char */
code > span.st { color: #4e9a06; } /* String */
code > span.co { color: #8f5902; font-style: italic; } /* Comment */
code > span.ot { color: #8f5902; } /* Other */
code > span.al { color: #ef2929; } /* Alert */
code > span.fu { color: #000000; } /* Function */
code > span.er { color: #a40000; font-weight: bold; } /* Error */
code > span.wa { color: #8f5902; font-weight: bold; font-style: italic; } /* Warning */
code > span.cn { color: #000000; } /* Constant */
code > span.sc { color: #000000; } /* SpecialChar */
code > span.vs { color: #4e9a06; } /* VerbatimString */
code > span.ss { color: #4e9a06; } /* SpecialString */
code > span.im { } /* Import */
code > span.va { color: #000000; } /* Variable */
code > span.cf { color: #204a87; font-weight: bold; } /* ControlFlow */
code > span.op { color: #ce5c00; font-weight: bold; } /* Operator */
code > span.pp { color: #8f5902; font-style: italic; } /* Preprocessor */
code > span.ex { } /* Extension */
code > span.at { color: #c4a000; } /* Attribute */
code > span.do { color: #8f5902; font-weight: bold; font-style: italic; } /* Documentation */
code > span.an { color: #8f5902; font-weight: bold; font-style: italic; } /* Annotation */
code > span.cv { color: #8f5902; font-weight: bold; font-style: italic; } /* CommentVar */
code > span.in { color: #8f5902; font-weight: bold; font-style: italic; } /* Information */
</style>

<link rel="stylesheet" href="style.css" type="text/css" />
</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li><a href="./">The SingleR book</a></li>

<li class="divider"></li>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html"><i class="fa fa-check"></i>Preface</a></li>
<li class="part"><span><b>I Basic usage</b></span></li>
<li class="chapter" data-level="1" data-path="introduction.html"><a href="introduction.html"><i class="fa fa-check"></i><b>1</b> Introduction</a><ul>
<li class="chapter" data-level="1.1" data-path="introduction.html"><a href="introduction.html#motivation"><i class="fa fa-check"></i><b>1.1</b> Motivation</a></li>
<li class="chapter" data-level="1.2" data-path="introduction.html"><a href="introduction.html#method-description"><i class="fa fa-check"></i><b>1.2</b> Method description</a></li>
<li class="chapter" data-level="1.3" data-path="introduction.html"><a href="introduction.html#quick-start"><i class="fa fa-check"></i><b>1.3</b> Quick start</a></li>
<li class="chapter" data-level="1.4" data-path="introduction.html"><a href="introduction.html#where-to-get-help"><i class="fa fa-check"></i><b>1.4</b> Where to get help</a></li>
<li class="chapter" data-level="" data-path="introduction.html"><a href="introduction.html#session-information"><i class="fa fa-check"></i>Session information</a></li>
</ul></li>
<li class="chapter" data-level="2" data-path="using-the-built-in-references.html"><a href="using-the-built-in-references.html"><i class="fa fa-check"></i><b>2</b> Using the built-in references</a><ul>
<li class="chapter" data-level="2.1" data-path="using-the-built-in-references.html"><a href="using-the-built-in-references.html#overview"><i class="fa fa-check"></i><b>2.1</b> Overview</a></li>
<li class="chapter" data-level="2.2" data-path="using-the-built-in-references.html"><a href="using-the-built-in-references.html#annotation-with-default-marker-detection"><i class="fa fa-check"></i><b>2.2</b> Annotation with default marker detection</a></li>
<li class="chapter" data-level="2.3" data-path="using-the-built-in-references.html"><a href="using-the-built-in-references.html#interaction-with-quality-control"><i class="fa fa-check"></i><b>2.3</b> Interaction with quality control</a></li>
<li class="chapter" data-level="2.4" data-path="using-the-built-in-references.html"><a href="using-the-built-in-references.html#choices-of-assay-data"><i class="fa fa-check"></i><b>2.4</b> Choices of assay data</a></li>
<li class="chapter" data-level="" data-path="using-the-built-in-references.html"><a href="using-the-built-in-references.html#session-information-1"><i class="fa fa-check"></i>Session information</a></li>
</ul></li>
<li class="chapter" data-level="3" data-path="using-single-cell-references.html"><a href="using-single-cell-references.html"><i class="fa fa-check"></i><b>3</b> Using single-cell references</a><ul>
<li class="chapter" data-level="3.1" data-path="using-single-cell-references.html"><a href="using-single-cell-references.html#overview-1"><i class="fa fa-check"></i><b>3.1</b> Overview</a></li>
<li class="chapter" data-level="3.2" data-path="using-single-cell-references.html"><a href="using-single-cell-references.html#annotation-with-test-based-marker-detection"><i class="fa fa-check"></i><b>3.2</b> Annotation with test-based marker detection</a></li>
<li class="chapter" data-level="3.3" data-path="using-single-cell-references.html"><a href="using-single-cell-references.html#defining-custom-markers"><i class="fa fa-check"></i><b>3.3</b> Defining custom markers</a></li>
<li class="chapter" data-level="3.4" data-path="using-single-cell-references.html"><a href="using-single-cell-references.html#pseudo-bulk-aggregation"><i class="fa fa-check"></i><b>3.4</b> Pseudo-bulk aggregation</a></li>
<li class="chapter" data-level="" data-path="using-single-cell-references.html"><a href="using-single-cell-references.html#session-information-2"><i class="fa fa-check"></i>Session information</a></li>
</ul></li>
<li class="chapter" data-level="4" data-path="annotation-diagnostics.html"><a href="annotation-diagnostics.html"><i class="fa fa-check"></i><b>4</b> Annotation diagnostics</a><ul>
<li class="chapter" data-level="4.1" data-path="annotation-diagnostics.html"><a href="annotation-diagnostics.html#overview-2"><i class="fa fa-check"></i><b>4.1</b> Overview</a></li>
<li class="chapter" data-level="4.2" data-path="annotation-diagnostics.html"><a href="annotation-diagnostics.html#based-on-the-scores-within-cells"><i class="fa fa-check"></i><b>4.2</b> Based on the scores within cells</a></li>
<li class="chapter" data-level="4.3" data-path="annotation-diagnostics.html"><a href="annotation-diagnostics.html#based-on-the-deltas-across-cells"><i class="fa fa-check"></i><b>4.3</b> Based on the deltas across cells</a></li>
<li class="chapter" data-level="4.4" data-path="annotation-diagnostics.html"><a href="annotation-diagnostics.html#based-on-marker-gene-expression"><i class="fa fa-check"></i><b>4.4</b> Based on marker gene expression</a></li>
<li class="chapter" data-level="" data-path="annotation-diagnostics.html"><a href="annotation-diagnostics.html#session-information-3"><i class="fa fa-check"></i>Session information</a></li>
</ul></li>
<li class="part"><span><b>II Advanced usage</b></span></li>
<li class="chapter" data-level="5" data-path="using-multiple-references.html"><a href="using-multiple-references.html"><i class="fa fa-check"></i><b>5</b> Using multiple references</a><ul>
<li class="chapter" data-level="5.1" data-path="using-multiple-references.html"><a href="using-multiple-references.html#overview-3"><i class="fa fa-check"></i><b>5.1</b> Overview</a></li>
<li class="chapter" data-level="5.2" data-path="using-multiple-references.html"><a href="using-multiple-references.html#using-reference-specific-labels"><i class="fa fa-check"></i><b>5.2</b> Using reference-specific labels</a></li>
<li class="chapter" data-level="5.3" data-path="using-multiple-references.html"><a href="using-multiple-references.html#comparing-scores-across-references"><i class="fa fa-check"></i><b>5.3</b> Comparing scores across references</a><ul>
<li class="chapter" data-level="5.3.1" data-path="using-multiple-references.html"><a href="using-multiple-references.html#combining-inferences-from-individual-references"><i class="fa fa-check"></i><b>5.3.1</b> Combining inferences from individual references</a></li>
<li class="chapter" data-level="5.3.2" data-path="using-multiple-references.html"><a href="using-multiple-references.html#combined-diagnostics"><i class="fa fa-check"></i><b>5.3.2</b> Combined diagnostics</a></li>
</ul></li>
<li class="chapter" data-level="5.4" data-path="using-multiple-references.html"><a href="using-multiple-references.html#using-harmonized-labels"><i class="fa fa-check"></i><b>5.4</b> Using harmonized labels</a><ul>
<li class="chapter" data-level="5.4.1" data-path="using-multiple-references.html"><a href="using-multiple-references.html#sharing-information-during-marker-detection"><i class="fa fa-check"></i><b>5.4.1</b> Sharing information during marker detection</a></li>
<li class="chapter" data-level="5.4.2" data-path="using-multiple-references.html"><a href="using-multiple-references.html#manual-label-harmonization"><i class="fa fa-check"></i><b>5.4.2</b> Manual label harmonization</a></li>
</ul></li>
<li class="chapter" data-level="" data-path="using-multiple-references.html"><a href="using-multiple-references.html#session-info"><i class="fa fa-check"></i>Session info</a></li>
</ul></li>
<li class="chapter" data-level="6" data-path="exploiting-the-cell-ontology.html"><a href="exploiting-the-cell-ontology.html"><i class="fa fa-check"></i><b>6</b> Exploiting the cell ontology</a><ul>
<li class="chapter" data-level="6.1" data-path="exploiting-the-cell-ontology.html"><a href="exploiting-the-cell-ontology.html#motivation-1"><i class="fa fa-check"></i><b>6.1</b> Motivation</a></li>
<li class="chapter" data-level="6.2" data-path="exploiting-the-cell-ontology.html"><a href="exploiting-the-cell-ontology.html#basic-manipulation"><i class="fa fa-check"></i><b>6.2</b> Basic manipulation</a></li>
<li class="chapter" data-level="" data-path="exploiting-the-cell-ontology.html"><a href="exploiting-the-cell-ontology.html#session-information-4"><i class="fa fa-check"></i>Session information</a></li>
</ul></li>
<li class="chapter" data-level="7" data-path="advanced-options.html"><a href="advanced-options.html"><i class="fa fa-check"></i><b>7</b> Advanced options</a><ul>
<li class="chapter" data-level="7.1" data-path="advanced-options.html"><a href="advanced-options.html#preconstructed-indices"><i class="fa fa-check"></i><b>7.1</b> Preconstructed indices</a></li>
<li class="chapter" data-level="7.2" data-path="advanced-options.html"><a href="advanced-options.html#parallelization"><i class="fa fa-check"></i><b>7.2</b> Parallelization</a></li>
<li class="chapter" data-level="7.3" data-path="advanced-options.html"><a href="advanced-options.html#approximate-algorithms"><i class="fa fa-check"></i><b>7.3</b> Approximate algorithms</a></li>
<li class="chapter" data-level="" data-path="advanced-options.html"><a href="advanced-options.html#session-information-5"><i class="fa fa-check"></i>Session information</a></li>
</ul></li>
<li class="part"><span><b>III Workflows</b></span></li>
<li class="chapter" data-level="8" data-path="cross-annotating-pancreas.html"><a href="cross-annotating-pancreas.html"><i class="fa fa-check"></i><b>8</b> Cross-annotating pancreas</a><ul>
<li class="chapter" data-level="8.1" data-path="cross-annotating-pancreas.html"><a href="cross-annotating-pancreas.html#loading-the-data"><i class="fa fa-check"></i><b>8.1</b> Loading the data</a></li>
<li class="chapter" data-level="8.2" data-path="cross-annotating-pancreas.html"><a href="cross-annotating-pancreas.html#applying-the-annotation"><i class="fa fa-check"></i><b>8.2</b> Applying the annotation</a></li>
<li class="chapter" data-level="8.3" data-path="cross-annotating-pancreas.html"><a href="cross-annotating-pancreas.html#diagnostics"><i class="fa fa-check"></i><b>8.3</b> Diagnostics</a></li>
<li class="chapter" data-level="" data-path="cross-annotating-pancreas.html"><a href="cross-annotating-pancreas.html#session-information-6"><i class="fa fa-check"></i>Session information</a></li>
</ul></li>
<li class="part"><span><b>IV Appendix</b></span></li>
<li class="chapter" data-level="9" data-path="contributors.html"><a href="contributors.html"><i class="fa fa-check"></i><b>9</b> Contributors</a><ul>
<li><a href="contributors.html#aaron-lun"><em>Aaron Lun</em></a></li>
</ul></li>
<li class="chapter" data-level="10" data-path="bibliography.html"><a href="bibliography.html"><i class="fa fa-check"></i><b>10</b> Bibliography</a></li>
<li class="divider"></li>
<li><a href="https://bioconductor.org" target="blank">Published by Bioconductor</a></li>

</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">Assigning cell types with SingleR</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="annotation-diagnostics" class="section level1">
<h1><span class="header-section-number">Chapter 4</span> Annotation diagnostics</h1>
<script>
document.addEventListener("click", function (event) {
    if (event.target.classList.contains("aaron-collapse")) {
        event.target.classList.toggle("active");
        var content = event.target.nextElementSibling;
        if (content.style.display === "block") {
            content.style.display = "none";
        } else {
            content.style.display = "block";
        }
    }
})
</script>
<style>
.aaron-collapse {
  background-color: #eee;
  color: #444;
  cursor: pointer;
  padding: 18px;
  width: 100%;
  border: none;
  text-align: left;
  outline: none;
  font-size: 15px;
}

.aaron-content {
  padding: 0 18px;
  display: none;
  overflow: hidden;
  background-color: #f1f1f1;
}
</style>
<div id="overview-2" class="section level2">
<h2><span class="header-section-number">4.1</span> Overview</h2>
<p>In addition to the labels, <code>SingleR()</code> returns a number of helpful diagnostics about the annotation process that can be used to determine whether the assignments are appropriate. Unambiguous assignments corroborated by expression of canonical markers add confidence to the results; conversely, low-confidence assignments can be pruned out to avoid adding noise to downstream analyses. This chapter will demonstrate some of these common sanity checks on the pancreas datasets from Chapter <a href="using-single-cell-references.html#using-single-cell-references">3</a> <span class="citation">(Muraro et al. <a href="#ref-muraro2016singlecell">2016</a>; Grun et al. <a href="#ref-grun2016denovo">2016</a>)</span>.</p>
<button class="aaron-collapse">
View history
</button>
<div class="aaron-content">
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co">#--- loading-muraro ---#</span>
<span class="kw">library</span>(scRNAseq)
sceM &lt;-<span class="st"> </span><span class="kw">MuraroPancreasData</span>()

sceM &lt;-<span class="st"> </span>sceM[,<span class="op">!</span><span class="kw">is.na</span>(sceM<span class="op">$</span>label) <span class="op">&amp;</span><span class="st"> </span>sceM<span class="op">$</span>label<span class="op">!=</span><span class="st">&quot;unclear&quot;</span>] 

<span class="kw">library</span>(scater)
sceM &lt;-<span class="st"> </span><span class="kw">logNormCounts</span>(sceM)

<span class="co">#--- loading-grun ---#</span>
sceG &lt;-<span class="st"> </span><span class="kw">GrunPancreasData</span>()

sceG &lt;-<span class="st"> </span><span class="kw">addPerCellQC</span>(sceG)
qc &lt;-<span class="st"> </span><span class="kw">quickPerCellQC</span>(<span class="kw">colData</span>(sceG), 
    <span class="dt">percent_subsets=</span><span class="st">&quot;altexps_ERCC_percent&quot;</span>,
    <span class="dt">batch=</span>sceG<span class="op">$</span>donor,
    <span class="dt">subset=</span>sceG<span class="op">$</span>donor <span class="op">%in%</span><span class="st"> </span><span class="kw">c</span>(<span class="st">&quot;D17&quot;</span>, <span class="st">&quot;D7&quot;</span>, <span class="st">&quot;D2&quot;</span>))
sceG &lt;-<span class="st"> </span>sceG[,<span class="op">!</span>qc<span class="op">$</span>discard]

sceG &lt;-<span class="st"> </span><span class="kw">logNormCounts</span>(sceG)

<span class="co">#--- annotation ---#</span>
<span class="kw">library</span>(SingleR)
pred.grun &lt;-<span class="st"> </span><span class="kw">SingleR</span>(<span class="dt">test=</span>sceG, <span class="dt">ref=</span>sceM, <span class="dt">labels=</span>sceM<span class="op">$</span>label, <span class="dt">de.method=</span><span class="st">&quot;wilcox&quot;</span>)</code></pre></div>
</div>
</div>
<div id="based-on-the-scores-within-cells" class="section level2">
<h2><span class="header-section-number">4.2</span> Based on the scores within cells</h2>
<p>The most obvious diagnostic reported by <code>SingleR()</code> is the nested matrix of per-cell scores in the <code>scores</code> field. This contains the correlation-based scores prior to any fine-tuning for each cell (row) and reference label (column). Ideally, we would see unambiguous assignments where, for any given cell, one label’s score is clearly larger than the others.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">pred.grun<span class="op">$</span>scores[<span class="dv">1</span><span class="op">:</span><span class="dv">10</span>,]</code></pre></div>
<pre><code>##       acinar  alpha   beta  delta   duct endothelial epsilon mesenchymal
##  [1,] 0.6797 0.1463 0.1554 0.1029 0.4971      0.2030 0.04434      0.1848
##  [2,] 0.6832 0.1849 0.1611 0.1250 0.5133      0.2231 0.08909      0.2140
##  [3,] 0.6845 0.2032 0.2141 0.1875 0.5060      0.2286 0.08960      0.1961
##  [4,] 0.6472 0.2050 0.2115 0.1879 0.7300      0.2285 0.12107      0.2708
##  [5,] 0.6079 0.2285 0.2419 0.1938 0.6616      0.2659 0.16007      0.3228
##  [6,] 0.5816 0.2789 0.2611 0.2661 0.6827      0.2857 0.21781      0.3151
##  [7,] 0.5704 0.3268 0.2990 0.2618 0.6293      0.2947 0.20834      0.3263
##  [8,] 0.5692 0.2455 0.2152 0.2033 0.6495      0.2652 0.18469      0.2943
##  [9,] 0.6421 0.2101 0.1929 0.1894 0.7201      0.2795 0.14615      0.3142
## [10,] 0.7153 0.2109 0.1747 0.1606 0.5896      0.1687 0.15639      0.1778
##            pp
##  [1,] 0.08954
##  [2,] 0.11207
##  [3,] 0.15207
##  [4,] 0.14351
##  [5,] 0.18475
##  [6,] 0.21898
##  [7,] 0.25871
##  [8,] 0.19160
##  [9,] 0.17281
## [10,] 0.16314</code></pre>
<p>To check whether this is indeed the case, we use the <code>plotScoreHeatmap()</code> function to visualize the score matrix (Figure <a href="annotation-diagnostics.html#fig:score-heatmap-grun">4.1</a>). Here, the key is to examine the spread of scores within each cell, i.e., down the columns of the heatmap. Similar scores for a group of labels indicates that the assignment is uncertain for those columns, though this may be acceptable if the uncertainty is distributed across closely related cell types.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(SingleR)
<span class="kw">plotScoreHeatmap</span>(pred.grun)</code></pre></div>
<div class="figure"><span id="fig:score-heatmap-grun"></span>
<img src="diagnostics_files/figure-html/score-heatmap-grun-1.png" alt="Heatmap of normalized scores for the Grun dataset. Each cell is a column while each row is a label in the reference Muraro dataset. The final label (after fine-tuning) for each cell is shown in the top color bar." width="672" />
<p class="caption">
Figure 4.1: Heatmap of normalized scores for the Grun dataset. Each cell is a column while each row is a label in the reference Muraro dataset. The final label (after fine-tuning) for each cell is shown in the top color bar.
</p>
</div>
<p>We can also display other metadata information for each cell by setting <code>clusters=</code> or <code>annotation_col=</code>. This is occasionally useful for examining potential batch effects, differences in cell type composition between conditions, relationship to clusters from an unsupervised analysis and so on,. For example, Figure <a href="annotation-diagnostics.html#fig:score-heatmap-grun-donor">4.2</a> displays the donor of origin for each cell; we can see that each cell type has contributions from multiple donors, which is reassuring as it indicates that our assignments are not (purely) driven by donor effects.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">plotScoreHeatmap</span>(pred.grun, 
    <span class="dt">annotation_col=</span><span class="kw">as.data.frame</span>(<span class="kw">colData</span>(sceG)[,<span class="st">&quot;donor&quot;</span>,<span class="dt">drop=</span><span class="ot">FALSE</span>]))</code></pre></div>
<div class="figure"><span id="fig:score-heatmap-grun-donor"></span>
<img src="diagnostics_files/figure-html/score-heatmap-grun-donor-1.png" alt="Heatmap of normalized scores for the Grun dataset, including the donor of origin for each cell." width="672" />
<p class="caption">
Figure 4.2: Heatmap of normalized scores for the Grun dataset, including the donor of origin for each cell.
</p>
</div>
<p>The <code>scores</code> matrix has several caveats associated with its interpretation. Only the pre-tuned scores are stored in this matrix, as scores after fine-tuning are not comparable across all labels. This means that the label with the highest score for a cell may not be the cell’s final label if fine-tuning is applied. Moreover, the magnitude of differences in the scores has no clear interpretation; indeed, <code>plotScoreHeatmap()</code> dispenses with any faithful representation of the scores and instead adjusts the values to highlight any differences between labels within each cell.</p>
</div>
<div id="based-on-the-deltas-across-cells" class="section level2">
<h2><span class="header-section-number">4.3</span> Based on the deltas across cells</h2>
<p>We identify poor-quality or ambiguous assignments based on the per-cell “delta”, i.e., the difference between the score for the assigned label and the median across all labels for each cell. Our assumption is that most of the labels in the reference are not relevant to any given cell. Thus, the median across all labels can be used as a measure of the baseline correlation, while the gap from the assigned label to this baseline can be used as a measure of the assignment confidence.</p>
<p>Low deltas indicate that the assignment is uncertain, possibly because the cell’s true label does not exist in the reference. An obvious next step is to apply a threshold on the delta to filter out these low-confidence assignments. We use the delta rather than the assignment score as the latter is more sensitive to technical effects. For example, changes in library size affect the technical noise and can increase/decrease all scores for a given cell, while the delta is somewhat more robust as it focuses on the differences between scores within each cell.</p>
<p><code>SingleR()</code> will set a threshold on the delta for each label using an outlier-based strategy. Specifically, we identify cells with deltas that are small outliers relative to the deltas of other cells with the same label. This assumes that, for any given label, most cells assigned to that label are correct. We focus on outliers to avoid difficulties with setting a fixed threshold, especially given that the magnitudes of the deltas are about as uninterpretable as the scores themselves. Pruned labels are reported in the <code>pruned.labels</code> field where low-quality assignments are replaced with <code>NA</code>.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">to.remove &lt;-<span class="st"> </span><span class="kw">is.na</span>(pred.grun<span class="op">$</span>pruned.labels)
<span class="kw">table</span>(<span class="dt">Label=</span>pred.grun<span class="op">$</span>labels, <span class="dt">Removed=</span>to.remove)</code></pre></div>
<pre><code>##              Removed
## Label         FALSE TRUE
##   acinar        251   26
##   alpha         198    5
##   beta          180    1
##   delta          49    1
##   duct          301    5
##   endothelial     4    1
##   epsilon         1    0
##   mesenchymal    22    0
##   pp             17    2</code></pre>
<p>However, the default pruning parameters may not be appropriate for every dataset. For example, if one label is consistently misassigned, the assumption that most cells are correctly assigned will not be appropriate. In such cases, we can revert to a fixed threshold by manually calling the underlying <code>pruneScores()</code> function with <code>min.diff.med=</code>. The example below discards cells with deltas below an arbitrary threshold of 0.2, where higher thresholds correspond to greater assignment certainty.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">to.remove &lt;-<span class="st"> </span><span class="kw">pruneScores</span>(pred.grun, <span class="dt">min.diff.med=</span><span class="fl">0.2</span>)
<span class="kw">table</span>(<span class="dt">Label=</span>pred.grun<span class="op">$</span>labels, <span class="dt">Removed=</span>to.remove)</code></pre></div>
<pre><code>##              Removed
## Label         FALSE TRUE
##   acinar        250   27
##   alpha         155   48
##   beta          148   33
##   delta          33   17
##   duct          301    5
##   endothelial     4    1
##   epsilon         0    1
##   mesenchymal    22    0
##   pp              4   15</code></pre>
<p>This entire process can be visualized using the <code>plotScoreDistribution()</code> function, which displays the per-label distribution of the deltas across cells (Figure <a href="annotation-diagnostics.html#fig:score-dist-grun">4.3</a>). We can use this plot to check that outlier detection in <code>pruneScores()</code> behaved sensibly. Labels with especially low deltas may warrant some additional caution in their interpretation.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">plotDeltaDistribution</span>(pred.grun)</code></pre></div>
<div class="figure"><span id="fig:score-dist-grun"></span>
<img src="diagnostics_files/figure-html/score-dist-grun-1.png" alt="Distribution of deltas for the Grun dataset. Each facet represents a label in the Muraro dataset, and each point represents a cell assigned to that label (colored by whether it was pruned)." width="672" />
<p class="caption">
Figure 4.3: Distribution of deltas for the Grun dataset. Each facet represents a label in the Muraro dataset, and each point represents a cell assigned to that label (colored by whether it was pruned).
</p>
</div>
<p>If fine-tuning was performed, we can apply an even more stringent filter based on the difference between the highest and second-highest scores after fine-tuning. Cells will only pass the filter if they are assigned to a label that is clearly distinguishable from any other label. In practice, this approach tends to be too conservative as assignments involving closely related labels are heavily penalized.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">to.remove2 &lt;-<span class="st"> </span><span class="kw">pruneScores</span>(pred.grun, <span class="dt">min.diff.next=</span><span class="fl">0.1</span>)
<span class="kw">table</span>(<span class="dt">Label=</span>pred.grun<span class="op">$</span>labels, <span class="dt">Removed=</span>to.remove2)</code></pre></div>
<pre><code>##              Removed
## Label         FALSE TRUE
##   acinar        235   42
##   alpha         166   37
##   beta          117   64
##   delta          25   25
##   duct          157  149
##   endothelial     4    1
##   epsilon         0    1
##   mesenchymal    22    0
##   pp              9   10</code></pre>
</div>
<div id="based-on-marker-gene-expression" class="section level2">
<h2><span class="header-section-number">4.4</span> Based on marker gene expression</h2>
<p>Another simple yet effective diagnostic is to examine the expression of the marker genes for each label in the test dataset. The marker genes used for each label are reported in the <code>metadata()</code> of the <code>SingleR()</code> output, so we can simply retrieve them to visualize their (usually log-transformed) expression values across the test dataset. Here, we use the <code>plotHeatmap()</code> function from <em><a href="https://bioconductor.org/packages/3.12/scater">scater</a></em> to examine the expression of markers used to identify beta cells (Figure <a href="annotation-diagnostics.html#fig:grun-beta-heat">4.4</a>).</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">all.markers &lt;-<span class="st"> </span><span class="kw">metadata</span>(pred.grun)<span class="op">$</span>de.genes
sceG<span class="op">$</span>labels &lt;-<span class="st"> </span>pred.grun<span class="op">$</span>labels

<span class="kw">library</span>(scater)
<span class="kw">plotHeatmap</span>(sceG, <span class="dt">order_columns_by=</span><span class="st">&quot;labels&quot;</span>,
    <span class="dt">features=</span><span class="kw">unique</span>(<span class="kw">unlist</span>(all.markers<span class="op">$</span>beta))) </code></pre></div>
<div class="figure"><span id="fig:grun-beta-heat"></span>
<img src="diagnostics_files/figure-html/grun-beta-heat-1.png" alt="Heatmap of log-expression values in the Grun dataset for all marker genes upregulated in beta cells in the Muraro reference dataset. Assigned labels for each cell are shown at the top of the plot." width="672" />
<p class="caption">
Figure 4.4: Heatmap of log-expression values in the Grun dataset for all marker genes upregulated in beta cells in the Muraro reference dataset. Assigned labels for each cell are shown at the top of the plot.
</p>
</div>
<p>If a cell in the test dataset is confidently assigned to a particular label, we would expect it to have strong expression of that label’s markers. We would also hope that those label’s markers are biologically meaningful; in this case, we do observe strong upregulation of insulin (<em>INS</em>) in the beta cells, which is reassuring and gives greater confidence to the correctness of the assignment. If the identified markers are not meaningful or not consistently upregulated, some skepticism towards the quality of the assignments is warranted.</p>
<p>It is straightforward to repeat this process for all labels by wrapping this code in a loop, as shown below in Figure <a href="annotation-diagnostics.html#fig:grun-beta-heat-all">4.5</a>. Note that <code>plotHeatmap()</code> is not the only function that can be used for this visualization; we could also use <code>plotDots()</code> to create a <em><a href="https://CRAN.R-project.org/package=Seurat">Seurat</a></em>-style dot plot, or we could use other heatmap plotting functions such as <code>dittoHeatmap()</code> from <em><a href="https://bioconductor.org/packages/3.12/dittoSeq">dittoSeq</a></em>.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">collected &lt;-<span class="st"> </span><span class="kw">list</span>()
<span class="cf">for</span> (lab <span class="cf">in</span> <span class="kw">unique</span>(pred.grun<span class="op">$</span>labels)) {
    collected[[lab]] &lt;-<span class="st"> </span><span class="kw">plotHeatmap</span>(sceG, <span class="dt">silent=</span><span class="ot">TRUE</span>, 
        <span class="dt">order_columns_by=</span><span class="st">&quot;labels&quot;</span>, <span class="dt">main=</span>lab,
        <span class="dt">features=</span><span class="kw">unique</span>(<span class="kw">unlist</span>(all.markers[[lab]])))[[<span class="dv">4</span>]] 
}
<span class="kw">do.call</span>(gridExtra<span class="op">::</span>grid.arrange, collected)</code></pre></div>
<div class="figure"><span id="fig:grun-beta-heat-all"></span>
<img src="diagnostics_files/figure-html/grun-beta-heat-all-1.png" alt="Heatmaps of log-expression values in the Grun dataset for all marker genes upregulated in each label in the Muraro reference dataset. Assigned labels for each cell are shown at the top of each plot." width="1920" />
<p class="caption">
Figure 4.5: Heatmaps of log-expression values in the Grun dataset for all marker genes upregulated in each label in the Muraro reference dataset. Assigned labels for each cell are shown at the top of each plot.
</p>
</div>
<p>In general, the heatmap provides a more interpretable diagnostic visualization than the plots of scores and deltas. However, it does require more effort to inspect and may not be feasible for large numbers of labels. It is also difficult to use a heatmap to determine the correctness of assignment for closely related labels.</p>
</div>
<div id="session-information-3" class="section level2 unnumbered">
<h2>Session information</h2>
<button class="aaron-collapse">
View session info
</button>
<div class="aaron-content">
<pre><code>R version 4.0.0 Patched (2020-05-01 r78341)
Platform: x86_64-pc-linux-gnu (64-bit)
Running under: Ubuntu 18.04.4 LTS

Matrix products: default
BLAS:   /home/luna/Software/R/R-4-0-branch-dev/lib/libRblas.so
LAPACK: /home/luna/Software/R/R-4-0-branch-dev/lib/libRlapack.so

locale:
 [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              
 [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    
 [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   
 [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 
 [9] LC_ADDRESS=C               LC_TELEPHONE=C            
[11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       

attached base packages:
[1] parallel  stats4    stats     graphics  grDevices utils     datasets 
[8] methods   base     

other attached packages:
 [1] scater_1.17.0               ggplot2_3.3.0              
 [3] SingleCellExperiment_1.11.1 SingleR_1.3.4              
 [5] SummarizedExperiment_1.19.4 DelayedArray_0.15.1        
 [7] matrixStats_0.56.0          Biobase_2.49.0             
 [9] GenomicRanges_1.41.1        GenomeInfoDb_1.25.0        
[11] IRanges_2.23.4              S4Vectors_0.27.6           
[13] BiocGenerics_0.35.2         BiocStyle_2.17.0           
[15] rebook_0.99.0              

loaded via a namespace (and not attached):
 [1] bitops_1.0-6                  bit64_0.9-7                  
 [3] RColorBrewer_1.1-2            httr_1.4.1                   
 [5] tools_4.0.0                   R6_2.4.1                     
 [7] irlba_2.3.3                   vipor_0.4.5                  
 [9] colorspace_1.4-1              DBI_1.1.0                    
[11] withr_2.2.0                   gridExtra_2.3                
[13] tidyselect_1.1.0              processx_3.4.2               
[15] bit_1.1-15.2                  curl_4.3                     
[17] compiler_4.0.0                graph_1.67.0                 
[19] BiocNeighbors_1.7.0           labeling_0.3                 
[21] bookdown_0.19                 scales_1.1.1                 
[23] callr_3.4.3                   rappdirs_0.3.1               
[25] stringr_1.4.0                 digest_0.6.25                
[27] rmarkdown_2.1                 XVector_0.29.0               
[29] pkgconfig_2.0.3               htmltools_0.4.0              
[31] highr_0.8                     dbplyr_1.4.3                 
[33] fastmap_1.0.1                 rlang_0.4.6                  
[35] RSQLite_2.2.0                 shiny_1.4.0.2                
[37] DelayedMatrixStats_1.11.0     farver_2.0.3                 
[39] BiocParallel_1.23.0           dplyr_0.8.5                  
[41] RCurl_1.98-1.2                magrittr_1.5                 
[43] BiocSingular_1.5.0            GenomeInfoDbData_1.2.3       
[45] Matrix_1.2-18                 ggbeeswarm_0.6.0             
[47] munsell_0.5.0                 Rcpp_1.0.4.6                 
[49] viridis_0.5.1                 lifecycle_0.2.0              
[51] stringi_1.4.6                 yaml_2.2.1                   
[53] zlibbioc_1.35.0               BiocFileCache_1.13.0         
[55] AnnotationHub_2.21.0          grid_4.0.0                   
[57] blob_1.2.1                    promises_1.1.0               
[59] ExperimentHub_1.15.0          crayon_1.3.4                 
[61] lattice_0.20-41               CodeDepends_0.6.5            
[63] knitr_1.28                    ps_1.3.3                     
[65] pillar_1.4.4                  codetools_0.2-16             
[67] XML_3.99-0.3                  glue_1.4.1                   
[69] BiocVersion_3.12.0            evaluate_0.14                
[71] BiocManager_1.30.10           vctrs_0.3.0                  
[73] httpuv_1.5.2                  gtable_0.3.0                 
[75] purrr_0.3.4                   assertthat_0.2.1             
[77] xfun_0.13                     rsvd_1.0.3                   
[79] mime_0.9                      xtable_1.8-4                 
[81] later_1.0.0                   viridisLite_0.3.0            
[83] pheatmap_1.0.12               tibble_3.0.1                 
[85] beeswarm_0.2.3                AnnotationDbi_1.51.0         
[87] memoise_1.1.0                 ellipsis_0.3.1               
[89] interactiveDisplayBase_1.27.0</code></pre>
</div>

</div>
</div>



<h3> Bibliography</h3>
<div id="refs" class="references">
<div id="ref-grun2016denovo">
<p>Grun, D., M. J. Muraro, J. C. Boisset, K. Wiebrands, A. Lyubimova, G. Dharmadhikari, M. van den Born, et al. 2016. “De Novo Prediction of Stem Cell Identity using Single-Cell Transcriptome Data.” <em>Cell Stem Cell</em> 19 (2): 266–77.</p>
</div>
<div id="ref-muraro2016singlecell">
<p>Muraro, M. J., G. Dharmadhikari, D. Grun, N. Groen, T. Dielen, E. Jansen, L. van Gurp, et al. 2016. “A Single-Cell Transcriptome Atlas of the Human Pancreas.” <em>Cell Syst</em> 3 (4): 385–94.</p>
</div>
</div>
            </section>

          </div>
        </div>
      </div>
<a href="using-single-cell-references.html" class="navigation navigation-prev " aria-label="Previous page"><i class="fa fa-angle-left"></i></a>
<a href="using-multiple-references.html" class="navigation navigation-next " aria-label="Next page"><i class="fa fa-angle-right"></i></a>
    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/lunr.js"></script>
<script src="libs/gitbook-2.6.7/js/clipboard.min.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-clipboard.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": {
"github": false,
"facebook": false,
"twitter": false,
"linkedin": false,
"weibo": false,
"instapaper": false,
"vk": false,
"all": ["facebook", "twitter", "linkedin", "weibo", "instapaper"]
},
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": {
"link": null,
"text": null
},
"history": {
"link": null,
"text": null
},
"view": {
"link": null,
"text": null
},
"download": null,
"toc": {
"collapse": "subsection"
},
"bookdown::epub_book": "default",
"cover": "images/cover.png"
});
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    var src = "";
    if (src === "" || src === "true") src = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-MML-AM_CHTML";
    if (location.protocol !== "file:")
      if (/^https?:/.test(src))
        src = src.replace(/^https?:/, '');
    script.src = src;
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>
</body>

</html>
