<!DOCTYPE html>
<html lang="" xml:lang="">
<head>

  <meta charset="utf-8" />
  <meta http-equiv="X-UA-Compatible" content="IE=edge" />
  <title>Chapter 8 Cross-annotating human pancreas | Assigning cell types with SingleR</title>
  <meta name="description" content="The SingleR book. Because sometimes, a vignette just isn’t enough." />
  <meta name="generator" content="bookdown 0.20 and GitBook 2.6.7" />

  <meta property="og:title" content="Chapter 8 Cross-annotating human pancreas | Assigning cell types with SingleR" />
  <meta property="og:type" content="book" />
  
  
  <meta property="og:description" content="The SingleR book. Because sometimes, a vignette just isn’t enough." />
  

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="Chapter 8 Cross-annotating human pancreas | Assigning cell types with SingleR" />
  
  <meta name="twitter:description" content="The SingleR book. Because sometimes, a vignette just isn’t enough." />
  



<meta name="date" content="2020-07-09" />

  <meta name="viewport" content="width=device-width, initial-scale=1" />
  <meta name="apple-mobile-web-app-capable" content="yes" />
  <meta name="apple-mobile-web-app-status-bar-style" content="black" />
  
  
<link rel="prev" href="advanced-options.html"/>
<link rel="next" href="cross-annotating-mouse-brains.html"/>
<script src="libs/header-attrs-2.3/header-attrs.js"></script>
<script src="libs/jquery-2.2.3/jquery.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-table.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-clipboard.css" rel="stylesheet" />











<style type="text/css">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>

<link rel="stylesheet" href="style.css" type="text/css" />
</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li><a href="./">The SingleR book</a></li>

<li class="divider"></li>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html"><i class="fa fa-check"></i>Preface</a></li>
<li class="part"><span><b>I Basic usage</b></span></li>
<li class="chapter" data-level="1" data-path="introduction.html"><a href="introduction.html"><i class="fa fa-check"></i><b>1</b> Introduction</a>
<ul>
<li class="chapter" data-level="1.1" data-path="introduction.html"><a href="introduction.html#motivation"><i class="fa fa-check"></i><b>1.1</b> Motivation</a></li>
<li class="chapter" data-level="1.2" data-path="introduction.html"><a href="introduction.html#method-description"><i class="fa fa-check"></i><b>1.2</b> Method description</a></li>
<li class="chapter" data-level="1.3" data-path="introduction.html"><a href="introduction.html#quick-start"><i class="fa fa-check"></i><b>1.3</b> Quick start</a></li>
<li class="chapter" data-level="1.4" data-path="introduction.html"><a href="introduction.html#where-to-get-help"><i class="fa fa-check"></i><b>1.4</b> Where to get help</a></li>
<li class="chapter" data-level="" data-path="introduction.html"><a href="introduction.html#session-information"><i class="fa fa-check"></i>Session information</a></li>
</ul></li>
<li class="chapter" data-level="2" data-path="using-the-classic-mode.html"><a href="using-the-classic-mode.html"><i class="fa fa-check"></i><b>2</b> Using the classic mode</a>
<ul>
<li class="chapter" data-level="2.1" data-path="using-the-classic-mode.html"><a href="using-the-classic-mode.html#overview"><i class="fa fa-check"></i><b>2.1</b> Overview</a></li>
<li class="chapter" data-level="2.2" data-path="using-the-classic-mode.html"><a href="using-the-classic-mode.html#annotation-the-test-dataset"><i class="fa fa-check"></i><b>2.2</b> Annotation the test dataset</a></li>
<li class="chapter" data-level="2.3" data-path="using-the-classic-mode.html"><a href="using-the-classic-mode.html#interaction-with-quality-control"><i class="fa fa-check"></i><b>2.3</b> Interaction with quality control</a></li>
<li class="chapter" data-level="2.4" data-path="using-the-classic-mode.html"><a href="using-the-classic-mode.html#choices-of-assay-data"><i class="fa fa-check"></i><b>2.4</b> Choices of assay data</a></li>
<li class="chapter" data-level="2.5" data-path="using-the-classic-mode.html"><a href="using-the-classic-mode.html#reference-choice"><i class="fa fa-check"></i><b>2.5</b> Comments on choice of references</a></li>
<li class="chapter" data-level="" data-path="introduction.html"><a href="introduction.html#session-information"><i class="fa fa-check"></i>Session information</a></li>
</ul></li>
<li class="chapter" data-level="3" data-path="more-markers.html"><a href="more-markers.html"><i class="fa fa-check"></i><b>3</b> Controlling marker detection</a>
<ul>
<li class="chapter" data-level="3.1" data-path="using-the-classic-mode.html"><a href="using-the-classic-mode.html#overview"><i class="fa fa-check"></i><b>3.1</b> Overview</a></li>
<li class="chapter" data-level="3.2" data-path="more-markers.html"><a href="more-markers.html#annotation-with-test-based-marker-detection"><i class="fa fa-check"></i><b>3.2</b> Annotation with test-based marker detection</a></li>
<li class="chapter" data-level="3.3" data-path="more-markers.html"><a href="more-markers.html#defining-custom-markers"><i class="fa fa-check"></i><b>3.3</b> Defining custom markers</a></li>
<li class="chapter" data-level="3.4" data-path="more-markers.html"><a href="more-markers.html#pseudo-bulk-aggregation"><i class="fa fa-check"></i><b>3.4</b> Pseudo-bulk aggregation</a></li>
<li class="chapter" data-level="" data-path="introduction.html"><a href="introduction.html#session-information"><i class="fa fa-check"></i>Session information</a></li>
</ul></li>
<li class="chapter" data-level="4" data-path="annotation-diagnostics.html"><a href="annotation-diagnostics.html"><i class="fa fa-check"></i><b>4</b> Annotation diagnostics</a>
<ul>
<li class="chapter" data-level="4.1" data-path="using-the-classic-mode.html"><a href="using-the-classic-mode.html#overview"><i class="fa fa-check"></i><b>4.1</b> Overview</a></li>
<li class="chapter" data-level="4.2" data-path="annotation-diagnostics.html"><a href="annotation-diagnostics.html#based-on-the-scores-within-cells"><i class="fa fa-check"></i><b>4.2</b> Based on the scores within cells</a></li>
<li class="chapter" data-level="4.3" data-path="annotation-diagnostics.html"><a href="annotation-diagnostics.html#based-on-the-deltas-across-cells"><i class="fa fa-check"></i><b>4.3</b> Based on the deltas across cells</a></li>
<li class="chapter" data-level="4.4" data-path="annotation-diagnostics.html"><a href="annotation-diagnostics.html#based-on-marker-gene-expression"><i class="fa fa-check"></i><b>4.4</b> Based on marker gene expression</a></li>
<li class="chapter" data-level="" data-path="introduction.html"><a href="introduction.html#session-information"><i class="fa fa-check"></i>Session information</a></li>
</ul></li>
<li class="part"><span><b>II Advanced usage</b></span></li>
<li class="chapter" data-level="5" data-path="using-multiple-references.html"><a href="using-multiple-references.html"><i class="fa fa-check"></i><b>5</b> Using multiple references</a>
<ul>
<li class="chapter" data-level="5.1" data-path="using-the-classic-mode.html"><a href="using-the-classic-mode.html#overview"><i class="fa fa-check"></i><b>5.1</b> Overview</a></li>
<li class="chapter" data-level="5.2" data-path="using-multiple-references.html"><a href="using-multiple-references.html#using-reference-specific-labels"><i class="fa fa-check"></i><b>5.2</b> Using reference-specific labels</a></li>
<li class="chapter" data-level="5.3" data-path="using-multiple-references.html"><a href="using-multiple-references.html#comparing-scores-across-references"><i class="fa fa-check"></i><b>5.3</b> Comparing scores across references</a>
<ul>
<li class="chapter" data-level="5.3.1" data-path="using-multiple-references.html"><a href="using-multiple-references.html#combining-inferences-from-individual-references"><i class="fa fa-check"></i><b>5.3.1</b> Combining inferences from individual references</a></li>
<li class="chapter" data-level="5.3.2" data-path="using-multiple-references.html"><a href="using-multiple-references.html#combined-diagnostics"><i class="fa fa-check"></i><b>5.3.2</b> Combined diagnostics</a></li>
</ul></li>
<li class="chapter" data-level="5.4" data-path="using-multiple-references.html"><a href="using-multiple-references.html#using-harmonized-labels"><i class="fa fa-check"></i><b>5.4</b> Using harmonized labels</a>
<ul>
<li class="chapter" data-level="5.4.1" data-path="using-multiple-references.html"><a href="using-multiple-references.html#sharing-information-during-marker-detection"><i class="fa fa-check"></i><b>5.4.1</b> Sharing information during marker detection</a></li>
<li class="chapter" data-level="5.4.2" data-path="using-multiple-references.html"><a href="using-multiple-references.html#manual-label-harmonization"><i class="fa fa-check"></i><b>5.4.2</b> Manual label harmonization</a></li>
</ul></li>
<li class="chapter" data-level="" data-path="using-multiple-references.html"><a href="using-multiple-references.html#session-info"><i class="fa fa-check"></i>Session info</a></li>
</ul></li>
<li class="chapter" data-level="6" data-path="exploiting-the-cell-ontology.html"><a href="exploiting-the-cell-ontology.html"><i class="fa fa-check"></i><b>6</b> Exploiting the cell ontology</a>
<ul>
<li class="chapter" data-level="6.1" data-path="introduction.html"><a href="introduction.html#motivation"><i class="fa fa-check"></i><b>6.1</b> Motivation</a></li>
<li class="chapter" data-level="6.2" data-path="exploiting-the-cell-ontology.html"><a href="exploiting-the-cell-ontology.html#basic-manipulation"><i class="fa fa-check"></i><b>6.2</b> Basic manipulation</a></li>
<li class="chapter" data-level="6.3" data-path="exploiting-the-cell-ontology.html"><a href="exploiting-the-cell-ontology.html#adjusting-resolution"><i class="fa fa-check"></i><b>6.3</b> Adjusting resolution</a></li>
<li class="chapter" data-level="" data-path="introduction.html"><a href="introduction.html#session-information"><i class="fa fa-check"></i>Session information</a></li>
</ul></li>
<li class="chapter" data-level="7" data-path="advanced-options.html"><a href="advanced-options.html"><i class="fa fa-check"></i><b>7</b> Advanced options</a>
<ul>
<li class="chapter" data-level="7.1" data-path="advanced-options.html"><a href="advanced-options.html#preconstructed-indices"><i class="fa fa-check"></i><b>7.1</b> Preconstructed indices</a></li>
<li class="chapter" data-level="7.2" data-path="advanced-options.html"><a href="advanced-options.html#parallelization"><i class="fa fa-check"></i><b>7.2</b> Parallelization</a></li>
<li class="chapter" data-level="7.3" data-path="advanced-options.html"><a href="advanced-options.html#approximate-algorithms"><i class="fa fa-check"></i><b>7.3</b> Approximate algorithms</a></li>
<li class="chapter" data-level="" data-path="introduction.html"><a href="introduction.html#session-information"><i class="fa fa-check"></i>Session information</a></li>
</ul></li>
<li class="part"><span><b>III Case studies</b></span></li>
<li class="chapter" data-level="8" data-path="cross-annotating-human-pancreas.html"><a href="cross-annotating-human-pancreas.html"><i class="fa fa-check"></i><b>8</b> Cross-annotating human pancreas</a>
<ul>
<li class="chapter" data-level="8.1" data-path="cross-annotating-human-pancreas.html"><a href="cross-annotating-human-pancreas.html#loading-the-data"><i class="fa fa-check"></i><b>8.1</b> Loading the data</a></li>
<li class="chapter" data-level="8.2" data-path="cross-annotating-human-pancreas.html"><a href="cross-annotating-human-pancreas.html#applying-the-annotation"><i class="fa fa-check"></i><b>8.2</b> Applying the annotation</a></li>
<li class="chapter" data-level="8.3" data-path="cross-annotating-human-pancreas.html"><a href="cross-annotating-human-pancreas.html#diagnostics"><i class="fa fa-check"></i><b>8.3</b> Diagnostics</a></li>
<li class="chapter" data-level="8.4" data-path="cross-annotating-human-pancreas.html"><a href="cross-annotating-human-pancreas.html#comparison-to-clusters"><i class="fa fa-check"></i><b>8.4</b> Comparison to clusters</a></li>
<li class="chapter" data-level="" data-path="introduction.html"><a href="introduction.html#session-information"><i class="fa fa-check"></i>Session information</a></li>
</ul></li>
<li class="chapter" data-level="9" data-path="cross-annotating-mouse-brains.html"><a href="cross-annotating-mouse-brains.html"><i class="fa fa-check"></i><b>9</b> Cross-annotating mouse brains</a>
<ul>
<li class="chapter" data-level="9.1" data-path="cross-annotating-human-pancreas.html"><a href="cross-annotating-human-pancreas.html#loading-the-data"><i class="fa fa-check"></i><b>9.1</b> Loading the data</a></li>
<li class="chapter" data-level="9.2" data-path="cross-annotating-human-pancreas.html"><a href="cross-annotating-human-pancreas.html#applying-the-annotation"><i class="fa fa-check"></i><b>9.2</b> Applying the annotation</a></li>
<li class="chapter" data-level="9.3" data-path="cross-annotating-human-pancreas.html"><a href="cross-annotating-human-pancreas.html#diagnostics"><i class="fa fa-check"></i><b>9.3</b> Diagnostics</a></li>
<li class="chapter" data-level="9.4" data-path="cross-annotating-human-pancreas.html"><a href="cross-annotating-human-pancreas.html#comparison-to-clusters"><i class="fa fa-check"></i><b>9.4</b> Comparison to clusters</a></li>
<li class="chapter" data-level="" data-path="introduction.html"><a href="introduction.html#session-information"><i class="fa fa-check"></i>Session information</a></li>
</ul></li>
<li class="part"><span><b>IV Appendix</b></span></li>
<li class="chapter" data-level="10" data-path="contributors.html"><a href="contributors.html"><i class="fa fa-check"></i><b>10</b> Contributors</a>
<ul>
<li><a href="contributors.html#aaron-lun"><em>Aaron Lun</em></a></li>
</ul></li>
<li class="chapter" data-level="11" data-path="bibliography.html"><a href="bibliography.html"><i class="fa fa-check"></i><b>11</b> Bibliography</a></li>
<li class="divider"></li>
<li><a href="https://bioconductor.org" target="blank">Published by Bioconductor</a></li>

</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">Assigning cell types with SingleR</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="cross-annotating-human-pancreas" class="section level1" number="8">
<h1><span class="header-section-number">Chapter 8</span> Cross-annotating human pancreas</h1>
<script>
document.addEventListener("click", function (event) {
    if (event.target.classList.contains("aaron-collapse")) {
        event.target.classList.toggle("active");
        var content = event.target.nextElementSibling;
        if (content.style.display === "block") {
            content.style.display = "none";
        } else {
            content.style.display = "block";
        }
    }
})
</script>
<style>
.aaron-collapse {
  background-color: #eee;
  color: #444;
  cursor: pointer;
  padding: 18px;
  width: 100%;
  border: none;
  text-align: left;
  outline: none;
  font-size: 15px;
}

.aaron-content {
  padding: 0 18px;
  display: none;
  overflow: hidden;
  background-color: #f1f1f1;
}
</style>
<div id="loading-the-data" class="section level2" number="8.1">
<h2><span class="header-section-number">8.1</span> Loading the data</h2>
<p>We load the <span class="citation">Muraro et al. (<a href="#ref-muraro2016singlecell" role="doc-biblioref">2016</a>)</span> dataset as our reference, removing unlabelled cells or cells without a clear label.</p>
<div class="sourceCode" id="cb128"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb128-1"><a href="cross-annotating-human-pancreas.html#cb128-1" aria-hidden="true"></a><span class="kw">library</span>(scRNAseq)</span>
<span id="cb128-2"><a href="cross-annotating-human-pancreas.html#cb128-2" aria-hidden="true"></a>sceM &lt;-<span class="st"> </span><span class="kw">MuraroPancreasData</span>()</span>
<span id="cb128-3"><a href="cross-annotating-human-pancreas.html#cb128-3" aria-hidden="true"></a>sceM &lt;-<span class="st"> </span>sceM[,<span class="op">!</span><span class="kw">is.na</span>(sceM<span class="op">$</span>label) <span class="op">&amp;</span><span class="st"> </span>sceM<span class="op">$</span>label<span class="op">!=</span><span class="st">&quot;unclear&quot;</span>] </span></code></pre></div>
<p>We compute log-expression values for use in marker detection inside <code>SingleR()</code>.</p>
<div class="sourceCode" id="cb129"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb129-1"><a href="cross-annotating-human-pancreas.html#cb129-1" aria-hidden="true"></a><span class="kw">library</span>(scater)</span>
<span id="cb129-2"><a href="cross-annotating-human-pancreas.html#cb129-2" aria-hidden="true"></a>sceM &lt;-<span class="st"> </span><span class="kw">logNormCounts</span>(sceM)</span></code></pre></div>
<p>We examine the distribution of labels in this reference.</p>
<div class="sourceCode" id="cb130"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb130-1"><a href="cross-annotating-human-pancreas.html#cb130-1" aria-hidden="true"></a><span class="kw">table</span>(sceM<span class="op">$</span>label)</span></code></pre></div>
<pre><code>## 
##      acinar       alpha        beta       delta        duct endothelial 
##         219         812         448         193         245          21 
##     epsilon mesenchymal          pp 
##           3          80         101</code></pre>
<p>We load the <span class="citation">Grun et al. (<a href="#ref-grun2016denovo" role="doc-biblioref">2016</a>)</span> dataset as our test,
applying some basic quality control to remove low-quality cells in some of the batches
(see <a href="https://osca.bioconductor.org/grun-human-pancreas-cel-seq2.html#quality-control-8">here</a> for details).</p>
<div class="sourceCode" id="cb132"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb132-1"><a href="cross-annotating-human-pancreas.html#cb132-1" aria-hidden="true"></a>sceG &lt;-<span class="st"> </span><span class="kw">GrunPancreasData</span>()</span>
<span id="cb132-2"><a href="cross-annotating-human-pancreas.html#cb132-2" aria-hidden="true"></a></span>
<span id="cb132-3"><a href="cross-annotating-human-pancreas.html#cb132-3" aria-hidden="true"></a>sceG &lt;-<span class="st"> </span><span class="kw">addPerCellQC</span>(sceG)</span>
<span id="cb132-4"><a href="cross-annotating-human-pancreas.html#cb132-4" aria-hidden="true"></a>qc &lt;-<span class="st"> </span><span class="kw">quickPerCellQC</span>(<span class="kw">colData</span>(sceG), </span>
<span id="cb132-5"><a href="cross-annotating-human-pancreas.html#cb132-5" aria-hidden="true"></a>    <span class="dt">percent_subsets=</span><span class="st">&quot;altexps_ERCC_percent&quot;</span>,</span>
<span id="cb132-6"><a href="cross-annotating-human-pancreas.html#cb132-6" aria-hidden="true"></a>    <span class="dt">batch=</span>sceG<span class="op">$</span>donor,</span>
<span id="cb132-7"><a href="cross-annotating-human-pancreas.html#cb132-7" aria-hidden="true"></a>    <span class="dt">subset=</span>sceG<span class="op">$</span>donor <span class="op">%in%</span><span class="st"> </span><span class="kw">c</span>(<span class="st">&quot;D17&quot;</span>, <span class="st">&quot;D7&quot;</span>, <span class="st">&quot;D2&quot;</span>))</span>
<span id="cb132-8"><a href="cross-annotating-human-pancreas.html#cb132-8" aria-hidden="true"></a>sceG &lt;-<span class="st"> </span>sceG[,<span class="op">!</span>qc<span class="op">$</span>discard]</span></code></pre></div>
<p>Technically speaking, the test dataset does not need log-expression values but we compute them anyway for convenience.</p>
<div class="sourceCode" id="cb133"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb133-1"><a href="cross-annotating-human-pancreas.html#cb133-1" aria-hidden="true"></a>sceG &lt;-<span class="st"> </span><span class="kw">logNormCounts</span>(sceG)</span></code></pre></div>
</div>
<div id="applying-the-annotation" class="section level2" number="8.2">
<h2><span class="header-section-number">8.2</span> Applying the annotation</h2>
<p>We apply <code>SingleR()</code> with Wilcoxon rank sum test-based marker detection to annotate the Grun dataset with the Muraro labels.</p>
<div class="sourceCode" id="cb134"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb134-1"><a href="cross-annotating-human-pancreas.html#cb134-1" aria-hidden="true"></a><span class="kw">library</span>(SingleR)</span>
<span id="cb134-2"><a href="cross-annotating-human-pancreas.html#cb134-2" aria-hidden="true"></a>pred.grun &lt;-<span class="st"> </span><span class="kw">SingleR</span>(<span class="dt">test=</span>sceG, <span class="dt">ref=</span>sceM, <span class="dt">labels=</span>sceM<span class="op">$</span>label, <span class="dt">de.method=</span><span class="st">&quot;wilcox&quot;</span>)</span></code></pre></div>
<p>We examine the distribution of predicted labels:</p>
<div class="sourceCode" id="cb135"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb135-1"><a href="cross-annotating-human-pancreas.html#cb135-1" aria-hidden="true"></a><span class="kw">table</span>(pred.grun<span class="op">$</span>labels)</span></code></pre></div>
<pre><code>## 
##      acinar       alpha        beta       delta        duct endothelial 
##         277         203         181          50         306           5 
##     epsilon mesenchymal          pp 
##           1          22          19</code></pre>
<p>We can also examine the number of discarded cells for each label:</p>
<div class="sourceCode" id="cb137"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb137-1"><a href="cross-annotating-human-pancreas.html#cb137-1" aria-hidden="true"></a><span class="kw">table</span>(<span class="dt">Label=</span>pred.grun<span class="op">$</span>labels,</span>
<span id="cb137-2"><a href="cross-annotating-human-pancreas.html#cb137-2" aria-hidden="true"></a>    <span class="dt">Lost=</span><span class="kw">is.na</span>(pred.grun<span class="op">$</span>pruned.labels))</span></code></pre></div>
<pre><code>##              Lost
## Label         FALSE TRUE
##   acinar        251   26
##   alpha         198    5
##   beta          180    1
##   delta          49    1
##   duct          301    5
##   endothelial     4    1
##   epsilon         1    0
##   mesenchymal    22    0
##   pp             17    2</code></pre>
</div>
<div id="diagnostics" class="section level2" number="8.3">
<h2><span class="header-section-number">8.3</span> Diagnostics</h2>
<p>We visualize the assignment scores for each label in Figure <a href="cross-annotating-human-pancreas.html#fig:unref-pancreas-score-heatmap">8.1</a>.</p>
<div class="sourceCode" id="cb139"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb139-1"><a href="cross-annotating-human-pancreas.html#cb139-1" aria-hidden="true"></a><span class="kw">plotScoreHeatmap</span>(pred.grun)</span></code></pre></div>
<div class="figure"><span id="fig:unref-pancreas-score-heatmap"></span>
<img src="pancreas_files/figure-html/unref-pancreas-score-heatmap-1.png" alt="Heatmap of the (normalized) assignment scores for each cell (column) in the Grun test dataset with respect to each label (row) in the Muraro reference dataset. The final assignment for each cell is shown in the annotation bar at the top." width="672" />
<p class="caption">
Figure 8.1: Heatmap of the (normalized) assignment scores for each cell (column) in the Grun test dataset with respect to each label (row) in the Muraro reference dataset. The final assignment for each cell is shown in the annotation bar at the top.
</p>
</div>
<p>The delta for each cell is visualized in Figure <a href="cross-annotating-human-pancreas.html#fig:unref-pancreas-delta-dist">8.2</a>.</p>
<div class="sourceCode" id="cb140"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb140-1"><a href="cross-annotating-human-pancreas.html#cb140-1" aria-hidden="true"></a><span class="kw">plotDeltaDistribution</span>(pred.grun)</span></code></pre></div>
<div class="figure"><span id="fig:unref-pancreas-delta-dist"></span>
<img src="pancreas_files/figure-html/unref-pancreas-delta-dist-1.png" alt="Distributions of the deltas for each cell in the Grun dataset assigned to each label in the Muraro dataset. Each cell is represented by a point; low-quality assignments that were pruned out are colored in orange." width="672" />
<p class="caption">
Figure 8.2: Distributions of the deltas for each cell in the Grun dataset assigned to each label in the Muraro dataset. Each cell is represented by a point; low-quality assignments that were pruned out are colored in orange.
</p>
</div>
<p>Finally, we visualize the heatmaps of the marker genes for each label in Figure <a href="cross-annotating-human-pancreas.html#fig:unref-pancreas-marker-heat">8.3</a>.</p>
<div class="sourceCode" id="cb141"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb141-1"><a href="cross-annotating-human-pancreas.html#cb141-1" aria-hidden="true"></a><span class="kw">library</span>(scater)</span>
<span id="cb141-2"><a href="cross-annotating-human-pancreas.html#cb141-2" aria-hidden="true"></a>collected &lt;-<span class="st"> </span><span class="kw">list</span>()</span>
<span id="cb141-3"><a href="cross-annotating-human-pancreas.html#cb141-3" aria-hidden="true"></a>all.markers &lt;-<span class="st"> </span><span class="kw">metadata</span>(pred.grun)<span class="op">$</span>de.genes</span>
<span id="cb141-4"><a href="cross-annotating-human-pancreas.html#cb141-4" aria-hidden="true"></a></span>
<span id="cb141-5"><a href="cross-annotating-human-pancreas.html#cb141-5" aria-hidden="true"></a>sceG<span class="op">$</span>labels &lt;-<span class="st"> </span>pred.grun<span class="op">$</span>labels</span>
<span id="cb141-6"><a href="cross-annotating-human-pancreas.html#cb141-6" aria-hidden="true"></a><span class="cf">for</span> (lab <span class="cf">in</span> <span class="kw">unique</span>(pred.grun<span class="op">$</span>labels)) {</span>
<span id="cb141-7"><a href="cross-annotating-human-pancreas.html#cb141-7" aria-hidden="true"></a>    collected[[lab]] &lt;-<span class="st"> </span><span class="kw">plotHeatmap</span>(sceG, <span class="dt">silent=</span><span class="ot">TRUE</span>, </span>
<span id="cb141-8"><a href="cross-annotating-human-pancreas.html#cb141-8" aria-hidden="true"></a>        <span class="dt">order_columns_by=</span><span class="st">&quot;labels&quot;</span>, <span class="dt">main=</span>lab,</span>
<span id="cb141-9"><a href="cross-annotating-human-pancreas.html#cb141-9" aria-hidden="true"></a>        <span class="dt">features=</span><span class="kw">unique</span>(<span class="kw">unlist</span>(all.markers[[lab]])))[[<span class="dv">4</span>]] </span>
<span id="cb141-10"><a href="cross-annotating-human-pancreas.html#cb141-10" aria-hidden="true"></a>}</span>
<span id="cb141-11"><a href="cross-annotating-human-pancreas.html#cb141-11" aria-hidden="true"></a><span class="kw">do.call</span>(gridExtra<span class="op">::</span>grid.arrange, collected)</span></code></pre></div>
<div class="figure"><span id="fig:unref-pancreas-marker-heat"></span>
<img src="pancreas_files/figure-html/unref-pancreas-marker-heat-1.png" alt="Heatmaps of log-expression values in the Grun dataset for all marker genes upregulated in each label in the Muraro reference dataset. Assigned labels for each cell are shown at the top of each plot." width="1920" />
<p class="caption">
Figure 8.3: Heatmaps of log-expression values in the Grun dataset for all marker genes upregulated in each label in the Muraro reference dataset. Assigned labels for each cell are shown at the top of each plot.
</p>
</div>
</div>
<div id="comparison-to-clusters" class="section level2" number="8.4">
<h2><span class="header-section-number">8.4</span> Comparison to clusters</h2>
<p>For comparison, we will perform a quick unsupervised analysis of the Grun dataset.
We model the variances using the spike-in data and we perform graph-based clustering
(increasing the resolution by dropping <code>k=5</code>).</p>
<div class="sourceCode" id="cb142"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb142-1"><a href="cross-annotating-human-pancreas.html#cb142-1" aria-hidden="true"></a><span class="kw">library</span>(scran)</span>
<span id="cb142-2"><a href="cross-annotating-human-pancreas.html#cb142-2" aria-hidden="true"></a>decG &lt;-<span class="st"> </span><span class="kw">modelGeneVarWithSpikes</span>(sceG, <span class="st">&quot;ERCC&quot;</span>)</span>
<span id="cb142-3"><a href="cross-annotating-human-pancreas.html#cb142-3" aria-hidden="true"></a></span>
<span id="cb142-4"><a href="cross-annotating-human-pancreas.html#cb142-4" aria-hidden="true"></a><span class="kw">set.seed</span>(<span class="dv">1000100</span>)</span>
<span id="cb142-5"><a href="cross-annotating-human-pancreas.html#cb142-5" aria-hidden="true"></a>sceG &lt;-<span class="st"> </span><span class="kw">denoisePCA</span>(sceG, decG)</span>
<span id="cb142-6"><a href="cross-annotating-human-pancreas.html#cb142-6" aria-hidden="true"></a>sceG<span class="op">$</span>cluster &lt;-<span class="st"> </span><span class="kw">clusterSNNGraph</span>(sceG, <span class="dt">k=</span><span class="dv">5</span>, <span class="dt">use.dimred=</span><span class="st">&quot;PCA&quot;</span>)</span></code></pre></div>
<p>We see that the clusters map well to the labels in Figure <a href="cross-annotating-human-pancreas.html#fig:unref-pancreas-label-clusters">8.4</a>.</p>
<div class="sourceCode" id="cb143"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb143-1"><a href="cross-annotating-human-pancreas.html#cb143-1" aria-hidden="true"></a>tab &lt;-<span class="st"> </span><span class="kw">table</span>(<span class="dt">cluster=</span>sceG<span class="op">$</span>cluster, <span class="dt">label=</span>pred.grun<span class="op">$</span>labels) </span>
<span id="cb143-2"><a href="cross-annotating-human-pancreas.html#cb143-2" aria-hidden="true"></a>pheatmap<span class="op">::</span><span class="kw">pheatmap</span>(<span class="kw">log10</span>(tab<span class="op">+</span><span class="dv">10</span>))</span></code></pre></div>
<div class="figure"><span id="fig:unref-pancreas-label-clusters"></span>
<img src="pancreas_files/figure-html/unref-pancreas-label-clusters-1.png" alt="Heatmap of the log-transformed number of cells in each combination of label (column) and cluster (row) in the Grun dataset." width="672" />
<p class="caption">
Figure 8.4: Heatmap of the log-transformed number of cells in each combination of label (column) and cluster (row) in the Grun dataset.
</p>
</div>
<p>We proceed to the most important part of the analysis.
Yes, that’s right, the <span class="math inline">\(t\)</span>-SNE plot (Figure <a href="cross-annotating-human-pancreas.html#fig:unref-pancreas-label-tsne">8.5</a>).</p>
<div class="sourceCode" id="cb144"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb144-1"><a href="cross-annotating-human-pancreas.html#cb144-1" aria-hidden="true"></a><span class="kw">set.seed</span>(<span class="dv">101010100</span>)</span>
<span id="cb144-2"><a href="cross-annotating-human-pancreas.html#cb144-2" aria-hidden="true"></a>sceG &lt;-<span class="st"> </span><span class="kw">runTSNE</span>(sceG, <span class="dt">dimred=</span><span class="st">&quot;PCA&quot;</span>)</span>
<span id="cb144-3"><a href="cross-annotating-human-pancreas.html#cb144-3" aria-hidden="true"></a><span class="kw">plotTSNE</span>(sceG, <span class="dt">colour_by=</span><span class="st">&quot;cluster&quot;</span>, <span class="dt">text_colour=</span><span class="st">&quot;red&quot;</span>,</span>
<span id="cb144-4"><a href="cross-annotating-human-pancreas.html#cb144-4" aria-hidden="true"></a>    <span class="dt">text_by=</span><span class="kw">I</span>(pred.grun<span class="op">$</span>labels))</span></code></pre></div>
<div class="figure"><span id="fig:unref-pancreas-label-tsne"></span>
<img src="pancreas_files/figure-html/unref-pancreas-label-tsne-1.png" alt="$t$-SNE plot of the Grun dataset, where each point is a cell and is colored by the assigned cluster. Reference labels from the Muraro dataset are also placed on the median coordinate across all cells assigned with that label." width="672" />
<p class="caption">
Figure 8.5: <span class="math inline">\(t\)</span>-SNE plot of the Grun dataset, where each point is a cell and is colored by the assigned cluster. Reference labels from the Muraro dataset are also placed on the median coordinate across all cells assigned with that label.
</p>
</div>
</div>
<div id="session-information" class="section level2 unnumbered">
<h2>Session information</h2>
<button class="aaron-collapse">
View session info
</button>
<div class="aaron-content">
<pre><code>R version 4.0.0 Patched (2020-05-01 r78341)
Platform: x86_64-pc-linux-gnu (64-bit)
Running under: Ubuntu 18.04.4 LTS

Matrix products: default
BLAS:   /home/luna/Software/R/R-4-0-branch-dev/lib/libRblas.so
LAPACK: /home/luna/Software/R/R-4-0-branch-dev/lib/libRlapack.so

locale:
 [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              
 [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    
 [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   
 [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 
 [9] LC_ADDRESS=C               LC_TELEPHONE=C            
[11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       

attached base packages:
[1] parallel  stats4    stats     graphics  grDevices utils     datasets 
[8] methods   base     

other attached packages:
 [1] scran_1.17.4                SingleR_1.3.6              
 [3] scater_1.17.3               ggplot2_3.3.2              
 [5] scRNAseq_2.3.6              SingleCellExperiment_1.11.6
 [7] SummarizedExperiment_1.19.5 DelayedArray_0.15.6        
 [9] matrixStats_0.56.0          Matrix_1.2-18              
[11] Biobase_2.49.0              GenomicRanges_1.41.5       
[13] GenomeInfoDb_1.25.5         IRanges_2.23.10            
[15] S4Vectors_0.27.12           BiocGenerics_0.35.4        
[17] BiocStyle_2.17.0            rebook_0.99.0              

loaded via a namespace (and not attached):
 [1] bitops_1.0-6                  bit64_0.9-7                  
 [3] RColorBrewer_1.1-2            httr_1.4.1                   
 [5] tools_4.0.0                   R6_2.4.1                     
 [7] irlba_2.3.3                   vipor_0.4.5                  
 [9] DBI_1.1.0                     colorspace_1.4-1             
[11] withr_2.2.0                   gridExtra_2.3                
[13] tidyselect_1.1.0              processx_3.4.2               
[15] bit_1.1-15.2                  curl_4.3                     
[17] compiler_4.0.0                graph_1.67.1                 
[19] BiocNeighbors_1.7.0           labeling_0.3                 
[21] bookdown_0.20                 scales_1.1.1                 
[23] callr_3.4.3                   rappdirs_0.3.1               
[25] stringr_1.4.0                 digest_0.6.25                
[27] rmarkdown_2.3                 XVector_0.29.3               
[29] pkgconfig_2.0.3               htmltools_0.5.0              
[31] highr_0.8                     limma_3.45.7                 
[33] dbplyr_1.4.4                  fastmap_1.0.1                
[35] rlang_0.4.6                   RSQLite_2.2.0                
[37] shiny_1.5.0                   DelayedMatrixStats_1.11.1    
[39] farver_2.0.3                  generics_0.0.2               
[41] BiocParallel_1.23.0           dplyr_1.0.0                  
[43] RCurl_1.98-1.2                magrittr_1.5                 
[45] BiocSingular_1.5.0            GenomeInfoDbData_1.2.3       
[47] scuttle_0.99.10               Rcpp_1.0.4.6                 
[49] ggbeeswarm_0.6.0              munsell_0.5.0                
[51] viridis_0.5.1                 lifecycle_0.2.0              
[53] edgeR_3.31.4                  stringi_1.4.6                
[55] yaml_2.2.1                    zlibbioc_1.35.0              
[57] Rtsne_0.15                    BiocFileCache_1.13.0         
[59] AnnotationHub_2.21.1          grid_4.0.0                   
[61] blob_1.2.1                    dqrng_0.2.1                  
[63] promises_1.1.1                ExperimentHub_1.15.0         
[65] crayon_1.3.4                  lattice_0.20-41              
[67] cowplot_1.0.0                 locfit_1.5-9.4               
[69] CodeDepends_0.6.5             knitr_1.29                   
[71] ps_1.3.3                      pillar_1.4.4                 
[73] igraph_1.2.5                  codetools_0.2-16             
[75] XML_3.99-0.3                  glue_1.4.1                   
[77] BiocVersion_3.12.0            evaluate_0.14                
[79] BiocManager_1.30.10           vctrs_0.3.1                  
[81] httpuv_1.5.4                  gtable_0.3.0                 
[83] purrr_0.3.4                   assertthat_0.2.1             
[85] xfun_0.15                     rsvd_1.0.3                   
[87] mime_0.9                      xtable_1.8-4                 
[89] later_1.1.0.1                 viridisLite_0.3.0            
[91] pheatmap_1.0.12               tibble_3.0.1                 
[93] AnnotationDbi_1.51.1          beeswarm_0.2.3               
[95] memoise_1.1.0                 statmod_1.4.34               
[97] ellipsis_0.3.1                interactiveDisplayBase_1.27.5</code></pre>
</div>

</div>
</div>
<h3> Bibliography</h3>
<div id="refs" class="references">
<div id="ref-grun2016denovo">
<p>Grun, D., M. J. Muraro, J. C. Boisset, K. Wiebrands, A. Lyubimova, G. Dharmadhikari, M. van den Born, et al. 2016. “De Novo Prediction of Stem Cell Identity using Single-Cell Transcriptome Data.” <em>Cell Stem Cell</em> 19 (2): 266–77.</p>
</div>
<div id="ref-muraro2016singlecell">
<p>Muraro, M. J., G. Dharmadhikari, D. Grun, N. Groen, T. Dielen, E. Jansen, L. van Gurp, et al. 2016. “A Single-Cell Transcriptome Atlas of the Human Pancreas.” <em>Cell Syst</em> 3 (4): 385–94.</p>
</div>
</div>
            </section>

          </div>
        </div>
      </div>
<a href="advanced-options.html" class="navigation navigation-prev " aria-label="Previous page"><i class="fa fa-angle-left"></i></a>
<a href="cross-annotating-mouse-brains.html" class="navigation navigation-next " aria-label="Next page"><i class="fa fa-angle-right"></i></a>
    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/lunr.js"></script>
<script src="libs/gitbook-2.6.7/js/clipboard.min.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-clipboard.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": {
"github": false,
"facebook": false,
"twitter": false,
"linkedin": false,
"weibo": false,
"instapaper": false,
"vk": false,
"all": ["facebook", "twitter", "linkedin", "weibo", "instapaper"]
},
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": {
"link": "https://github.com/LTLA/SingleRBook-base/edit/master/pancreas.Rmd",
"text": "Edit"
},
"history": {
"link": "https://github.com/LTLA/SingleRBook-base/commits/master/pancreas.Rmd",
"text": null
},
"view": {
"link": "https://github.com/LTLA/SingleRBook-base/blob/master/pancreas.Rmd",
"text": null
},
"download": null,
"toc": {
"collapse": "subsection"
},
"bookdown::epub_book": "default",
"cover-image": "https://www.bioconductor.org/images/logo_bioconductor.gif",
"favicon": "https://bioc2019.bioconductor.org/images/favicon.ico"
});
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    var src = "true";
    if (src === "" || src === "true") src = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-MML-AM_CHTML";
    if (location.protocol !== "file:")
      if (/^https?:/.test(src))
        src = src.replace(/^https?:/, '');
    script.src = src;
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>
</body>

</html>
