<!DOCTYPE html>
<html lang="" xml:lang="">
<head>

  <meta charset="utf-8" />
  <meta http-equiv="X-UA-Compatible" content="IE=edge" />
  <title>Chapter 6 Exploiting the cell ontology | Assigning cell types with SingleR</title>
  <meta name="description" content="The SingleR book. Because sometimes, a vignette just isn’t enough." />
  <meta name="generator" content="bookdown 0.19 and GitBook 2.6.7" />

  <meta property="og:title" content="Chapter 6 Exploiting the cell ontology | Assigning cell types with SingleR" />
  <meta property="og:type" content="book" />
  
  
  <meta property="og:description" content="The SingleR book. Because sometimes, a vignette just isn’t enough." />
  

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="Chapter 6 Exploiting the cell ontology | Assigning cell types with SingleR" />
  
  <meta name="twitter:description" content="The SingleR book. Because sometimes, a vignette just isn’t enough." />
  



<meta name="date" content="2020-05-24" />

  <meta name="viewport" content="width=device-width, initial-scale=1" />
  <meta name="apple-mobile-web-app-capable" content="yes" />
  <meta name="apple-mobile-web-app-status-bar-style" content="black" />
  
  
<link rel="prev" href="using-multiple-references.html"/>
<link rel="next" href="advanced-options.html"/>
<script src="libs/jquery-2.2.3/jquery.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-table.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-clipboard.css" rel="stylesheet" />











<style type="text/css">
div.sourceCode { overflow-x: auto; }
table.sourceCode, tr.sourceCode, td.lineNumbers, td.sourceCode {
  margin: 0; padding: 0; vertical-align: baseline; border: none; }
table.sourceCode { width: 100%; line-height: 100%; background-color: #f8f8f8; }
td.lineNumbers { text-align: right; padding-right: 4px; padding-left: 4px; color: #aaaaaa; border-right: 1px solid #aaaaaa; }
td.sourceCode { padding-left: 5px; }
pre, code { background-color: #f8f8f8; }
code > span.kw { color: #204a87; font-weight: bold; } /* Keyword */
code > span.dt { color: #204a87; } /* DataType */
code > span.dv { color: #0000cf; } /* DecVal */
code > span.bn { color: #0000cf; } /* BaseN */
code > span.fl { color: #0000cf; } /* Float */
code > span.ch { color: #4e9a06; } /* Char */
code > span.st { color: #4e9a06; } /* String */
code > span.co { color: #8f5902; font-style: italic; } /* Comment */
code > span.ot { color: #8f5902; } /* Other */
code > span.al { color: #ef2929; } /* Alert */
code > span.fu { color: #000000; } /* Function */
code > span.er { color: #a40000; font-weight: bold; } /* Error */
code > span.wa { color: #8f5902; font-weight: bold; font-style: italic; } /* Warning */
code > span.cn { color: #000000; } /* Constant */
code > span.sc { color: #000000; } /* SpecialChar */
code > span.vs { color: #4e9a06; } /* VerbatimString */
code > span.ss { color: #4e9a06; } /* SpecialString */
code > span.im { } /* Import */
code > span.va { color: #000000; } /* Variable */
code > span.cf { color: #204a87; font-weight: bold; } /* ControlFlow */
code > span.op { color: #ce5c00; font-weight: bold; } /* Operator */
code > span.pp { color: #8f5902; font-style: italic; } /* Preprocessor */
code > span.ex { } /* Extension */
code > span.at { color: #c4a000; } /* Attribute */
code > span.do { color: #8f5902; font-weight: bold; font-style: italic; } /* Documentation */
code > span.an { color: #8f5902; font-weight: bold; font-style: italic; } /* Annotation */
code > span.cv { color: #8f5902; font-weight: bold; font-style: italic; } /* CommentVar */
code > span.in { color: #8f5902; font-weight: bold; font-style: italic; } /* Information */
</style>

<link rel="stylesheet" href="style.css" type="text/css" />
</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li><a href="./">The SingleR book</a></li>

<li class="divider"></li>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html"><i class="fa fa-check"></i>Preface</a></li>
<li class="part"><span><b>I Basic usage</b></span></li>
<li class="chapter" data-level="1" data-path="introduction.html"><a href="introduction.html"><i class="fa fa-check"></i><b>1</b> Introduction</a><ul>
<li class="chapter" data-level="1.1" data-path="introduction.html"><a href="introduction.html#motivation"><i class="fa fa-check"></i><b>1.1</b> Motivation</a></li>
<li class="chapter" data-level="1.2" data-path="introduction.html"><a href="introduction.html#method-description"><i class="fa fa-check"></i><b>1.2</b> Method description</a></li>
<li class="chapter" data-level="1.3" data-path="introduction.html"><a href="introduction.html#quick-start"><i class="fa fa-check"></i><b>1.3</b> Quick start</a></li>
<li class="chapter" data-level="1.4" data-path="introduction.html"><a href="introduction.html#where-to-get-help"><i class="fa fa-check"></i><b>1.4</b> Where to get help</a></li>
<li class="chapter" data-level="" data-path="introduction.html"><a href="introduction.html#session-information"><i class="fa fa-check"></i>Session information</a></li>
</ul></li>
<li class="chapter" data-level="2" data-path="using-the-built-in-references.html"><a href="using-the-built-in-references.html"><i class="fa fa-check"></i><b>2</b> Using the built-in references</a><ul>
<li class="chapter" data-level="2.1" data-path="using-the-built-in-references.html"><a href="using-the-built-in-references.html#overview"><i class="fa fa-check"></i><b>2.1</b> Overview</a></li>
<li class="chapter" data-level="2.2" data-path="using-the-built-in-references.html"><a href="using-the-built-in-references.html#annotation-with-default-marker-detection"><i class="fa fa-check"></i><b>2.2</b> Annotation with default marker detection</a></li>
<li class="chapter" data-level="2.3" data-path="using-the-built-in-references.html"><a href="using-the-built-in-references.html#interaction-with-quality-control"><i class="fa fa-check"></i><b>2.3</b> Interaction with quality control</a></li>
<li class="chapter" data-level="2.4" data-path="using-the-built-in-references.html"><a href="using-the-built-in-references.html#choices-of-assay-data"><i class="fa fa-check"></i><b>2.4</b> Choices of assay data</a></li>
<li class="chapter" data-level="" data-path="using-the-built-in-references.html"><a href="using-the-built-in-references.html#session-information-1"><i class="fa fa-check"></i>Session information</a></li>
</ul></li>
<li class="chapter" data-level="3" data-path="using-single-cell-references.html"><a href="using-single-cell-references.html"><i class="fa fa-check"></i><b>3</b> Using single-cell references</a><ul>
<li class="chapter" data-level="3.1" data-path="using-single-cell-references.html"><a href="using-single-cell-references.html#overview-1"><i class="fa fa-check"></i><b>3.1</b> Overview</a></li>
<li class="chapter" data-level="3.2" data-path="using-single-cell-references.html"><a href="using-single-cell-references.html#annotation-with-test-based-marker-detection"><i class="fa fa-check"></i><b>3.2</b> Annotation with test-based marker detection</a></li>
<li class="chapter" data-level="3.3" data-path="using-single-cell-references.html"><a href="using-single-cell-references.html#defining-custom-markers"><i class="fa fa-check"></i><b>3.3</b> Defining custom markers</a></li>
<li class="chapter" data-level="3.4" data-path="using-single-cell-references.html"><a href="using-single-cell-references.html#pseudo-bulk-aggregation"><i class="fa fa-check"></i><b>3.4</b> Pseudo-bulk aggregation</a></li>
<li class="chapter" data-level="" data-path="using-single-cell-references.html"><a href="using-single-cell-references.html#session-information-2"><i class="fa fa-check"></i>Session information</a></li>
</ul></li>
<li class="chapter" data-level="4" data-path="annotation-diagnostics.html"><a href="annotation-diagnostics.html"><i class="fa fa-check"></i><b>4</b> Annotation diagnostics</a><ul>
<li class="chapter" data-level="4.1" data-path="annotation-diagnostics.html"><a href="annotation-diagnostics.html#overview-2"><i class="fa fa-check"></i><b>4.1</b> Overview</a></li>
<li class="chapter" data-level="4.2" data-path="annotation-diagnostics.html"><a href="annotation-diagnostics.html#based-on-the-scores-within-cells"><i class="fa fa-check"></i><b>4.2</b> Based on the scores within cells</a></li>
<li class="chapter" data-level="4.3" data-path="annotation-diagnostics.html"><a href="annotation-diagnostics.html#based-on-the-deltas-across-cells"><i class="fa fa-check"></i><b>4.3</b> Based on the deltas across cells</a></li>
<li class="chapter" data-level="4.4" data-path="annotation-diagnostics.html"><a href="annotation-diagnostics.html#based-on-marker-gene-expression"><i class="fa fa-check"></i><b>4.4</b> Based on marker gene expression</a></li>
<li class="chapter" data-level="" data-path="annotation-diagnostics.html"><a href="annotation-diagnostics.html#session-information-3"><i class="fa fa-check"></i>Session information</a></li>
</ul></li>
<li class="part"><span><b>II Advanced usage</b></span></li>
<li class="chapter" data-level="5" data-path="using-multiple-references.html"><a href="using-multiple-references.html"><i class="fa fa-check"></i><b>5</b> Using multiple references</a><ul>
<li class="chapter" data-level="5.1" data-path="using-multiple-references.html"><a href="using-multiple-references.html#overview-3"><i class="fa fa-check"></i><b>5.1</b> Overview</a></li>
<li class="chapter" data-level="5.2" data-path="using-multiple-references.html"><a href="using-multiple-references.html#using-reference-specific-labels"><i class="fa fa-check"></i><b>5.2</b> Using reference-specific labels</a></li>
<li class="chapter" data-level="5.3" data-path="using-multiple-references.html"><a href="using-multiple-references.html#comparing-scores-across-references"><i class="fa fa-check"></i><b>5.3</b> Comparing scores across references</a><ul>
<li class="chapter" data-level="5.3.1" data-path="using-multiple-references.html"><a href="using-multiple-references.html#combining-inferences-from-individual-references"><i class="fa fa-check"></i><b>5.3.1</b> Combining inferences from individual references</a></li>
<li class="chapter" data-level="5.3.2" data-path="using-multiple-references.html"><a href="using-multiple-references.html#combined-diagnostics"><i class="fa fa-check"></i><b>5.3.2</b> Combined diagnostics</a></li>
</ul></li>
<li class="chapter" data-level="5.4" data-path="using-multiple-references.html"><a href="using-multiple-references.html#using-harmonized-labels"><i class="fa fa-check"></i><b>5.4</b> Using harmonized labels</a><ul>
<li class="chapter" data-level="5.4.1" data-path="using-multiple-references.html"><a href="using-multiple-references.html#sharing-information-during-marker-detection"><i class="fa fa-check"></i><b>5.4.1</b> Sharing information during marker detection</a></li>
<li class="chapter" data-level="5.4.2" data-path="using-multiple-references.html"><a href="using-multiple-references.html#manual-label-harmonization"><i class="fa fa-check"></i><b>5.4.2</b> Manual label harmonization</a></li>
</ul></li>
<li class="chapter" data-level="" data-path="using-multiple-references.html"><a href="using-multiple-references.html#session-info"><i class="fa fa-check"></i>Session info</a></li>
</ul></li>
<li class="chapter" data-level="6" data-path="exploiting-the-cell-ontology.html"><a href="exploiting-the-cell-ontology.html"><i class="fa fa-check"></i><b>6</b> Exploiting the cell ontology</a><ul>
<li class="chapter" data-level="6.1" data-path="exploiting-the-cell-ontology.html"><a href="exploiting-the-cell-ontology.html#motivation-1"><i class="fa fa-check"></i><b>6.1</b> Motivation</a></li>
<li class="chapter" data-level="6.2" data-path="exploiting-the-cell-ontology.html"><a href="exploiting-the-cell-ontology.html#basic-manipulation"><i class="fa fa-check"></i><b>6.2</b> Basic manipulation</a></li>
<li class="chapter" data-level="6.3" data-path="exploiting-the-cell-ontology.html"><a href="exploiting-the-cell-ontology.html#adjusting-resolution"><i class="fa fa-check"></i><b>6.3</b> Adjusting resolution</a></li>
<li class="chapter" data-level="" data-path="exploiting-the-cell-ontology.html"><a href="exploiting-the-cell-ontology.html#session-information-4"><i class="fa fa-check"></i>Session information</a></li>
</ul></li>
<li class="chapter" data-level="7" data-path="advanced-options.html"><a href="advanced-options.html"><i class="fa fa-check"></i><b>7</b> Advanced options</a><ul>
<li class="chapter" data-level="7.1" data-path="advanced-options.html"><a href="advanced-options.html#preconstructed-indices"><i class="fa fa-check"></i><b>7.1</b> Preconstructed indices</a></li>
<li class="chapter" data-level="7.2" data-path="advanced-options.html"><a href="advanced-options.html#parallelization"><i class="fa fa-check"></i><b>7.2</b> Parallelization</a></li>
<li class="chapter" data-level="7.3" data-path="advanced-options.html"><a href="advanced-options.html#approximate-algorithms"><i class="fa fa-check"></i><b>7.3</b> Approximate algorithms</a></li>
<li class="chapter" data-level="" data-path="advanced-options.html"><a href="advanced-options.html#session-information-5"><i class="fa fa-check"></i>Session information</a></li>
</ul></li>
<li class="part"><span><b>III Workflows</b></span></li>
<li class="chapter" data-level="8" data-path="cross-annotating-pancreas.html"><a href="cross-annotating-pancreas.html"><i class="fa fa-check"></i><b>8</b> Cross-annotating pancreas</a><ul>
<li class="chapter" data-level="8.1" data-path="cross-annotating-pancreas.html"><a href="cross-annotating-pancreas.html#loading-the-data"><i class="fa fa-check"></i><b>8.1</b> Loading the data</a></li>
<li class="chapter" data-level="8.2" data-path="cross-annotating-pancreas.html"><a href="cross-annotating-pancreas.html#applying-the-annotation"><i class="fa fa-check"></i><b>8.2</b> Applying the annotation</a></li>
<li class="chapter" data-level="8.3" data-path="cross-annotating-pancreas.html"><a href="cross-annotating-pancreas.html#diagnostics"><i class="fa fa-check"></i><b>8.3</b> Diagnostics</a></li>
<li class="chapter" data-level="" data-path="cross-annotating-pancreas.html"><a href="cross-annotating-pancreas.html#session-information-6"><i class="fa fa-check"></i>Session information</a></li>
</ul></li>
<li class="part"><span><b>IV Appendix</b></span></li>
<li class="chapter" data-level="9" data-path="contributors.html"><a href="contributors.html"><i class="fa fa-check"></i><b>9</b> Contributors</a><ul>
<li><a href="contributors.html#aaron-lun"><em>Aaron Lun</em></a></li>
</ul></li>
<li class="chapter" data-level="10" data-path="bibliography.html"><a href="bibliography.html"><i class="fa fa-check"></i><b>10</b> Bibliography</a></li>
<li class="divider"></li>
<li><a href="https://bioconductor.org" target="blank">Published by Bioconductor</a></li>

</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">Assigning cell types with SingleR</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="exploiting-the-cell-ontology" class="section level1">
<h1><span class="header-section-number">Chapter 6</span> Exploiting the cell ontology</h1>
<script>
document.addEventListener("click", function (event) {
    if (event.target.classList.contains("aaron-collapse")) {
        event.target.classList.toggle("active");
        var content = event.target.nextElementSibling;
        if (content.style.display === "block") {
            content.style.display = "none";
        } else {
            content.style.display = "block";
        }
    }
})
</script>
<style>
.aaron-collapse {
  background-color: #eee;
  color: #444;
  cursor: pointer;
  padding: 18px;
  width: 100%;
  border: none;
  text-align: left;
  outline: none;
  font-size: 15px;
}

.aaron-content {
  padding: 0 18px;
  display: none;
  overflow: hidden;
  background-color: #f1f1f1;
}
</style>
<div id="motivation-1" class="section level2">
<h2><span class="header-section-number">6.1</span> Motivation</h2>
<p>As previously discussed in Section <a href="using-multiple-references.html#using-harmonized-labels">5.4</a>, <em><a href="https://bioconductor.org/packages/3.12/SingleR">SingleR</a></em> maps the labels in its references to the <a href="https://www.ebi.ac.uk/ols/ontologies/cl">Cell Ontology</a>. The most obvious advantage of doing this is to provide a standardized vocabulary with which to describe cell types, thus facilitating integrated analyses with multiple references. However, another useful feature of the Cell Ontology is its hierarchical organization of terms, allowing us to adjust cell type annotations to the desired resolution. This represents a more dynamic alternative to the static <code>label.main</code> and <code>label.fine</code> options in each reference.</p>
</div>
<div id="basic-manipulation" class="section level2">
<h2><span class="header-section-number">6.2</span> Basic manipulation</h2>
<p>We use the <em><a href="https://bioconductor.org/packages/3.12/ontoProc">ontoProc</a></em> package to load in the Cell Ontology. This produces an <code>ontology_index</code> object (from the <em><a href="https://CRAN.R-project.org/package=ontologyIndex">ontologyIndex</a></em> package) that we can query for various pieces of information.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># </span><span class="al">TODO</span><span class="co">: wrap in utility function.</span>
<span class="kw">library</span>(ontoProc)
bfc &lt;-<span class="st"> </span>BiocFileCache<span class="op">::</span><span class="kw">BiocFileCache</span>(<span class="dt">ask=</span><span class="ot">FALSE</span>)
path &lt;-<span class="st"> </span>BiocFileCache<span class="op">::</span><span class="kw">bfcrpath</span>(bfc, <span class="st">&quot;http://purl.obolibrary.org/obo/cl.obo&quot;</span>)
cl &lt;-<span class="st"> </span><span class="kw">get_ontology</span>(path, <span class="dt">extract_tags=</span><span class="st">&quot;everything&quot;</span>)
cl</code></pre></div>
<pre><code>## Ontology with 2236 terms
## 
## format-version: 1.2
## data-version: cl/2020-05-21/cl-simple.owl
## ontology: cl/cl-simple
## 
## Properties:
##  id: character
##  name: character
##  parents: list
##  children: list
##  ancestors: list
##  obsolete: logical
##  alt_id: list
##  comment: character
##  consider: list
##  created_by: character
##  creation_date: character
##  data-version: list
##  def: character
##  develops_from: list
##  format-version: list
##  holds_over_chain: list
##  is_a: list
##  is_transitive: list
##  namespace: list
##  ontology: list
##  property_value: list
##  remark: list
##  replaced_by: list
##  subset: list
##  subsetdef: list
##  synonym: list
##  synonymtypedef: list
##  transitive_over: list
##  xref: list
## Roots:
##  CL:0000000 - cell
##  BFO:0000050 - NA
##  develops_from - develops_from</code></pre>
<p>The most immediate use of this object lies in mapping ontology terms to their plain-English descriptions. We can use this to translate annotations produced by <code>SingleR()</code> from the <code>label.ont</code> labels into a more interpretable form. We demonstrate this approach using <em><a href="https://bioconductor.org/packages/3.12/SingleR">SingleR</a></em>’s collection of mouse RNA-seq references <span class="citation">(Aran et al. <a href="#ref-aran2019reference">2019</a>)</span>.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">head</span>(cl<span class="op">$</span>name) <span class="co"># short name</span></code></pre></div>
<pre><code>##                         CL:0000000                         CL:0000001 
##                             &quot;cell&quot;            &quot;primary cultured cell&quot; 
##                         CL:0000002                         CL:0000003 
## &quot;obsolete immortal cell line cell&quot;                      &quot;native cell&quot; 
##                         CL:0000004                         CL:0000005 
##        &quot;obsolete cell by organism&quot;  &quot;fibroblast neural crest derived&quot;</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">head</span>(cl<span class="op">$</span>def) <span class="co"># longer definition</span></code></pre></div>
<pre><code>##                                                                                                                                                                                                                     CL:0000000 
##                                   &quot;\&quot;A material entity of anatomical origin (part of or deriving from an organism) that has as its parts a maximally connected cell compartment surrounded by a plasma membrane.\&quot; [CARO:mah]&quot; 
##                                                                                                                                                                                                                     CL:0000001 
##                                                                 &quot;\&quot;A cultured cell that is freshly isolated from a organismal source, or derives in culture from such a cell prior to the culture being passaged.\&quot; [ReO:mhb]&quot; 
##                                                                                                                                                                                                                     CL:0000002 
##             &quot;\&quot;OBSOLETE: A cell line cell that is expected to be capable of an unlimited number of divisions, and is thus able to support indefinite growth/propagation in vitro as part of a immortal cell line.\&quot; [ReO:mhb]&quot; 
##                                                                                                                                                                                                                     CL:0000003 
## &quot;\&quot;A cell that is found in a natural setting, which includes multicellular organism cells &#39;in vivo&#39; (i.e. part of an organism), and unicellular organisms &#39;in environment&#39; (i.e. part of a natural environment).\&quot; [CARO:mah]&quot; 
##                                                                                                                                                                                                                     CL:0000004 
##                                                                                                                            &quot;\&quot;OBSOLETE: A classification of cells by the organisms within which they are contained.\&quot; [FB:ma]&quot; 
##                                                                                                                                                                                                                     CL:0000005 
##                                                                                                                           &quot;\&quot;Any fibroblast that is deriived from the neural crest.\&quot; [https://orcid.org/0000-0001-5208-3432]&quot;</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(SingleR)
ref &lt;-<span class="st"> </span><span class="kw">MouseRNAseqData</span>(<span class="dt">cell.ont=</span><span class="st">&quot;nonna&quot;</span>)
translated &lt;-<span class="st"> </span>cl<span class="op">$</span>name[ref<span class="op">$</span>label.ont]
<span class="kw">head</span>(translated)</code></pre></div>
<pre><code>## &lt;NA&gt; &lt;NA&gt; &lt;NA&gt; &lt;NA&gt; &lt;NA&gt; &lt;NA&gt; 
##   NA   NA   NA   NA   NA   NA</code></pre>
<p>Another interesting application involves examining the relationship between different terms. The ontology itself is a directed acyclic graph, so we can can convert it into <code>graph</code> object for advanced queries using the <em><a href="https://CRAN.R-project.org/package=igraph">igraph</a></em> package. Each edge represents an “is a” relationship where each vertex represents a specialized case of the concept of the parent node.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># </span><span class="al">TODO</span><span class="co">: wrap in utility function.</span>
parents &lt;-<span class="st"> </span>cl<span class="op">$</span>parents
self &lt;-<span class="st"> </span><span class="kw">rep</span>(<span class="kw">names</span>(parents), <span class="kw">lengths</span>(parents))

<span class="kw">library</span>(igraph)
g &lt;-<span class="st"> </span><span class="kw">make_graph</span>(<span class="kw">rbind</span>(<span class="kw">unlist</span>(parents), self))
g</code></pre></div>
<pre><code>## IGRAPH 045f37f DN-- 2234 3141 -- 
## + attr: name (v/c)
## + edges from 045f37f (vertex names):
##  [1] CL:0000010-&gt;CL:0000001 CL:0000000-&gt;CL:0000003 CL:0000057-&gt;CL:0000005
##  [4] CL:0000101-&gt;CL:0000006 CL:0000197-&gt;CL:0000006 CL:0002321-&gt;CL:0000007
##  [7] CL:0000333-&gt;CL:0000008 CL:0000578-&gt;CL:0000010 CL:0000333-&gt;CL:0000011
## [10] CL:0000034-&gt;CL:0000014 CL:0000039-&gt;CL:0000014 CL:0000586-&gt;CL:0000015
## [13] CL:0000014-&gt;CL:0000016 CL:0000015-&gt;CL:0000016 CL:0000015-&gt;CL:0000017
## [16] CL:0000015-&gt;CL:0000018 CL:0000413-&gt;CL:0000018 CL:0000408-&gt;CL:0000019
## [19] CL:0000015-&gt;CL:0000020 CL:0000586-&gt;CL:0000021 CL:0000014-&gt;CL:0000022
## [22] CL:0000021-&gt;CL:0000022 CL:0000021-&gt;CL:0000023 CL:0000021-&gt;CL:0000024
## + ... omitted several edges</code></pre>
<p>One query involves identifying all descendents of a particular term of interest. This can be useful when searching for a cell type in the presence of variable annotation resolution; for example, a search for “epithelial cell” can be configured to pick up all child terms such as “endothelial cell” and “ependymal cell”.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">term &lt;-<span class="st"> &quot;CL:0000624&quot;</span>
cl<span class="op">$</span>name[term]</code></pre></div>
<pre><code>##                        CL:0000624 
## &quot;CD4-positive, alpha-beta T cell&quot;</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">all.kids &lt;-<span class="st"> </span><span class="kw">names</span>(<span class="kw">subcomponent</span>(g, term))
<span class="kw">head</span>(cl<span class="op">$</span>name[all.kids])</code></pre></div>
<pre><code>##                                                       CL:0000624 
##                                &quot;CD4-positive, alpha-beta T cell&quot; 
##                                                       CL:0000492 
##                                     &quot;CD4-positive helper T cell&quot; 
##                                                       CL:0001051 
## &quot;CD4-positive, CXCR3-negative, CCR6-negative, alpha-beta T cell&quot; 
##                                                       CL:0000791 
##                                       &quot;mature alpha-beta T cell&quot; 
##                                                       CL:0000792 
##      &quot;CD4-positive, CD25-positive, alpha-beta regulatory T cell&quot; 
##                                                       CL:0000793 
##                &quot;CD4-positive, alpha-beta intraepithelial T cell&quot;</code></pre>
<p>Alternatively, we might be interested in the last common ancestor (LCA) for a set of terms. This is the furthest term - or, in some cases, multiple terms - from the root of the ontology that is also an ancestor of all of the terms of interest. We will use this LCA concept in the next section to adjust resolution across multiple references.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">terms &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="st">&quot;CL:0000624&quot;</span>, <span class="st">&quot;CL:0000785&quot;</span>, <span class="st">&quot;CL:0000623&quot;</span>)
cl<span class="op">$</span>name[terms]</code></pre></div>
<pre><code>##                        CL:0000624                        CL:0000785 
## &quot;CD4-positive, alpha-beta T cell&quot;                   &quot;mature B cell&quot; 
##                        CL:0000623 
##             &quot;natural killer cell&quot;</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># </span><span class="al">TODO</span><span class="co">: god, put this in a function somewhere.</span>
all.ancestors &lt;-<span class="st"> </span><span class="kw">lapply</span>(terms, subcomponent, <span class="dt">graph=</span>g, <span class="dt">mode=</span><span class="st">&quot;in&quot;</span>)
all.ancestors &lt;-<span class="st"> </span><span class="kw">lapply</span>(all.ancestors, names)
common.ancestors &lt;-<span class="st"> </span><span class="kw">Reduce</span>(intersect, all.ancestors)

ancestors.of.ancestors &lt;-<span class="st"> </span><span class="kw">lapply</span>(common.ancestors, subcomponent, <span class="dt">graph=</span>g, <span class="dt">mode=</span><span class="st">&quot;in&quot;</span>)
ancestors.of.ancestors &lt;-<span class="st"> </span><span class="kw">lapply</span>(ancestors.of.ancestors, names)
ancestors.of.ancestors &lt;-<span class="st"> </span><span class="kw">mapply</span>(setdiff, ancestors.of.ancestors, common.ancestors) 

latest.common.ancestors &lt;-<span class="st"> </span><span class="kw">setdiff</span>(common.ancestors, <span class="kw">unlist</span>(ancestors.of.ancestors))
cl<span class="op">$</span>name[latest.common.ancestors]</code></pre></div>
<pre><code>##   CL:0000542 
## &quot;lymphocyte&quot;</code></pre>
</div>
<div id="adjusting-resolution" class="section level2">
<h2><span class="header-section-number">6.3</span> Adjusting resolution</h2>
<p>We can use the ontology graph to adjust the resolution of the reference labels by rolling up overly-specific terms to their LCA. The <code>findCommonAncestors()</code> utility takes a set of terms and returns a list of potential LCAs for various subsets of those terms. Users can inspect this list to identify LCAs at the desired resolution and then map their descendent terms to those LCAs.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">findCommonAncestors &lt;-<span class="st"> </span><span class="cf">function</span>(..., g, <span class="dt">remove.self=</span><span class="ot">TRUE</span>, <span class="dt">names=</span><span class="ot">NULL</span>) {
    terms &lt;-<span class="st"> </span><span class="kw">list</span>(...)
    <span class="cf">if</span> (<span class="kw">is.null</span>(<span class="kw">names</span>(terms))) {
        <span class="kw">names</span>(terms) &lt;-<span class="st"> </span><span class="kw">sprintf</span>(<span class="st">&quot;set%i&quot;</span>, <span class="kw">seq_along</span>(terms))
    }

    all.terms &lt;-<span class="st"> </span><span class="kw">unique</span>(<span class="kw">unlist</span>(terms))
    all.ancestors &lt;-<span class="st"> </span><span class="kw">lapply</span>(all.terms, subcomponent, <span class="dt">graph=</span>g, <span class="dt">mode=</span><span class="st">&quot;in&quot;</span>)
    all.ancestors &lt;-<span class="st"> </span><span class="kw">lapply</span>(all.ancestors, names)
    by.ancestor &lt;-<span class="st"> </span><span class="kw">split</span>(
        <span class="kw">rep</span>(all.terms, <span class="kw">lengths</span>(all.ancestors)),
        <span class="kw">unlist</span>(all.ancestors)
    )

    <span class="co"># Removing ancestor nodes with the same count as its children.</span>
    available &lt;-<span class="st"> </span><span class="kw">names</span>(by.ancestor)
    <span class="cf">for</span> (i <span class="cf">in</span> available) {
        <span class="cf">if</span> (<span class="op">!</span>i <span class="op">%in%</span><span class="st"> </span><span class="kw">names</span>(by.ancestor)) {
            <span class="cf">next</span>
        }

        counts &lt;-<span class="st"> </span><span class="kw">lengths</span>(by.ancestor)
        cur.ancestors &lt;-<span class="st"> </span><span class="kw">subcomponent</span>(g, i, <span class="dt">mode=</span><span class="st">&quot;in&quot;</span>)
        cur.ancestors &lt;-<span class="st"> </span><span class="kw">setdiff</span>(<span class="kw">names</span>(cur.ancestors), i)
        drop &lt;-<span class="st"> </span>cur.ancestors[counts[i]<span class="op">==</span>counts[cur.ancestors]]
        by.ancestor &lt;-<span class="st"> </span>by.ancestor[<span class="op">!</span><span class="kw">names</span>(by.ancestor) <span class="op">%in%</span><span class="st"> </span>drop]
    }

    <span class="cf">if</span> (remove.self) {
        by.ancestor &lt;-<span class="st"> </span>by.ancestor[<span class="kw">lengths</span>(by.ancestor) <span class="op">&gt;</span><span class="st"> </span>1L]
    }
    by.ancestor &lt;-<span class="st"> </span>by.ancestor[<span class="kw">order</span>(<span class="kw">lengths</span>(by.ancestor))] <span class="co"># most specific terms first.</span>

    <span class="co"># Decorating the output.</span>
    <span class="cf">for</span> (i <span class="cf">in</span> <span class="kw">names</span>(by.ancestor)) {
        current &lt;-<span class="st"> </span>by.ancestor[[i]]
        df &lt;-<span class="st"> </span><span class="kw">DataFrame</span>(<span class="dt">row.names=</span>current)

        curout &lt;-<span class="st"> </span><span class="kw">list</span>()
        <span class="cf">if</span> (<span class="op">!</span><span class="kw">is.null</span>(names)) {
            curout<span class="op">$</span>name &lt;-<span class="st"> </span><span class="kw">unname</span>(names[i])
            df<span class="op">$</span>name &lt;-<span class="st"> </span>names[current]
        }

        presence &lt;-<span class="st"> </span><span class="kw">list</span>()
        <span class="cf">for</span> (b <span class="cf">in</span> <span class="kw">names</span>(terms)) {
            presence[[b]] &lt;-<span class="st"> </span>current <span class="op">%in%</span><span class="st"> </span>terms[[b]]
        }
        df &lt;-<span class="st"> </span><span class="kw">cbind</span>(df, <span class="kw">do.call</span>(DataFrame, presence))

        curout<span class="op">$</span>descendents &lt;-<span class="st"> </span>df
        by.ancestor[[i]] &lt;-<span class="st"> </span>curout
    }

    by.ancestor
}

lca &lt;-<span class="st"> </span><span class="kw">findCommonAncestors</span>(ref<span class="op">$</span>label.ont, <span class="dt">g=</span>g, <span class="dt">names=</span>cl<span class="op">$</span>name)
<span class="kw">head</span>(lca)</code></pre></div>
<pre><code>## $`CL:0000081`
## $`CL:0000081`$name
## [1] &quot;blood cell&quot;
## 
## $`CL:0000081`$descendents
## DataFrame with 2 rows and 2 columns
##                   name      set1
##            &lt;character&gt; &lt;logical&gt;
## CL:0000232 erythrocyte      TRUE
## CL:0000094 granulocyte      TRUE
## 
## 
## $`CL:0000126`
## $`CL:0000126`$name
## [1] &quot;macroglial cell&quot;
## 
## $`CL:0000126`$descendents
## DataFrame with 2 rows and 2 columns
##                       name      set1
##                &lt;character&gt; &lt;logical&gt;
## CL:0000127       astrocyte      TRUE
## CL:0000128 oligodendrocyte      TRUE
## 
## 
## $`CL:0000393`
## $`CL:0000393`$name
## [1] &quot;electrically responsive cell&quot;
## 
## $`CL:0000393`$descendents
## DataFrame with 2 rows and 2 columns
##                           name      set1
##                    &lt;character&gt; &lt;logical&gt;
## CL:0000540              neuron      TRUE
## CL:0000746 cardiac muscle cell      TRUE
## 
## 
## $`CL:0002320`
## $`CL:0002320`$name
## [1] &quot;connective tissue cell&quot;
## 
## $`CL:0002320`$descendents
## DataFrame with 2 rows and 2 columns
##                   name      set1
##            &lt;character&gt; &lt;logical&gt;
## CL:0000136    fat cell      TRUE
## CL:0000057  fibroblast      TRUE
## 
## 
## $`CL:0011115`
## $`CL:0011115`$name
## [1] &quot;precursor cell&quot;
## 
## $`CL:0011115`$descendents
## DataFrame with 2 rows and 2 columns
##                          name      set1
##                   &lt;character&gt; &lt;logical&gt;
## CL:0000047 neuronal stem cell      TRUE
## CL:0000576           monocyte      TRUE
## 
## 
## $`CL:0000066`
## $`CL:0000066`$name
## [1] &quot;epithelial cell&quot;
## 
## $`CL:0000066`$descendents
## DataFrame with 3 rows and 2 columns
##                        name      set1
##                 &lt;character&gt; &lt;logical&gt;
## CL:0000115 endothelial cell      TRUE
## CL:0000182       hepatocyte      TRUE
## CL:0000065   ependymal cell      TRUE</code></pre>
<p>We can also use this function to synchronize multiple sets of terms to the same resolution. Here, we consider the ImmGen dataset <span class="citation">(Heng et al. <a href="#ref-ImmGenRef">2008</a>)</span>, which provides highly resolved annotation of immune cell types. The <code>findCommonAncestors()</code> function specifies the origins of the descendents for each LCA, allowing us to focus on LCAs that have representatives in both sets of terms.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">ref2 &lt;-<span class="st"> </span><span class="kw">ImmGenData</span>(<span class="dt">cell.ont=</span><span class="st">&quot;nonna&quot;</span>)
lca2 &lt;-<span class="st"> </span><span class="kw">findCommonAncestors</span>(<span class="dt">MouseRNA=</span>ref<span class="op">$</span>label.ont,
    <span class="dt">ImmGen=</span>ref2<span class="op">$</span>label.ont, <span class="dt">g=</span>g, <span class="dt">names=</span>cl<span class="op">$</span>name)
<span class="kw">head</span>(lca2)</code></pre></div>
<pre><code>## $`CL:0000126`
## $`CL:0000126`$name
## [1] &quot;macroglial cell&quot;
## 
## $`CL:0000126`$descendents
## DataFrame with 2 rows and 3 columns
##                       name  MouseRNA    ImmGen
##                &lt;character&gt; &lt;logical&gt; &lt;logical&gt;
## CL:0000127       astrocyte      TRUE     FALSE
## CL:0000128 oligodendrocyte      TRUE     FALSE
## 
## 
## $`CL:0000393`
## $`CL:0000393`$name
## [1] &quot;electrically responsive cell&quot;
## 
## $`CL:0000393`$descendents
## DataFrame with 2 rows and 3 columns
##                           name  MouseRNA    ImmGen
##                    &lt;character&gt; &lt;logical&gt; &lt;logical&gt;
## CL:0000540              neuron      TRUE     FALSE
## CL:0000746 cardiac muscle cell      TRUE     FALSE
## 
## 
## $`CL:0000623`
## $`CL:0000623`$name
## [1] &quot;natural killer cell&quot;
## 
## $`CL:0000623`$descendents
## DataFrame with 2 rows and 3 columns
##                                          name  MouseRNA    ImmGen
##                                   &lt;character&gt; &lt;logical&gt; &lt;logical&gt;
## CL:0000623                natural killer cell      TRUE      TRUE
## CL:0002438 NK1.1-positive natural killer cell     FALSE      TRUE
## 
## 
## $`CL:0000813`
## $`CL:0000813`$name
## [1] &quot;memory T cell&quot;
## 
## $`CL:0000813`$descendents
## DataFrame with 2 rows and 3 columns
##                                              name  MouseRNA    ImmGen
##                                       &lt;character&gt; &lt;logical&gt; &lt;logical&gt;
## CL:0000897 CD4-positive, alpha-beta memory T cell     FALSE      TRUE
## CL:0000909 CD8-positive, alpha-beta memory T cell     FALSE      TRUE
## 
## 
## $`CL:0000815`
## $`CL:0000815`$name
## [1] &quot;regulatory T cell&quot;
## 
## $`CL:0000815`$descendents
## DataFrame with 2 rows and 3 columns
##                                                                 name  MouseRNA
##                                                          &lt;character&gt; &lt;logical&gt;
## CL:0000792 CD4-positive, CD25-positive, alpha-beta regulatory T cell     FALSE
## CL:0000815                                         regulatory T cell     FALSE
##               ImmGen
##            &lt;logical&gt;
## CL:0000792      TRUE
## CL:0000815      TRUE
## 
## 
## $`CL:0000819`
## $`CL:0000819`$name
## [1] &quot;B-1 B cell&quot;
## 
## $`CL:0000819`$descendents
## DataFrame with 2 rows and 3 columns
##                   name  MouseRNA    ImmGen
##            &lt;character&gt; &lt;logical&gt; &lt;logical&gt;
## CL:0000820 B-1a B cell     FALSE      TRUE
## CL:0000821 B-1b B cell     FALSE      TRUE</code></pre>
<p>For example, we might notice that the mouse RNA-seq reference only has a single “T cell” term. To synchronize resolution across references, we would need to roll up all of the ImmGen’s finely resolved subsets into that LCA as shown below. Of course, this results in some loss of precision and information; whether this is an acceptable price for simpler interpretation is a decision that is left to the user.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">children &lt;-<span class="st"> </span>lca2<span class="op">$</span><span class="st">`</span><span class="dt">CL:0000084</span><span class="st">`</span><span class="op">$</span>descendents
children</code></pre></div>
<pre><code>## DataFrame with 35 rows and 3 columns
##                                                                 name  MouseRNA
##                                                          &lt;character&gt; &lt;logical&gt;
## CL:0000084                                                    T cell      TRUE
## CL:0002427                         resting double-positive thymocyte     FALSE
## CL:0000809                     double-positive, alpha-beta thymocyte     FALSE
## CL:0002429                   CD69-positive double-positive thymocyte     FALSE
## CL:0000624                           CD4-positive, alpha-beta T cell     FALSE
## ...                                                              ...       ...
## CL:0002415 immature Vgamma1.1-positive, Vdelta6.3-positive thymocyte     FALSE
## CL:0002411          Vgamma1.1-positive, Vdelta6.3-negative thymocyte     FALSE
## CL:0002416   mature Vgamma1.1-positive, Vdelta6.3-positive thymocyte     FALSE
## CL:0002407                         mature Vgamma2-positive thymocyte     FALSE
## CL:0000815                                         regulatory T cell     FALSE
##               ImmGen
##            &lt;logical&gt;
## CL:0000084      TRUE
## CL:0002427      TRUE
## CL:0000809      TRUE
## CL:0002429      TRUE
## CL:0000624      TRUE
## ...              ...
## CL:0002415      TRUE
## CL:0002411      TRUE
## CL:0002416      TRUE
## CL:0002407      TRUE
## CL:0000815      TRUE</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># Synchronization:</span>
synced.mm &lt;-<span class="st"> </span>ref<span class="op">$</span>label.ont
synced.mm[synced.mm <span class="op">%in%</span><span class="st"> </span><span class="kw">rownames</span>(children)] &lt;-<span class="st"> &quot;CL:0000084&quot;</span>
synced.ig &lt;-<span class="st"> </span>ref2<span class="op">$</span>label.ont
synced.ig[synced.ig <span class="op">%in%</span><span class="st"> </span><span class="kw">rownames</span>(children)] &lt;-<span class="st"> &quot;CL:0000084&quot;</span></code></pre></div>
</div>
<div id="session-information-4" class="section level2 unnumbered">
<h2>Session information</h2>
<button class="aaron-collapse">
View session info
</button>
<div class="aaron-content">
<pre><code>R version 4.0.0 Patched (2020-05-01 r78341)
Platform: x86_64-pc-linux-gnu (64-bit)
Running under: Ubuntu 18.04.4 LTS

Matrix products: default
BLAS:   /home/luna/Software/R/R-4-0-branch-dev/lib/libRblas.so
LAPACK: /home/luna/Software/R/R-4-0-branch-dev/lib/libRlapack.so

locale:
 [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              
 [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    
 [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   
 [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 
 [9] LC_ADDRESS=C               LC_TELEPHONE=C            
[11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       

attached base packages:
[1] stats4    parallel  stats     graphics  grDevices utils     datasets 
[8] methods   base     

other attached packages:
 [1] igraph_1.2.5                SingleR_1.3.4              
 [3] SummarizedExperiment_1.19.4 DelayedArray_0.15.1        
 [5] matrixStats_0.56.0          Biobase_2.49.0             
 [7] GenomicRanges_1.41.1        GenomeInfoDb_1.25.0        
 [9] IRanges_2.23.5              S4Vectors_0.27.6           
[11] BiocGenerics_0.35.2         ontoProc_1.9.2             
[13] ontologyIndex_2.5           BiocStyle_2.17.0           
[15] rebook_0.99.0              

loaded via a namespace (and not attached):
 [1] httr_1.4.1                    BiocSingular_1.5.0           
 [3] AnnotationHub_2.21.0          bit64_0.9-7                  
 [5] DelayedMatrixStats_1.11.0     ontologyPlot_1.4             
 [7] paintmap_1.0                  shiny_1.4.0.2                
 [9] assertthat_0.2.1              interactiveDisplayBase_1.27.2
[11] BiocFileCache_1.13.0          BiocManager_1.30.10          
[13] blob_1.2.1                    GenomeInfoDbData_1.2.3       
[15] yaml_2.2.1                    BiocVersion_3.12.0           
[17] pillar_1.4.4                  RSQLite_2.2.0                
[19] lattice_0.20-41               glue_1.4.1                   
[21] digest_0.6.25                 promises_1.1.0               
[23] XVector_0.29.1                htmltools_0.4.0              
[25] httpuv_1.5.2                  Matrix_1.2-18                
[27] XML_3.99-0.3                  pkgconfig_2.0.3              
[29] bookdown_0.19                 zlibbioc_1.35.0              
[31] purrr_0.3.4                   xtable_1.8-4                 
[33] processx_3.4.2                later_1.0.0                  
[35] BiocParallel_1.23.0           tibble_3.0.1                 
[37] ellipsis_0.3.1                magrittr_1.5                 
[39] crayon_1.3.4                  CodeDepends_0.6.5            
[41] mime_0.9                      memoise_1.1.0                
[43] evaluate_0.14                 ps_1.3.3                     
[45] graph_1.67.0                  tools_4.0.0                  
[47] lifecycle_0.2.0               stringr_1.4.0                
[49] irlba_2.3.3                   AnnotationDbi_1.51.0         
[51] callr_3.4.3                   compiler_4.0.0               
[53] rsvd_1.0.3                    rlang_0.4.6                  
[55] grid_4.0.0                    RCurl_1.98-1.2               
[57] BiocNeighbors_1.7.0           rappdirs_0.3.1               
[59] bitops_1.0-6                  rmarkdown_2.1                
[61] ExperimentHub_1.15.0          codetools_0.2-16             
[63] curl_4.3                      DBI_1.1.0                    
[65] R6_2.4.1                      knitr_1.28                   
[67] dplyr_0.8.5                   fastmap_1.0.1                
[69] bit_1.1-15.2                  Rgraphviz_2.33.0             
[71] stringi_1.4.6                 Rcpp_1.0.4.6                 
[73] vctrs_0.3.0                   dbplyr_1.4.3                 
[75] tidyselect_1.1.0              xfun_0.14                    </code></pre>
</div>

</div>
</div>
<h3> Bibliography</h3>
<div id="refs" class="references">
<div id="ref-aran2019reference">
<p>Aran, D., A. P. Looney, L. Liu, E. Wu, V. Fong, A. Hsu, S. Chak, et al. 2019. “Reference-based analysis of lung single-cell sequencing reveals a transitional profibrotic macrophage.” <em>Nat. Immunol.</em> 20 (2): 163–72.</p>
</div>
<div id="ref-ImmGenRef">
<p>Heng, Tracy S.P., Michio W. Painter, Kutlu Elpek, Veronika Lukacs-Kornek, Nora Mauermann, Shannon J. Turley, Daphne Koller, et al. 2008. “The immunological genome project: Networks of gene expression in immune cells.” <em>Nature Immunology</em> 9 (10): 1091–4. doi:<a href="https://doi.org/10.1038/ni1008-1091">10.1038/ni1008-1091</a>.</p>
</div>
</div>
            </section>

          </div>
        </div>
      </div>
<a href="using-multiple-references.html" class="navigation navigation-prev " aria-label="Previous page"><i class="fa fa-angle-left"></i></a>
<a href="advanced-options.html" class="navigation navigation-next " aria-label="Next page"><i class="fa fa-angle-right"></i></a>
    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/lunr.js"></script>
<script src="libs/gitbook-2.6.7/js/clipboard.min.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-clipboard.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": {
"github": false,
"facebook": false,
"twitter": false,
"linkedin": false,
"weibo": false,
"instapaper": false,
"vk": false,
"all": ["facebook", "twitter", "linkedin", "weibo", "instapaper"]
},
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": {
"link": null,
"text": null
},
"history": {
"link": null,
"text": null
},
"view": {
"link": null,
"text": null
},
"download": null,
"toc": {
"collapse": "subsection"
},
"bookdown::epub_book": "default",
"cover": "images/cover.png"
});
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    var src = "";
    if (src === "" || src === "true") src = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-MML-AM_CHTML";
    if (location.protocol !== "file:")
      if (/^https?:/.test(src))
        src = src.replace(/^https?:/, '');
    script.src = src;
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>
</body>

</html>
