<!DOCTYPE html>
<html lang="" xml:lang="">
<head>

  <meta charset="utf-8" />
  <meta http-equiv="X-UA-Compatible" content="IE=edge" />
  <title>Chapter 6 Harmonizing labels | Assigning cell types with SingleR</title>
  <meta name="description" content="The SingleR book. Because sometimes, a vignette just isn’t enough." />
  <meta name="generator" content="bookdown 0.18 and GitBook 2.6.7" />

  <meta property="og:title" content="Chapter 6 Harmonizing labels | Assigning cell types with SingleR" />
  <meta property="og:type" content="book" />
  
  
  <meta property="og:description" content="The SingleR book. Because sometimes, a vignette just isn’t enough." />
  

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="Chapter 6 Harmonizing labels | Assigning cell types with SingleR" />
  
  <meta name="twitter:description" content="The SingleR book. Because sometimes, a vignette just isn’t enough." />
  



<meta name="date" content="2020-05-13" />

  <meta name="viewport" content="width=device-width, initial-scale=1" />
  <meta name="apple-mobile-web-app-capable" content="yes" />
  <meta name="apple-mobile-web-app-status-bar-style" content="black" />
  
  
<link rel="prev" href="using-multiple-references.html"/>
<link rel="next" href="advanced-options.html"/>
<script src="libs/jquery-2.2.3/jquery.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-table.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-clipboard.css" rel="stylesheet" />











<style type="text/css">
div.sourceCode { overflow-x: auto; }
table.sourceCode, tr.sourceCode, td.lineNumbers, td.sourceCode {
  margin: 0; padding: 0; vertical-align: baseline; border: none; }
table.sourceCode { width: 100%; line-height: 100%; background-color: #f8f8f8; }
td.lineNumbers { text-align: right; padding-right: 4px; padding-left: 4px; color: #aaaaaa; border-right: 1px solid #aaaaaa; }
td.sourceCode { padding-left: 5px; }
pre, code { background-color: #f8f8f8; }
code > span.kw { color: #204a87; font-weight: bold; } /* Keyword */
code > span.dt { color: #204a87; } /* DataType */
code > span.dv { color: #0000cf; } /* DecVal */
code > span.bn { color: #0000cf; } /* BaseN */
code > span.fl { color: #0000cf; } /* Float */
code > span.ch { color: #4e9a06; } /* Char */
code > span.st { color: #4e9a06; } /* String */
code > span.co { color: #8f5902; font-style: italic; } /* Comment */
code > span.ot { color: #8f5902; } /* Other */
code > span.al { color: #ef2929; } /* Alert */
code > span.fu { color: #000000; } /* Function */
code > span.er { color: #a40000; font-weight: bold; } /* Error */
code > span.wa { color: #8f5902; font-weight: bold; font-style: italic; } /* Warning */
code > span.cn { color: #000000; } /* Constant */
code > span.sc { color: #000000; } /* SpecialChar */
code > span.vs { color: #4e9a06; } /* VerbatimString */
code > span.ss { color: #4e9a06; } /* SpecialString */
code > span.im { } /* Import */
code > span.va { color: #000000; } /* Variable */
code > span.cf { color: #204a87; font-weight: bold; } /* ControlFlow */
code > span.op { color: #ce5c00; font-weight: bold; } /* Operator */
code > span.pp { color: #8f5902; font-style: italic; } /* Preprocessor */
code > span.ex { } /* Extension */
code > span.at { color: #c4a000; } /* Attribute */
code > span.do { color: #8f5902; font-weight: bold; font-style: italic; } /* Documentation */
code > span.an { color: #8f5902; font-weight: bold; font-style: italic; } /* Annotation */
code > span.cv { color: #8f5902; font-weight: bold; font-style: italic; } /* CommentVar */
code > span.in { color: #8f5902; font-weight: bold; font-style: italic; } /* Information */
</style>

<link rel="stylesheet" href="style.css" type="text/css" />
</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li><a href="./">The SingleR book</a></li>

<li class="divider"></li>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html"><i class="fa fa-check"></i>Preface</a></li>
<li class="chapter" data-level="1" data-path="introduction.html"><a href="introduction.html"><i class="fa fa-check"></i><b>1</b> Introduction</a><ul>
<li class="chapter" data-level="1.1" data-path="introduction.html"><a href="introduction.html#motivation"><i class="fa fa-check"></i><b>1.1</b> Motivation</a></li>
<li class="chapter" data-level="1.2" data-path="introduction.html"><a href="introduction.html#method-description"><i class="fa fa-check"></i><b>1.2</b> Method description</a></li>
<li class="chapter" data-level="1.3" data-path="introduction.html"><a href="introduction.html#quick-start"><i class="fa fa-check"></i><b>1.3</b> Quick start</a></li>
<li class="chapter" data-level="1.4" data-path="introduction.html"><a href="introduction.html#where-to-get-help"><i class="fa fa-check"></i><b>1.4</b> Where to get help</a></li>
<li class="chapter" data-level="" data-path="introduction.html"><a href="introduction.html#session-information"><i class="fa fa-check"></i>Session information</a></li>
<li class="chapter" data-level="" data-path="introduction.html"><a href="introduction.html#references"><i class="fa fa-check"></i>References</a></li>
</ul></li>
<li class="chapter" data-level="2" data-path="using-the-built-in-references.html"><a href="using-the-built-in-references.html"><i class="fa fa-check"></i><b>2</b> Using the built-in references</a><ul>
<li class="chapter" data-level="2.1" data-path="using-the-built-in-references.html"><a href="using-the-built-in-references.html#overview"><i class="fa fa-check"></i><b>2.1</b> Overview</a></li>
<li class="chapter" data-level="2.2" data-path="using-the-built-in-references.html"><a href="using-the-built-in-references.html#annotation-with-default-marker-detection"><i class="fa fa-check"></i><b>2.2</b> Annotation with default marker detection</a></li>
<li class="chapter" data-level="2.3" data-path="using-the-built-in-references.html"><a href="using-the-built-in-references.html#choices-of-assay-data"><i class="fa fa-check"></i><b>2.3</b> Choices of assay data</a></li>
<li class="chapter" data-level="" data-path="using-the-built-in-references.html"><a href="using-the-built-in-references.html#session-information-1"><i class="fa fa-check"></i>Session information</a></li>
<li class="chapter" data-level="" data-path="using-the-built-in-references.html"><a href="using-the-built-in-references.html#references-1"><i class="fa fa-check"></i>References</a></li>
</ul></li>
<li class="chapter" data-level="3" data-path="using-single-cell-references.html"><a href="using-single-cell-references.html"><i class="fa fa-check"></i><b>3</b> Using single-cell references</a><ul>
<li class="chapter" data-level="3.1" data-path="using-single-cell-references.html"><a href="using-single-cell-references.html#overview-1"><i class="fa fa-check"></i><b>3.1</b> Overview</a></li>
<li class="chapter" data-level="3.2" data-path="using-single-cell-references.html"><a href="using-single-cell-references.html#annotation-with-test-based-marker-detection"><i class="fa fa-check"></i><b>3.2</b> Annotation with test-based marker detection</a></li>
<li class="chapter" data-level="3.3" data-path="using-single-cell-references.html"><a href="using-single-cell-references.html#defining-custom-markers"><i class="fa fa-check"></i><b>3.3</b> Defining custom markers</a></li>
<li class="chapter" data-level="3.4" data-path="using-single-cell-references.html"><a href="using-single-cell-references.html#pseudo-bulk-aggregation"><i class="fa fa-check"></i><b>3.4</b> Pseudo-bulk aggregation</a></li>
<li class="chapter" data-level="" data-path="using-single-cell-references.html"><a href="using-single-cell-references.html#session-information-2"><i class="fa fa-check"></i>Session information</a></li>
<li class="chapter" data-level="3.5" data-path="using-single-cell-references.html"><a href="using-single-cell-references.html#references-2"><i class="fa fa-check"></i><b>3.5</b> References</a></li>
</ul></li>
<li class="chapter" data-level="4" data-path="annotation-diagnostics.html"><a href="annotation-diagnostics.html"><i class="fa fa-check"></i><b>4</b> Annotation diagnostics</a><ul>
<li class="chapter" data-level="4.1" data-path="annotation-diagnostics.html"><a href="annotation-diagnostics.html#based-on-the-scores-within-cells"><i class="fa fa-check"></i><b>4.1</b> Based on the scores within cells</a></li>
<li class="chapter" data-level="4.2" data-path="annotation-diagnostics.html"><a href="annotation-diagnostics.html#based-on-the-deltas-across-cells"><i class="fa fa-check"></i><b>4.2</b> Based on the deltas across cells</a></li>
<li class="chapter" data-level="4.3" data-path="annotation-diagnostics.html"><a href="annotation-diagnostics.html#based-on-marker-gene-expression"><i class="fa fa-check"></i><b>4.3</b> Based on marker gene expression</a></li>
</ul></li>
<li class="chapter" data-level="5" data-path="using-multiple-references.html"><a href="using-multiple-references.html"><i class="fa fa-check"></i><b>5</b> Using multiple references</a></li>
<li class="chapter" data-level="6" data-path="harmonizing-labels.html"><a href="harmonizing-labels.html"><i class="fa fa-check"></i><b>6</b> Harmonizing labels</a></li>
<li class="chapter" data-level="7" data-path="advanced-options.html"><a href="advanced-options.html"><i class="fa fa-check"></i><b>7</b> Advanced options</a><ul>
<li class="chapter" data-level="7.1" data-path="advanced-options.html"><a href="advanced-options.html#preconstructed-indices"><i class="fa fa-check"></i><b>7.1</b> Preconstructed indices</a></li>
<li class="chapter" data-level="7.2" data-path="advanced-options.html"><a href="advanced-options.html#parallelization"><i class="fa fa-check"></i><b>7.2</b> Parallelization</a></li>
<li class="chapter" data-level="7.3" data-path="advanced-options.html"><a href="advanced-options.html#approximate-algorithms"><i class="fa fa-check"></i><b>7.3</b> Approximate algorithms</a></li>
<li class="chapter" data-level="" data-path="advanced-options.html"><a href="advanced-options.html#session-information-3"><i class="fa fa-check"></i>Session information</a></li>
<li class="chapter" data-level="" data-path="advanced-options.html"><a href="advanced-options.html#references-3"><i class="fa fa-check"></i>References</a></li>
</ul></li>
<li class="divider"></li>
<li><a href="https://bioconductor.org" target="blank">Published by Bioconductor</a></li>

</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">Assigning cell types with SingleR</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="harmonizing-labels" class="section level1">
<h1><span class="header-section-number">Chapter 6</span> Harmonizing labels</h1>
<p>The <code>matchReferences()</code> function provides a simple yet elegant approach for label harmonization between two references. Each reference is used to annotate the other and the probability of mutual assignment between each pair of labels is computed. Probabilities close to 1 indicate there is a 1:1 relation between that pair of labels; on the other hand, an all-zero probability vector indicates that a label is unique to a particular reference.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(SingleR)
bp.se &lt;-<span class="st"> </span><span class="kw">BlueprintEncodeData</span>()
hpca.se &lt;-<span class="st"> </span><span class="kw">HumanPrimaryCellAtlasData</span>()

matched &lt;-<span class="st"> </span><span class="kw">matchReferences</span>(bp.se, hpca.se,
    bp.se<span class="op">$</span>label.main, hpca.se<span class="op">$</span>label.main)
pheatmap<span class="op">::</span><span class="kw">pheatmap</span>(matched, <span class="dt">col=</span>viridis<span class="op">::</span><span class="kw">plasma</span>(<span class="dv">100</span>))</code></pre></div>
<p><img src="06-ontology_files/figure-html/unnamed-chunk-2-1.png" width="672" /></p>
<p>A heatmap like the one above can be used to guide harmonization to enforce a consistent vocabulary across all labels representing the same cell type or state. The most obvious benefit of harmonization is that interpretation of the results is simplified. However, an even more important effect is that the presence of harmonized labels from multiple references allows the classification machinery to protect against irrelevant batch effects between references. For example, in <code>SingleR()</code>’s case, marker genes are favored if they are consistently upregulated across multiple references, improving robustness to technical idiosyncrasies in any test dataset.</p>
<p>We stress that some manual intervention is still required in this process, given the risks posed by differences in biological systems and technologies. For example, neurons are considered unique to each reference while smooth muscle cells in the HPCA data are incorrectly matched to fibroblasts in the Blueprint/ENCODE data. CD4<sup>+</sup> and CD8<sup>+</sup> T cells are also both assigned to “T cells”, so some decision about the acceptable resolution of the harmonized labels is required here.</p>
<p>As an aside, we can also use this function to identify the matching clusters between two independent scRNA-seq analyses. This is an “off-label” use that involves substituting the cluster assignments as proxies for the labels. We can then match up clusters and integrate conclusions from multiple datasets without the difficulty of batch correction and reclustering.</p>

</div>
            </section>

          </div>
        </div>
      </div>
<a href="using-multiple-references.html" class="navigation navigation-prev " aria-label="Previous page"><i class="fa fa-angle-left"></i></a>
<a href="advanced-options.html" class="navigation navigation-next " aria-label="Next page"><i class="fa fa-angle-right"></i></a>
    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/lunr.js"></script>
<script src="libs/gitbook-2.6.7/js/clipboard.min.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-clipboard.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": {
"github": false,
"facebook": false,
"twitter": false,
"linkedin": false,
"weibo": false,
"instapaper": false,
"vk": false,
"all": ["facebook", "twitter", "linkedin", "weibo", "instapaper"]
},
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": {
"link": null,
"text": null
},
"history": {
"link": null,
"text": null
},
"view": {
"link": null,
"text": null
},
"download": null,
"toc": {
"collapse": "subsection"
},
"bookdown::epub_book": "default",
"cover": "images/cover.png"
});
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    var src = "";
    if (src === "" || src === "true") src = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-MML-AM_CHTML";
    if (location.protocol !== "file:")
      if (/^https?:/.test(src))
        src = src.replace(/^https?:/, '');
    script.src = src;
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>
</body>

</html>
