<!DOCTYPE html>
<html lang="" xml:lang="">
<head>

  <meta charset="utf-8" />
  <meta http-equiv="X-UA-Compatible" content="IE=edge" />
  <title>Chapter 3 Using single-cell references | Assigning cell types with SingleR</title>
  <meta name="description" content="The SingleR book. Because sometimes, a vignette just isn’t enough." />
  <meta name="generator" content="bookdown 0.18 and GitBook 2.6.7" />

  <meta property="og:title" content="Chapter 3 Using single-cell references | Assigning cell types with SingleR" />
  <meta property="og:type" content="book" />
  
  
  <meta property="og:description" content="The SingleR book. Because sometimes, a vignette just isn’t enough." />
  

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="Chapter 3 Using single-cell references | Assigning cell types with SingleR" />
  
  <meta name="twitter:description" content="The SingleR book. Because sometimes, a vignette just isn’t enough." />
  



<meta name="date" content="2020-05-13" />

  <meta name="viewport" content="width=device-width, initial-scale=1" />
  <meta name="apple-mobile-web-app-capable" content="yes" />
  <meta name="apple-mobile-web-app-status-bar-style" content="black" />
  
  
<link rel="prev" href="using-the-built-in-references.html"/>
<link rel="next" href="annotation-diagnostics.html"/>
<script src="libs/jquery-2.2.3/jquery.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-table.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-clipboard.css" rel="stylesheet" />











<style type="text/css">
div.sourceCode { overflow-x: auto; }
table.sourceCode, tr.sourceCode, td.lineNumbers, td.sourceCode {
  margin: 0; padding: 0; vertical-align: baseline; border: none; }
table.sourceCode { width: 100%; line-height: 100%; background-color: #f8f8f8; }
td.lineNumbers { text-align: right; padding-right: 4px; padding-left: 4px; color: #aaaaaa; border-right: 1px solid #aaaaaa; }
td.sourceCode { padding-left: 5px; }
pre, code { background-color: #f8f8f8; }
code > span.kw { color: #204a87; font-weight: bold; } /* Keyword */
code > span.dt { color: #204a87; } /* DataType */
code > span.dv { color: #0000cf; } /* DecVal */
code > span.bn { color: #0000cf; } /* BaseN */
code > span.fl { color: #0000cf; } /* Float */
code > span.ch { color: #4e9a06; } /* Char */
code > span.st { color: #4e9a06; } /* String */
code > span.co { color: #8f5902; font-style: italic; } /* Comment */
code > span.ot { color: #8f5902; } /* Other */
code > span.al { color: #ef2929; } /* Alert */
code > span.fu { color: #000000; } /* Function */
code > span.er { color: #a40000; font-weight: bold; } /* Error */
code > span.wa { color: #8f5902; font-weight: bold; font-style: italic; } /* Warning */
code > span.cn { color: #000000; } /* Constant */
code > span.sc { color: #000000; } /* SpecialChar */
code > span.vs { color: #4e9a06; } /* VerbatimString */
code > span.ss { color: #4e9a06; } /* SpecialString */
code > span.im { } /* Import */
code > span.va { color: #000000; } /* Variable */
code > span.cf { color: #204a87; font-weight: bold; } /* ControlFlow */
code > span.op { color: #ce5c00; font-weight: bold; } /* Operator */
code > span.pp { color: #8f5902; font-style: italic; } /* Preprocessor */
code > span.ex { } /* Extension */
code > span.at { color: #c4a000; } /* Attribute */
code > span.do { color: #8f5902; font-weight: bold; font-style: italic; } /* Documentation */
code > span.an { color: #8f5902; font-weight: bold; font-style: italic; } /* Annotation */
code > span.cv { color: #8f5902; font-weight: bold; font-style: italic; } /* CommentVar */
code > span.in { color: #8f5902; font-weight: bold; font-style: italic; } /* Information */
</style>

<link rel="stylesheet" href="style.css" type="text/css" />
</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li><a href="./">The SingleR book</a></li>

<li class="divider"></li>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html"><i class="fa fa-check"></i>Preface</a></li>
<li class="chapter" data-level="1" data-path="introduction.html"><a href="introduction.html"><i class="fa fa-check"></i><b>1</b> Introduction</a><ul>
<li class="chapter" data-level="1.1" data-path="introduction.html"><a href="introduction.html#motivation"><i class="fa fa-check"></i><b>1.1</b> Motivation</a></li>
<li class="chapter" data-level="1.2" data-path="introduction.html"><a href="introduction.html#method-description"><i class="fa fa-check"></i><b>1.2</b> Method description</a></li>
<li class="chapter" data-level="1.3" data-path="introduction.html"><a href="introduction.html#quick-start"><i class="fa fa-check"></i><b>1.3</b> Quick start</a></li>
<li class="chapter" data-level="1.4" data-path="introduction.html"><a href="introduction.html#where-to-get-help"><i class="fa fa-check"></i><b>1.4</b> Where to get help</a></li>
<li class="chapter" data-level="" data-path="introduction.html"><a href="introduction.html#session-information"><i class="fa fa-check"></i>Session information</a></li>
<li class="chapter" data-level="" data-path="introduction.html"><a href="introduction.html#references"><i class="fa fa-check"></i>References</a></li>
</ul></li>
<li class="chapter" data-level="2" data-path="using-the-built-in-references.html"><a href="using-the-built-in-references.html"><i class="fa fa-check"></i><b>2</b> Using the built-in references</a><ul>
<li class="chapter" data-level="2.1" data-path="using-the-built-in-references.html"><a href="using-the-built-in-references.html#overview"><i class="fa fa-check"></i><b>2.1</b> Overview</a></li>
<li class="chapter" data-level="2.2" data-path="using-the-built-in-references.html"><a href="using-the-built-in-references.html#annotation-with-default-marker-detection"><i class="fa fa-check"></i><b>2.2</b> Annotation with default marker detection</a></li>
<li class="chapter" data-level="2.3" data-path="using-the-built-in-references.html"><a href="using-the-built-in-references.html#choices-of-assay-data"><i class="fa fa-check"></i><b>2.3</b> Choices of assay data</a></li>
<li class="chapter" data-level="" data-path="using-the-built-in-references.html"><a href="using-the-built-in-references.html#session-information-1"><i class="fa fa-check"></i>Session information</a></li>
<li class="chapter" data-level="" data-path="using-the-built-in-references.html"><a href="using-the-built-in-references.html#references-1"><i class="fa fa-check"></i>References</a></li>
</ul></li>
<li class="chapter" data-level="3" data-path="using-single-cell-references.html"><a href="using-single-cell-references.html"><i class="fa fa-check"></i><b>3</b> Using single-cell references</a><ul>
<li class="chapter" data-level="3.1" data-path="using-single-cell-references.html"><a href="using-single-cell-references.html#overview-1"><i class="fa fa-check"></i><b>3.1</b> Overview</a></li>
<li class="chapter" data-level="3.2" data-path="using-single-cell-references.html"><a href="using-single-cell-references.html#annotation-with-test-based-marker-detection"><i class="fa fa-check"></i><b>3.2</b> Annotation with test-based marker detection</a></li>
<li class="chapter" data-level="3.3" data-path="using-single-cell-references.html"><a href="using-single-cell-references.html#defining-custom-markers"><i class="fa fa-check"></i><b>3.3</b> Defining custom markers</a></li>
<li class="chapter" data-level="3.4" data-path="using-single-cell-references.html"><a href="using-single-cell-references.html#pseudo-bulk-aggregation"><i class="fa fa-check"></i><b>3.4</b> Pseudo-bulk aggregation</a></li>
<li class="chapter" data-level="" data-path="using-single-cell-references.html"><a href="using-single-cell-references.html#session-information-2"><i class="fa fa-check"></i>Session information</a></li>
<li class="chapter" data-level="3.5" data-path="using-single-cell-references.html"><a href="using-single-cell-references.html#references-2"><i class="fa fa-check"></i><b>3.5</b> References</a></li>
</ul></li>
<li class="chapter" data-level="4" data-path="annotation-diagnostics.html"><a href="annotation-diagnostics.html"><i class="fa fa-check"></i><b>4</b> Annotation diagnostics</a><ul>
<li class="chapter" data-level="4.1" data-path="annotation-diagnostics.html"><a href="annotation-diagnostics.html#based-on-the-scores-within-cells"><i class="fa fa-check"></i><b>4.1</b> Based on the scores within cells</a></li>
<li class="chapter" data-level="4.2" data-path="annotation-diagnostics.html"><a href="annotation-diagnostics.html#based-on-the-deltas-across-cells"><i class="fa fa-check"></i><b>4.2</b> Based on the deltas across cells</a></li>
<li class="chapter" data-level="4.3" data-path="annotation-diagnostics.html"><a href="annotation-diagnostics.html#based-on-marker-gene-expression"><i class="fa fa-check"></i><b>4.3</b> Based on marker gene expression</a></li>
</ul></li>
<li class="chapter" data-level="5" data-path="using-multiple-references.html"><a href="using-multiple-references.html"><i class="fa fa-check"></i><b>5</b> Using multiple references</a></li>
<li class="chapter" data-level="6" data-path="harmonizing-labels.html"><a href="harmonizing-labels.html"><i class="fa fa-check"></i><b>6</b> Harmonizing labels</a></li>
<li class="chapter" data-level="7" data-path="advanced-options.html"><a href="advanced-options.html"><i class="fa fa-check"></i><b>7</b> Advanced options</a><ul>
<li class="chapter" data-level="7.1" data-path="advanced-options.html"><a href="advanced-options.html#preconstructed-indices"><i class="fa fa-check"></i><b>7.1</b> Preconstructed indices</a></li>
<li class="chapter" data-level="7.2" data-path="advanced-options.html"><a href="advanced-options.html#parallelization"><i class="fa fa-check"></i><b>7.2</b> Parallelization</a></li>
<li class="chapter" data-level="7.3" data-path="advanced-options.html"><a href="advanced-options.html#approximate-algorithms"><i class="fa fa-check"></i><b>7.3</b> Approximate algorithms</a></li>
<li class="chapter" data-level="" data-path="advanced-options.html"><a href="advanced-options.html#session-information-3"><i class="fa fa-check"></i>Session information</a></li>
<li class="chapter" data-level="" data-path="advanced-options.html"><a href="advanced-options.html#references-3"><i class="fa fa-check"></i>References</a></li>
</ul></li>
<li class="divider"></li>
<li><a href="https://bioconductor.org" target="blank">Published by Bioconductor</a></li>

</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">Assigning cell types with SingleR</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="using-single-cell-references" class="section level1">
<h1><span class="header-section-number">Chapter 3</span> Using single-cell references</h1>
<div id="overview-1" class="section level2">
<h2><span class="header-section-number">3.1</span> Overview</h2>
<p>Another application of <em><a href="https://bioconductor.org/packages/3.12/SingleR">SingleR</a></em> involves using one single-cell dataset as a reference to annotate others. This is almost identical to the use of the built-in bulk references described in Chapter <a href="using-the-built-in-references.html#using-the-built-in-references">2</a>, with the main difference involving the marker detection strategy that is applied to the reference labels. In particular, we identify top-ranked markers based on pairwise Wilcoxon rank sum tests or <span class="math inline">\(t\)</span>-tests between labels; this allows us to account for the variability across cells to choose genes that are robustly upregulated in each label.</p>
<p>Despite the name of this chapter, the functionality described here is not limited to single-cell references. The determining factor is whether a dataset contains enough samples for each label to permit the use of statistical tests for marker detection. This scenario is most common for single-cell references but the same approach can be applied to bulk references with many replicate samples.</p>
</div>
<div id="annotation-with-test-based-marker-detection" class="section level2">
<h2><span class="header-section-number">3.2</span> Annotation with test-based marker detection</h2>
<p>Here, we will use two human pancreas scRNA-seq datasets from the <em><a href="https://bioconductor.org/packages/3.12/scRNAseq">scRNAseq</a></em> package. The aim is to use one pre-labelled dataset to annotate the other unlabelled dataset. First, we set up the <span class="citation">Muraro et al. (<a href="#ref-muraro2016singlecell">2016</a>)</span> dataset to be our reference, computing log-normalized expression values as discussed in Section <a href="using-the-built-in-references.html#choices-of-assay-data">2.3</a>.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(scRNAseq)
sceM &lt;-<span class="st"> </span><span class="kw">MuraroPancreasData</span>()

<span class="co"># Removing unlabelled cells or cells without a clear label.</span>
sceM &lt;-<span class="st"> </span>sceM[,<span class="op">!</span><span class="kw">is.na</span>(sceM<span class="op">$</span>label) <span class="op">&amp;</span><span class="st"> </span>sceM<span class="op">$</span>label<span class="op">!=</span><span class="st">&quot;unclear&quot;</span>] 

<span class="kw">library</span>(scater)
sceM &lt;-<span class="st"> </span><span class="kw">logNormCounts</span>(sceM)
sceM</code></pre></div>
<pre><code>## class: SingleCellExperiment 
## dim: 19059 2122 
## metadata(0):
## assays(2): counts logcounts
## rownames(19059): A1BG-AS1__chr19 A1BG__chr19 ... ZZEF1__chr17
##   ZZZ3__chr1
## rowData names(2): symbol chr
## colnames(2122): D28-1_1 D28-1_2 ... D30-8_93 D30-8_94
## colData names(4): label donor plate sizeFactor
## reducedDimNames(0):
## altExpNames(1): ERCC</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># Seeing the available labels in this dataset.</span>
<span class="kw">table</span>(sceM<span class="op">$</span>label)</code></pre></div>
<pre><code>## 
##      acinar       alpha        beta       delta        duct endothelial 
##         219         812         448         193         245          21 
##     epsilon mesenchymal          pp 
##           3          80         101</code></pre>
<p>We then set up our test dataset from <span class="citation">Grun et al. (<a href="#ref-grun2016denovo">2016</a>)</span>, applying some basic quality control as discusssed <a href="https://osca.bioconductor.org/grun-human-pancreas-cel-seq2.html#quality-control-8">here</a>. (As <code>SingleR()</code> operates independently on each cell, quality control on the test dataset can actually be applied after annotation is performed. We only apply it beforehand here to simplify the interpretation of the downstream results.) We also compute the log-transformed values here, not because it is strictly necessary but so that we don’t have to keep on typing <code>assay.type.test=1</code> in later calls to <code>SingleR()</code>.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">sceG &lt;-<span class="st"> </span><span class="kw">GrunPancreasData</span>()

sceG &lt;-<span class="st"> </span><span class="kw">addPerCellQC</span>(sceG)
qc &lt;-<span class="st"> </span><span class="kw">quickPerCellQC</span>(<span class="kw">colData</span>(sceG), 
    <span class="dt">percent_subsets=</span><span class="st">&quot;altexps_ERCC_percent&quot;</span>,
    <span class="dt">batch=</span>sceG<span class="op">$</span>donor,
    <span class="dt">subset=</span>sceG<span class="op">$</span>donor <span class="op">%in%</span><span class="st"> </span><span class="kw">c</span>(<span class="st">&quot;D17&quot;</span>, <span class="st">&quot;D7&quot;</span>, <span class="st">&quot;D2&quot;</span>))
sceG &lt;-<span class="st"> </span>sceG[,<span class="op">!</span>qc<span class="op">$</span>discard]

sceG &lt;-<span class="st"> </span><span class="kw">logNormCounts</span>(sceG)
sceG</code></pre></div>
<pre><code>## class: SingleCellExperiment 
## dim: 20064 1064 
## metadata(0):
## assays(2): counts logcounts
## rownames(20064): A1BG-AS1__chr19 A1BG__chr19 ... ZZEF1__chr17
##   ZZZ3__chr1
## rowData names(2): symbol chr
## colnames(1064): D2ex_1 D2ex_2 ... D17TGFB_94 D17TGFB_95
## colData names(13): donor sample ... total sizeFactor
## reducedDimNames(0):
## altExpNames(1): ERCC</code></pre>
<p>We run <code>SingleR()</code> as described previously but with a marker detection mode that considers the variance of expression across cells. Here, we will use the Wilcoxon ranked sum test to identify the top markers for each pairwise comparison between labels. This is slower but more appropriate for single-cell data compared to the default marker detection algorithm, which may fail for low-coverage data where the median for each label is often zero.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(SingleR)
pred.grun &lt;-<span class="st"> </span><span class="kw">SingleR</span>(<span class="dt">test=</span>sceG, <span class="dt">ref=</span>sceM, <span class="dt">labels=</span>sceM<span class="op">$</span>label, <span class="dt">de.method=</span><span class="st">&quot;wilcox&quot;</span>)
<span class="kw">table</span>(pred.grun<span class="op">$</span>labels)</code></pre></div>
<pre><code>## 
## acinar   beta  delta   duct 
##     53      4      2     41</code></pre>
<p>By default, the function will take the top <code>de.n</code> (default: 10) genes from each pairwise comparison between labels. A larger number of markers increases the robustness of the annotation by ensuring that relevant genes are not omitted, especially if the reference dataset has study-specific effects that cause uninteresting genes to dominate the top set. However, this comes at the cost of increasing noise and computational time.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(SingleR)
pred.grun &lt;-<span class="st"> </span><span class="kw">SingleR</span>(<span class="dt">test=</span>sceG, <span class="dt">ref=</span>sceM, <span class="dt">labels=</span>sceM<span class="op">$</span>label, 
    <span class="dt">de.method=</span><span class="st">&quot;wilcox&quot;</span>, <span class="dt">de.n=</span><span class="dv">50</span>)
<span class="kw">table</span>(pred.grun<span class="op">$</span>labels)</code></pre></div>
<pre><code>## 
##      acinar       alpha        beta       delta        duct endothelial 
##         275         203         177          55         307           5 
##     epsilon mesenchymal          pp 
##           1          23          18</code></pre>
</div>
<div id="defining-custom-markers" class="section level2">
<h2><span class="header-section-number">3.3</span> Defining custom markers</h2>
<p>The marker detection in <code>SingleR()</code> is built on top of the testing framework in <em><a href="https://bioconductor.org/packages/3.12/scran">scran</a></em>, so most options in <code>?pairwiseWilcox</code> and friends can be applied via the <code>de.args=</code> option. For example, we could use the <span class="math inline">\(t\)</span>-test and test against a log-fold change threshold with <code>de.args=list(lfc=1)</code>.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(SingleR)
pred.grun2 &lt;-<span class="st"> </span><span class="kw">SingleR</span>(<span class="dt">test=</span>sceG, <span class="dt">ref=</span>sceM, <span class="dt">labels=</span>sceM<span class="op">$</span>label, 
    <span class="dt">de.method=</span><span class="st">&quot;t&quot;</span>, <span class="dt">de.args=</span><span class="kw">list</span>(<span class="dt">lfc=</span><span class="dv">1</span>))
<span class="kw">table</span>(pred.grun2<span class="op">$</span>labels)</code></pre></div>
<pre><code>## 
##      acinar       alpha        beta       delta        duct endothelial 
##         285         200         177          54         296           5 
##     epsilon mesenchymal          pp 
##           5          24          18</code></pre>
<p>However, users can also construct their own marker lists with any DE testing machinery. For example, we can perform pairwise binomial tests to identify genes that are differentially detected (i.e., have differences in the proportion of cells with non-zero counts) between labels in the reference Muraro dataset. We then take the top 10 marker genes from each pairwise comparison, obtaining a list of lists of character vectors containing the identities of the markers for that comparison.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(scran)
out &lt;-<span class="st"> </span><span class="kw">pairwiseBinom</span>(<span class="kw">counts</span>(sceM), sceM<span class="op">$</span>label, <span class="dt">direction=</span><span class="st">&quot;up&quot;</span>)
markers &lt;-<span class="st"> </span><span class="kw">getTopMarkers</span>(out<span class="op">$</span>statistics, out<span class="op">$</span>pairs, <span class="dt">n=</span><span class="dv">10</span>)

<span class="co"># Upregulated in acinar compared to alpha:</span>
markers<span class="op">$</span>acinar<span class="op">$</span>alpha</code></pre></div>
<pre><code>##  [1] &quot;KCNQ1__chr11&quot;  &quot;FAM129A__chr1&quot; &quot;KLK1__chr19&quot;   &quot;NTN4__chr12&quot;  
##  [5] &quot;RASEF__chr9&quot;   &quot;CTRL__chr16&quot;   &quot;LGALS2__chr22&quot; &quot;NUPR1__chr16&quot; 
##  [9] &quot;LGALS3__chr14&quot; &quot;NR5A2__chr1&quot;</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># Upregulated in alpha compared to acinar:</span>
markers<span class="op">$</span>alpha<span class="op">$</span>acinar</code></pre></div>
<pre><code>##  [1] &quot;SLC38A4__chr12&quot; &quot;ARX__chrX&quot;      &quot;CRYBA2__chr2&quot;   &quot;FSTL5__chr4&quot;   
##  [5] &quot;GNG2__chr14&quot;    &quot;NOL4__chr18&quot;    &quot;IRX2__chr5&quot;     &quot;KCNMB2__chr3&quot;  
##  [9] &quot;CFC1__chr2&quot;     &quot;KCNJ6__chr21&quot;</code></pre>
<p>Once we have this list of lists, we supply it to <code>SingleR()</code> via the <code>genes=</code> argument, which causes the function to bypass the internal marker detection to use the supplied gene sets instead. The most obvious benefit of this approach is that the user can achieve greater control of the markers, allowing integration of prior biological knowledge to obtain more relevant genes and a more robust annotation.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">pred.grun2b &lt;-<span class="st"> </span><span class="kw">SingleR</span>(<span class="dt">test=</span>sceG, <span class="dt">ref=</span>sceM, <span class="dt">labels=</span>sceM<span class="op">$</span>label, <span class="dt">genes=</span>markers)
<span class="kw">table</span>(pred.grun2b<span class="op">$</span>labels)</code></pre></div>
<pre><code>## 
##      acinar       alpha        beta       delta        duct endothelial 
##         276         202         175          54         302           5 
##     epsilon mesenchymal          pp 
##           2          25          23</code></pre>
<p>In some cases, markers may only be available for specific labels rather than for pairwise comparisons between labels. This is accommodated by supplying a named list of character vectors to <code>genes</code>. Note that this is likely to be less powerful than the list-of-lists approach as information about pairwise differences is discarded.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># Creating label-specific markers.</span>
label.markers &lt;-<span class="st"> </span><span class="kw">lapply</span>(markers, unlist)
label.markers &lt;-<span class="st"> </span><span class="kw">lapply</span>(label.markers, unique)
<span class="kw">str</span>(label.markers)</code></pre></div>
<pre><code>## List of 9
##  $ acinar     : chr [1:40] &quot;KCNQ1__chr11&quot; &quot;FAM129A__chr1&quot; &quot;KLK1__chr19&quot; &quot;NTN4__chr12&quot; ...
##  $ alpha      : chr [1:41] &quot;SLC38A4__chr12&quot; &quot;ARX__chrX&quot; &quot;CRYBA2__chr2&quot; &quot;FSTL5__chr4&quot; ...
##  $ beta       : chr [1:47] &quot;ELAVL4__chr1&quot; &quot;PRUNE2__chr9&quot; &quot;NMNAT2__chr1&quot; &quot;PLCB4__chr20&quot; ...
##  $ delta      : chr [1:44] &quot;NOL4__chr18&quot; &quot;CABP7__chr22&quot; &quot;UNC80__chr2&quot; &quot;HEPACAM2__chr7&quot; ...
##  $ duct       : chr [1:50] &quot;ADCY5__chr3&quot; &quot;PDE3A__chr12&quot; &quot;SLC3A1__chr2&quot; &quot;BICC1__chr10&quot; ...
##  $ endothelial: chr [1:26] &quot;GPR4__chr19&quot; &quot;TMEM204__chr16&quot; &quot;GPR116__chr6&quot; &quot;CYYR1__chr21&quot; ...
##  $ epsilon    : chr [1:14] &quot;BHMT__chr5&quot; &quot;JPH3__chr16&quot; &quot;SERPINA10__chr14&quot; &quot;UGT2B4__chr4&quot; ...
##  $ mesenchymal: chr [1:34] &quot;TNFAIP6__chr2&quot; &quot;THBS2__chr6&quot; &quot;CDH11__chr16&quot; &quot;SRPX2__chrX&quot; ...
##  $ pp         : chr [1:44] &quot;SERTM1__chr13&quot; &quot;ETV1__chr7&quot; &quot;ARX__chrX&quot; &quot;ELAVL4__chr1&quot; ...</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">pred.grun2c &lt;-<span class="st"> </span><span class="kw">SingleR</span>(<span class="dt">test=</span>sceG, <span class="dt">ref=</span>sceM, <span class="dt">labels=</span>sceM<span class="op">$</span>label, <span class="dt">genes=</span>label.markers)
<span class="kw">table</span>(pred.grun2c<span class="op">$</span>labels)</code></pre></div>
<pre><code>## 
##      acinar       alpha        beta       delta        duct endothelial 
##         262         204         169          59         317           6 
##     epsilon mesenchymal          pp 
##           2          24          21</code></pre>
</div>
<div id="pseudo-bulk-aggregation" class="section level2">
<h2><span class="header-section-number">3.4</span> Pseudo-bulk aggregation</h2>
<p>Single-cell reference datasets provide a like-for-like comparison to our test single-cell datasets, yielding a more accurate classification of the cells in the latter (hopefully). However, there are frequently many more samples in single-cell references compared to bulk references, increasing the computational work involved in classification. We avoid this by aggregating cells into one “pseudo-bulk” sample per label (e.g., by averaging across log-expression values) and using those as the reference, which allows us to achieve the same efficiency as the use of bulk references.</p>
<p>The obvious cost of this approach is that we discard potentially useful information about the distribution of cells within each label. Cells that belong to a heterogeneous population may not be correctly assigned if they are far from the population center. We attempt to preserve some of this information by using <span class="math inline">\(k\)</span>-means clustering within each cell to create pseudo-bulk samples that are representative of a particular region of the expression space (i.e., vector quantization). We create <span class="math inline">\(\sqrt{N}\)</span> clusters given a label with <span class="math inline">\(N\)</span> cells, which provides a reasonable compromise between reducing computational work and preserving the label’s internal distribution.</p>
<p>To enable this aggregation, we simply set <code>aggr.ref=TRUE</code> in the <code>SingleR()</code> call. This uses the <code>aggregateReference()</code> function to perform <span class="math inline">\(k\)</span>-means clustering <em>within</em> each label (typically after principal components analysis on the log-expression matrix, for greater speed) and average expression values for each within-label cluster. Note that marker detection is still performed on the unaggregated data so as to make full use of the distribution of expression values across cells.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">set.seed</span>(<span class="dv">100</span>) <span class="co"># for the k-means step.</span>
pred.grun3 &lt;-<span class="st"> </span><span class="kw">SingleR</span>(<span class="dt">test=</span>sceG, <span class="dt">ref=</span>sceM, <span class="dt">labels=</span>sceM<span class="op">$</span>label, 
    <span class="dt">de.method=</span><span class="st">&quot;wilcox&quot;</span>, <span class="dt">aggr.ref=</span><span class="ot">TRUE</span>)
<span class="kw">table</span>(pred.grun3<span class="op">$</span>labels)</code></pre></div>
<pre><code>## 
##      acinar       alpha        beta       delta        duct endothelial 
##         277         202         184          47         306           5 
##     epsilon mesenchymal          pp 
##           1          22          20</code></pre>
<p>Obviously, the aggregation itself requires computational work so setting <code>aggr.ref=TRUE</code> in <code>SingleR()</code> itself may not improve speed. Rather, the real power of this approach lies in pre-aggregating the reference dataset so that it can be repeatedly applied to quickly annotate multiple test datasets. This approach is discussed in more detail in Chapter <a href="advanced-options.html#advanced-options">7</a>.</p>
</div>
<div id="session-information-2" class="section level2 unnumbered">
<h2>Session information</h2>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">sessionInfo</span>()</code></pre></div>
<pre><code>## R version 4.0.0 Patched (2020-05-01 r78341)
## Platform: x86_64-pc-linux-gnu (64-bit)
## Running under: Ubuntu 18.04.4 LTS
## 
## Matrix products: default
## BLAS:   /home/luna/Software/R/R-4-0-branch-dev/lib/libRblas.so
## LAPACK: /home/luna/Software/R/R-4-0-branch-dev/lib/libRlapack.so
## 
## locale:
##  [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              
##  [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    
##  [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   
##  [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 
##  [9] LC_ADDRESS=C               LC_TELEPHONE=C            
## [11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       
## 
## attached base packages:
## [1] stats4    parallel  stats     graphics  grDevices utils     datasets 
## [8] methods   base     
## 
## other attached packages:
##  [1] scran_1.17.0                SingleR_1.3.2              
##  [3] scater_1.17.0               ggplot2_3.3.0              
##  [5] scRNAseq_2.3.0              SingleCellExperiment_1.11.1
##  [7] SummarizedExperiment_1.19.2 DelayedArray_0.15.1        
##  [9] matrixStats_0.56.0          Biobase_2.49.0             
## [11] GenomicRanges_1.41.1        GenomeInfoDb_1.25.0        
## [13] IRanges_2.23.4              S4Vectors_0.27.5           
## [15] BiocGenerics_0.35.2         BiocStyle_2.17.0           
## 
## loaded via a namespace (and not attached):
##  [1] bitops_1.0-6                  bit64_0.9-7                  
##  [3] httr_1.4.1                    tools_4.0.0                  
##  [5] R6_2.4.1                      irlba_2.3.3                  
##  [7] vipor_0.4.5                   DBI_1.1.0                    
##  [9] colorspace_1.4-1              withr_2.2.0                  
## [11] tidyselect_1.0.0              gridExtra_2.3                
## [13] bit_1.1-15.2                  curl_4.3                     
## [15] compiler_4.0.0                BiocNeighbors_1.7.0          
## [17] bookdown_0.18                 scales_1.1.0                 
## [19] rappdirs_0.3.1                stringr_1.4.0                
## [21] digest_0.6.25                 rmarkdown_2.1                
## [23] XVector_0.29.0                pkgconfig_2.0.3              
## [25] htmltools_0.4.0               limma_3.45.0                 
## [27] dbplyr_1.4.3                  fastmap_1.0.1                
## [29] rlang_0.4.6                   RSQLite_2.2.0                
## [31] shiny_1.4.0.2                 DelayedMatrixStats_1.11.0    
## [33] BiocParallel_1.23.0           dplyr_0.8.5                  
## [35] RCurl_1.98-1.2                magrittr_1.5                 
## [37] BiocSingular_1.5.0            scuttle_0.99.5               
## [39] GenomeInfoDbData_1.2.3        Matrix_1.2-18                
## [41] Rcpp_1.0.4.6                  ggbeeswarm_0.6.0             
## [43] munsell_0.5.0                 viridis_0.5.1                
## [45] lifecycle_0.2.0               stringi_1.4.6                
## [47] yaml_2.2.1                    edgeR_3.31.0                 
## [49] zlibbioc_1.35.0               BiocFileCache_1.13.0         
## [51] AnnotationHub_2.21.0          grid_4.0.0                   
## [53] blob_1.2.1                    promises_1.1.0               
## [55] dqrng_0.2.1                   ExperimentHub_1.15.0         
## [57] crayon_1.3.4                  lattice_0.20-41              
## [59] locfit_1.5-9.4                knitr_1.28                   
## [61] pillar_1.4.4                  igraph_1.2.5                 
## [63] codetools_0.2-16              glue_1.4.0                   
## [65] BiocVersion_3.12.0            evaluate_0.14                
## [67] BiocManager_1.30.10           vctrs_0.2.4                  
## [69] httpuv_1.5.2                  gtable_0.3.0                 
## [71] purrr_0.3.4                   assertthat_0.2.1             
## [73] xfun_0.13                     rsvd_1.0.3                   
## [75] mime_0.9                      xtable_1.8-4                 
## [77] later_1.0.0                   viridisLite_0.3.0            
## [79] tibble_3.0.1                  AnnotationDbi_1.51.0         
## [81] beeswarm_0.2.3                memoise_1.1.0                
## [83] statmod_1.4.34                ellipsis_0.3.0               
## [85] interactiveDisplayBase_1.27.0</code></pre>
</div>
<div id="references-2" class="section level2">
<h2><span class="header-section-number">3.5</span> References</h2>

</div>
</div>
<h3>References</h3>
<div id="refs" class="references">
<div id="ref-grun2016denovo">
<p>Grun, D., M. J. Muraro, J. C. Boisset, K. Wiebrands, A. Lyubimova, G. Dharmadhikari, M. van den Born, et al. 2016. “De Novo Prediction of Stem Cell Identity using Single-Cell Transcriptome Data.” <em>Cell Stem Cell</em> 19 (2): 266–77.</p>
</div>
<div id="ref-muraro2016singlecell">
<p>Muraro, M. J., G. Dharmadhikari, D. Grun, N. Groen, T. Dielen, E. Jansen, L. van Gurp, et al. 2016. “A Single-Cell Transcriptome Atlas of the Human Pancreas.” <em>Cell Syst</em> 3 (4): 385–94.</p>
</div>
</div>
            </section>

          </div>
        </div>
      </div>
<a href="using-the-built-in-references.html" class="navigation navigation-prev " aria-label="Previous page"><i class="fa fa-angle-left"></i></a>
<a href="annotation-diagnostics.html" class="navigation navigation-next " aria-label="Next page"><i class="fa fa-angle-right"></i></a>
    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/lunr.js"></script>
<script src="libs/gitbook-2.6.7/js/clipboard.min.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-clipboard.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": {
"github": false,
"facebook": false,
"twitter": false,
"linkedin": false,
"weibo": false,
"instapaper": false,
"vk": false,
"all": ["facebook", "twitter", "linkedin", "weibo", "instapaper"]
},
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": {
"link": null,
"text": null
},
"history": {
"link": null,
"text": null
},
"view": {
"link": null,
"text": null
},
"download": null,
"toc": {
"collapse": "subsection"
},
"bookdown::epub_book": "default",
"cover": "images/cover.png"
});
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    var src = "";
    if (src === "" || src === "true") src = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-MML-AM_CHTML";
    if (location.protocol !== "file:")
      if (/^https?:/.test(src))
        src = src.replace(/^https?:/, '');
    script.src = src;
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>
</body>

</html>
