<!DOCTYPE html>
<html lang="" xml:lang="">
<head>

  <meta charset="utf-8" />
  <meta http-equiv="X-UA-Compatible" content="IE=edge" />
  <title>Chapter 2 Using the built-in references | Assigning cell types with SingleR</title>
  <meta name="description" content="The SingleR book. Because sometimes, a vignette just isn’t enough." />
  <meta name="generator" content="bookdown 0.19 and GitBook 2.6.7" />

  <meta property="og:title" content="Chapter 2 Using the built-in references | Assigning cell types with SingleR" />
  <meta property="og:type" content="book" />
  
  
  <meta property="og:description" content="The SingleR book. Because sometimes, a vignette just isn’t enough." />
  

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="Chapter 2 Using the built-in references | Assigning cell types with SingleR" />
  
  <meta name="twitter:description" content="The SingleR book. Because sometimes, a vignette just isn’t enough." />
  



<meta name="date" content="2020-05-24" />

  <meta name="viewport" content="width=device-width, initial-scale=1" />
  <meta name="apple-mobile-web-app-capable" content="yes" />
  <meta name="apple-mobile-web-app-status-bar-style" content="black" />
  
  
<link rel="prev" href="introduction.html"/>
<link rel="next" href="using-single-cell-references.html"/>
<script src="libs/jquery-2.2.3/jquery.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-table.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-clipboard.css" rel="stylesheet" />











<style type="text/css">
div.sourceCode { overflow-x: auto; }
table.sourceCode, tr.sourceCode, td.lineNumbers, td.sourceCode {
  margin: 0; padding: 0; vertical-align: baseline; border: none; }
table.sourceCode { width: 100%; line-height: 100%; background-color: #f8f8f8; }
td.lineNumbers { text-align: right; padding-right: 4px; padding-left: 4px; color: #aaaaaa; border-right: 1px solid #aaaaaa; }
td.sourceCode { padding-left: 5px; }
pre, code { background-color: #f8f8f8; }
code > span.kw { color: #204a87; font-weight: bold; } /* Keyword */
code > span.dt { color: #204a87; } /* DataType */
code > span.dv { color: #0000cf; } /* DecVal */
code > span.bn { color: #0000cf; } /* BaseN */
code > span.fl { color: #0000cf; } /* Float */
code > span.ch { color: #4e9a06; } /* Char */
code > span.st { color: #4e9a06; } /* String */
code > span.co { color: #8f5902; font-style: italic; } /* Comment */
code > span.ot { color: #8f5902; } /* Other */
code > span.al { color: #ef2929; } /* Alert */
code > span.fu { color: #000000; } /* Function */
code > span.er { color: #a40000; font-weight: bold; } /* Error */
code > span.wa { color: #8f5902; font-weight: bold; font-style: italic; } /* Warning */
code > span.cn { color: #000000; } /* Constant */
code > span.sc { color: #000000; } /* SpecialChar */
code > span.vs { color: #4e9a06; } /* VerbatimString */
code > span.ss { color: #4e9a06; } /* SpecialString */
code > span.im { } /* Import */
code > span.va { color: #000000; } /* Variable */
code > span.cf { color: #204a87; font-weight: bold; } /* ControlFlow */
code > span.op { color: #ce5c00; font-weight: bold; } /* Operator */
code > span.pp { color: #8f5902; font-style: italic; } /* Preprocessor */
code > span.ex { } /* Extension */
code > span.at { color: #c4a000; } /* Attribute */
code > span.do { color: #8f5902; font-weight: bold; font-style: italic; } /* Documentation */
code > span.an { color: #8f5902; font-weight: bold; font-style: italic; } /* Annotation */
code > span.cv { color: #8f5902; font-weight: bold; font-style: italic; } /* CommentVar */
code > span.in { color: #8f5902; font-weight: bold; font-style: italic; } /* Information */
</style>

<link rel="stylesheet" href="style.css" type="text/css" />
</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li><a href="./">The SingleR book</a></li>

<li class="divider"></li>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html"><i class="fa fa-check"></i>Preface</a></li>
<li class="part"><span><b>I Basic usage</b></span></li>
<li class="chapter" data-level="1" data-path="introduction.html"><a href="introduction.html"><i class="fa fa-check"></i><b>1</b> Introduction</a><ul>
<li class="chapter" data-level="1.1" data-path="introduction.html"><a href="introduction.html#motivation"><i class="fa fa-check"></i><b>1.1</b> Motivation</a></li>
<li class="chapter" data-level="1.2" data-path="introduction.html"><a href="introduction.html#method-description"><i class="fa fa-check"></i><b>1.2</b> Method description</a></li>
<li class="chapter" data-level="1.3" data-path="introduction.html"><a href="introduction.html#quick-start"><i class="fa fa-check"></i><b>1.3</b> Quick start</a></li>
<li class="chapter" data-level="1.4" data-path="introduction.html"><a href="introduction.html#where-to-get-help"><i class="fa fa-check"></i><b>1.4</b> Where to get help</a></li>
<li class="chapter" data-level="" data-path="introduction.html"><a href="introduction.html#session-information"><i class="fa fa-check"></i>Session information</a></li>
</ul></li>
<li class="chapter" data-level="2" data-path="using-the-built-in-references.html"><a href="using-the-built-in-references.html"><i class="fa fa-check"></i><b>2</b> Using the built-in references</a><ul>
<li class="chapter" data-level="2.1" data-path="using-the-built-in-references.html"><a href="using-the-built-in-references.html#overview"><i class="fa fa-check"></i><b>2.1</b> Overview</a></li>
<li class="chapter" data-level="2.2" data-path="using-the-built-in-references.html"><a href="using-the-built-in-references.html#annotation-with-default-marker-detection"><i class="fa fa-check"></i><b>2.2</b> Annotation with default marker detection</a></li>
<li class="chapter" data-level="2.3" data-path="using-the-built-in-references.html"><a href="using-the-built-in-references.html#interaction-with-quality-control"><i class="fa fa-check"></i><b>2.3</b> Interaction with quality control</a></li>
<li class="chapter" data-level="2.4" data-path="using-the-built-in-references.html"><a href="using-the-built-in-references.html#choices-of-assay-data"><i class="fa fa-check"></i><b>2.4</b> Choices of assay data</a></li>
<li class="chapter" data-level="" data-path="using-the-built-in-references.html"><a href="using-the-built-in-references.html#session-information-1"><i class="fa fa-check"></i>Session information</a></li>
</ul></li>
<li class="chapter" data-level="3" data-path="using-single-cell-references.html"><a href="using-single-cell-references.html"><i class="fa fa-check"></i><b>3</b> Using single-cell references</a><ul>
<li class="chapter" data-level="3.1" data-path="using-single-cell-references.html"><a href="using-single-cell-references.html#overview-1"><i class="fa fa-check"></i><b>3.1</b> Overview</a></li>
<li class="chapter" data-level="3.2" data-path="using-single-cell-references.html"><a href="using-single-cell-references.html#annotation-with-test-based-marker-detection"><i class="fa fa-check"></i><b>3.2</b> Annotation with test-based marker detection</a></li>
<li class="chapter" data-level="3.3" data-path="using-single-cell-references.html"><a href="using-single-cell-references.html#defining-custom-markers"><i class="fa fa-check"></i><b>3.3</b> Defining custom markers</a></li>
<li class="chapter" data-level="3.4" data-path="using-single-cell-references.html"><a href="using-single-cell-references.html#pseudo-bulk-aggregation"><i class="fa fa-check"></i><b>3.4</b> Pseudo-bulk aggregation</a></li>
<li class="chapter" data-level="" data-path="using-single-cell-references.html"><a href="using-single-cell-references.html#session-information-2"><i class="fa fa-check"></i>Session information</a></li>
</ul></li>
<li class="chapter" data-level="4" data-path="annotation-diagnostics.html"><a href="annotation-diagnostics.html"><i class="fa fa-check"></i><b>4</b> Annotation diagnostics</a><ul>
<li class="chapter" data-level="4.1" data-path="annotation-diagnostics.html"><a href="annotation-diagnostics.html#overview-2"><i class="fa fa-check"></i><b>4.1</b> Overview</a></li>
<li class="chapter" data-level="4.2" data-path="annotation-diagnostics.html"><a href="annotation-diagnostics.html#based-on-the-scores-within-cells"><i class="fa fa-check"></i><b>4.2</b> Based on the scores within cells</a></li>
<li class="chapter" data-level="4.3" data-path="annotation-diagnostics.html"><a href="annotation-diagnostics.html#based-on-the-deltas-across-cells"><i class="fa fa-check"></i><b>4.3</b> Based on the deltas across cells</a></li>
<li class="chapter" data-level="4.4" data-path="annotation-diagnostics.html"><a href="annotation-diagnostics.html#based-on-marker-gene-expression"><i class="fa fa-check"></i><b>4.4</b> Based on marker gene expression</a></li>
<li class="chapter" data-level="" data-path="annotation-diagnostics.html"><a href="annotation-diagnostics.html#session-information-3"><i class="fa fa-check"></i>Session information</a></li>
</ul></li>
<li class="part"><span><b>II Advanced usage</b></span></li>
<li class="chapter" data-level="5" data-path="using-multiple-references.html"><a href="using-multiple-references.html"><i class="fa fa-check"></i><b>5</b> Using multiple references</a><ul>
<li class="chapter" data-level="5.1" data-path="using-multiple-references.html"><a href="using-multiple-references.html#overview-3"><i class="fa fa-check"></i><b>5.1</b> Overview</a></li>
<li class="chapter" data-level="5.2" data-path="using-multiple-references.html"><a href="using-multiple-references.html#using-reference-specific-labels"><i class="fa fa-check"></i><b>5.2</b> Using reference-specific labels</a></li>
<li class="chapter" data-level="5.3" data-path="using-multiple-references.html"><a href="using-multiple-references.html#comparing-scores-across-references"><i class="fa fa-check"></i><b>5.3</b> Comparing scores across references</a><ul>
<li class="chapter" data-level="5.3.1" data-path="using-multiple-references.html"><a href="using-multiple-references.html#combining-inferences-from-individual-references"><i class="fa fa-check"></i><b>5.3.1</b> Combining inferences from individual references</a></li>
<li class="chapter" data-level="5.3.2" data-path="using-multiple-references.html"><a href="using-multiple-references.html#combined-diagnostics"><i class="fa fa-check"></i><b>5.3.2</b> Combined diagnostics</a></li>
</ul></li>
<li class="chapter" data-level="5.4" data-path="using-multiple-references.html"><a href="using-multiple-references.html#using-harmonized-labels"><i class="fa fa-check"></i><b>5.4</b> Using harmonized labels</a><ul>
<li class="chapter" data-level="5.4.1" data-path="using-multiple-references.html"><a href="using-multiple-references.html#sharing-information-during-marker-detection"><i class="fa fa-check"></i><b>5.4.1</b> Sharing information during marker detection</a></li>
<li class="chapter" data-level="5.4.2" data-path="using-multiple-references.html"><a href="using-multiple-references.html#manual-label-harmonization"><i class="fa fa-check"></i><b>5.4.2</b> Manual label harmonization</a></li>
</ul></li>
<li class="chapter" data-level="" data-path="using-multiple-references.html"><a href="using-multiple-references.html#session-info"><i class="fa fa-check"></i>Session info</a></li>
</ul></li>
<li class="chapter" data-level="6" data-path="exploiting-the-cell-ontology.html"><a href="exploiting-the-cell-ontology.html"><i class="fa fa-check"></i><b>6</b> Exploiting the cell ontology</a><ul>
<li class="chapter" data-level="6.1" data-path="exploiting-the-cell-ontology.html"><a href="exploiting-the-cell-ontology.html#motivation-1"><i class="fa fa-check"></i><b>6.1</b> Motivation</a></li>
<li class="chapter" data-level="6.2" data-path="exploiting-the-cell-ontology.html"><a href="exploiting-the-cell-ontology.html#basic-manipulation"><i class="fa fa-check"></i><b>6.2</b> Basic manipulation</a></li>
<li class="chapter" data-level="6.3" data-path="exploiting-the-cell-ontology.html"><a href="exploiting-the-cell-ontology.html#adjusting-resolution"><i class="fa fa-check"></i><b>6.3</b> Adjusting resolution</a></li>
<li class="chapter" data-level="" data-path="exploiting-the-cell-ontology.html"><a href="exploiting-the-cell-ontology.html#session-information-4"><i class="fa fa-check"></i>Session information</a></li>
</ul></li>
<li class="chapter" data-level="7" data-path="advanced-options.html"><a href="advanced-options.html"><i class="fa fa-check"></i><b>7</b> Advanced options</a><ul>
<li class="chapter" data-level="7.1" data-path="advanced-options.html"><a href="advanced-options.html#preconstructed-indices"><i class="fa fa-check"></i><b>7.1</b> Preconstructed indices</a></li>
<li class="chapter" data-level="7.2" data-path="advanced-options.html"><a href="advanced-options.html#parallelization"><i class="fa fa-check"></i><b>7.2</b> Parallelization</a></li>
<li class="chapter" data-level="7.3" data-path="advanced-options.html"><a href="advanced-options.html#approximate-algorithms"><i class="fa fa-check"></i><b>7.3</b> Approximate algorithms</a></li>
<li class="chapter" data-level="" data-path="advanced-options.html"><a href="advanced-options.html#session-information-5"><i class="fa fa-check"></i>Session information</a></li>
</ul></li>
<li class="part"><span><b>III Workflows</b></span></li>
<li class="chapter" data-level="8" data-path="cross-annotating-pancreas.html"><a href="cross-annotating-pancreas.html"><i class="fa fa-check"></i><b>8</b> Cross-annotating pancreas</a><ul>
<li class="chapter" data-level="8.1" data-path="cross-annotating-pancreas.html"><a href="cross-annotating-pancreas.html#loading-the-data"><i class="fa fa-check"></i><b>8.1</b> Loading the data</a></li>
<li class="chapter" data-level="8.2" data-path="cross-annotating-pancreas.html"><a href="cross-annotating-pancreas.html#applying-the-annotation"><i class="fa fa-check"></i><b>8.2</b> Applying the annotation</a></li>
<li class="chapter" data-level="8.3" data-path="cross-annotating-pancreas.html"><a href="cross-annotating-pancreas.html#diagnostics"><i class="fa fa-check"></i><b>8.3</b> Diagnostics</a></li>
<li class="chapter" data-level="" data-path="cross-annotating-pancreas.html"><a href="cross-annotating-pancreas.html#session-information-6"><i class="fa fa-check"></i>Session information</a></li>
</ul></li>
<li class="part"><span><b>IV Appendix</b></span></li>
<li class="chapter" data-level="9" data-path="contributors.html"><a href="contributors.html"><i class="fa fa-check"></i><b>9</b> Contributors</a><ul>
<li><a href="contributors.html#aaron-lun"><em>Aaron Lun</em></a></li>
</ul></li>
<li class="chapter" data-level="10" data-path="bibliography.html"><a href="bibliography.html"><i class="fa fa-check"></i><b>10</b> Bibliography</a></li>
<li class="divider"></li>
<li><a href="https://bioconductor.org" target="blank">Published by Bioconductor</a></li>

</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">Assigning cell types with SingleR</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="using-the-built-in-references" class="section level1">
<h1><span class="header-section-number">Chapter 2</span> Using the built-in references</h1>
<script>
document.addEventListener("click", function (event) {
    if (event.target.classList.contains("aaron-collapse")) {
        event.target.classList.toggle("active");
        var content = event.target.nextElementSibling;
        if (content.style.display === "block") {
            content.style.display = "none";
        } else {
            content.style.display = "block";
        }
    }
})
</script>
<style>
.aaron-collapse {
  background-color: #eee;
  color: #444;
  cursor: pointer;
  padding: 18px;
  width: 100%;
  border: none;
  text-align: left;
  outline: none;
  font-size: 15px;
}

.aaron-content {
  padding: 0 18px;
  display: none;
  overflow: hidden;
  background-color: #f1f1f1;
}
</style>
<div id="overview" class="section level2">
<h2><span class="header-section-number">2.1</span> Overview</h2>
<p><em><a href="https://bioconductor.org/packages/3.12/SingleR">SingleR</a></em> detects markers in a pairwise manner between labels in the reference dataset. Specifically, for each label of interest, it performs pairwise comparisons to every other label in the reference and identifies the genes that are upregulated in the label of interest for each comparison. The initial score calculation is then performed on the union of marker genes across all comparisons for all label. This approach ensures that the selected subset of features will contain genes that distinguish each label from any other label. (In contrast, other approaches that treat the “other” labels as a single group do not offer this guarantee; see <a href="https://osca.bioconductor.org/marker-detection.html#standard-application">here</a> for a discussion.) It also allows the fine-tuning step to aggressively improve resolution by only using marker genes from comparisons where both labels have scores close to the maximum.</p>
<p>The original (“classic”) marker detection algorithm used in <span class="citation">Aran et al. (<a href="#ref-aran2019reference">2019</a>)</span> identified marker genes based on their log-fold changes in each pairwise comparison. Specifically, it used the genes with the largest positive differences in the per-label median log-expression values between labels. The number of genes taken from each pairwise comparison was defined as <span class="math inline">\(500 (\frac{2}{3})^{\log_{2}(n)}\)</span>, where <span class="math inline">\(n\)</span> is the number of unique labels in the reference; this aimed to reduce the number of genes (and thus the computational time) as the number of labels and pairwise comparisons increased. This classic mode is primarily intended for reference datasets that have little or no replication, a description that covers most of the built-in references and precludes more complicated marker detection procedures (Chapter <a href="using-single-cell-references.html#using-single-cell-references">3</a>).</p>
</div>
<div id="annotation-with-default-marker-detection" class="section level2">
<h2><span class="header-section-number">2.2</span> Annotation with default marker detection</h2>
<p>For demonstration purposes, we will use the <span class="citation">Grun et al. (<a href="#ref-grun2016denovo">2016</a>)</span> haematopoietic stem cell (HSC) dataset from the <em><a href="https://bioconductor.org/packages/3.12/scRNAseq">scRNAseq</a></em> package. The <code>GrunHSCData()</code> function conveniently returns a <code>SingleCellExperiment</code> object containing the count matrix for this dataset.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(scRNAseq)
sce &lt;-<span class="st"> </span><span class="kw">GrunHSCData</span>(<span class="dt">ensembl=</span><span class="ot">TRUE</span>)
sce</code></pre></div>
<pre><code>## class: SingleCellExperiment 
## dim: 21817 1915 
## metadata(0):
## assays(1): counts
## rownames(21817): ENSMUSG00000109644 ENSMUSG00000007777 ...
##   ENSMUSG00000055670 ENSMUSG00000039068
## rowData names(3): symbol chr originalName
## colnames(1915): JC4_349_HSC_FE_S13_ JC4_350_HSC_FE_S13_ ...
##   JC48P6_1203_HSC_FE_S8_ JC48P6_1204_HSC_FE_S8_
## colData names(2): sample protocol
## reducedDimNames(0):
## altExpNames(0):</code></pre>
<p>Our plan is to annotate each cell with the built-in ImmGen reference dataset <span class="citation">(Heng et al. <a href="#ref-ImmGenRef">2008</a>)</span>. Calling the <code>ImmGenData()</code> function returns a <code>SummarizedExperiment</code> object containing a matrix of log-expression values with sample-level labels. We set <code>ensembl=TRUE</code> to match the reference’s gene annotation with that in the <code>sce</code> object (the default behavior is to use the gene symbol).</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(SingleR)
immgen &lt;-<span class="st"> </span><span class="kw">ImmGenData</span>(<span class="dt">ensembl=</span><span class="ot">TRUE</span>)
immgen</code></pre></div>
<pre><code>## class: SummarizedExperiment 
## dim: 21352 830 
## metadata(0):
## assays(1): logcounts
## rownames(21352): ENSMUSG00000079681 ENSMUSG00000066372 ...
##   ENSMUSG00000034640 ENSMUSG00000036940
## rowData names(0):
## colnames(830):
##   GSM1136119_EA07068_260297_MOGENE-1_0-ST-V1_MF.11C-11B+.LU_1.CEL
##   GSM1136120_EA07068_260298_MOGENE-1_0-ST-V1_MF.11C-11B+.LU_2.CEL ...
##   GSM920654_EA07068_201214_MOGENE-1_0-ST-V1_TGD.VG4+24ALO.E17.TH_1.CEL
##   GSM920655_EA07068_201215_MOGENE-1_0-ST-V1_TGD.VG4+24ALO.E17.TH_2.CEL
## colData names(3): label.main label.fine label.ont</code></pre>
<p>Technically speaking, each built-in dataset actually has three sets of labels that primarily differ in their resolution. For the purposes of this demonstration, we will use the “fine” labels in the <code>label.fine</code> metadata field, which represents the highest resolution of annotation available in ImmGen.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">head</span>(immgen<span class="op">$</span>label.fine)</code></pre></div>
<pre><code>## [1] &quot;Macrophages (MF.11C-11B+)&quot; &quot;Macrophages (MF.11C-11B+)&quot;
## [3] &quot;Macrophages (MF.11C-11B+)&quot; &quot;Macrophages (MF.ALV)&quot;     
## [5] &quot;Macrophages (MF.ALV)&quot;      &quot;Macrophages (MF.ALV)&quot;</code></pre>
<p>Annotation is then a simple matter of calling <code>SingleR()</code> on our test (Grun) dataset and the reference (ImmGen) dataset, leaving the default of <code>de.method=&quot;classic&quot;</code> to use the original marker detection scheme. This applies the algorithm described in Section <a href="introduction.html#method-description">1.2</a>, returning a <code>DataFrame</code> where each row contains prediction results for a single cell in the <code>sce</code> object. Labels are provided before fine-tuning (<code>first.labels</code>), after fine-tuning (<code>labels</code>) and after pruning (<code>pruned.labels</code>); some of the other fields are discussed in more detail in Chapter <a href="annotation-diagnostics.html#annotation-diagnostics">4</a>.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># See &#39;Choices of assay data&#39; for &#39;assay.type.test=&#39; explanation.</span>
pred &lt;-<span class="st"> </span><span class="kw">SingleR</span>(<span class="dt">test =</span> sce, <span class="dt">ref =</span> immgen, 
    <span class="dt">labels =</span> immgen<span class="op">$</span>label.fine, <span class="dt">assay.type.test=</span><span class="dv">1</span>)
<span class="kw">colnames</span>(pred)</code></pre></div>
<pre><code>## [1] &quot;scores&quot;        &quot;first.labels&quot;  &quot;tuning.scores&quot; &quot;labels&quot;       
## [5] &quot;pruned.labels&quot;</code></pre>
</div>
<div id="interaction-with-quality-control" class="section level2">
<h2><span class="header-section-number">2.3</span> Interaction with quality control</h2>
<p>Upon examining the distribution of assigned labels, we see that many of them are related to stem cells. However, there are quite a large number of more differentiated labels mixed in, which is not what we expect from a sorted population of HSCs.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">head</span>(<span class="kw">sort</span>(<span class="kw">table</span>(pred<span class="op">$</span>labels), <span class="dt">decreasing=</span><span class="ot">TRUE</span>))</code></pre></div>
<pre><code>## 
##   Stem cells (SC.MEP) Neutrophils (GN.ARTH)      Macrophages (MF) 
##                   362                   314                   165 
##  Stem cells (SC.STSL)    B cells (proB.FrA) Stem cells (SC.LT34F) 
##                   142                   121                   103</code></pre>
<p>This is probably because - despite what its name might suggest - the dataset obtained by <code>GrunHSCData()</code> actually contains more than HSCs. If we restrict our analysis to the sorted HSCs (obviously) and remove one low-quality batch (see <a href="https://osca.bioconductor.org/merged-hcsc.html#quality-control-12">the analysis here</a> for the rationale) we can see that the distribution of cell type labels is more similar to what we might expect. Low-quality cells - typically those with low library sizes and low numbers of detected genes - lack information for accurate label assignment and often need to be removed to enable interpretation of the results.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">actual.hsc &lt;-<span class="st"> </span>pred<span class="op">$</span>labels[sce<span class="op">$</span>protocol<span class="op">==</span><span class="st">&quot;sorted hematopoietic stem cells&quot;</span> <span class="op">&amp;</span><span class="st"> </span>sce<span class="op">$</span>sample<span class="op">!=</span><span class="st">&quot;JC4&quot;</span>]
<span class="kw">head</span>(<span class="kw">sort</span>(<span class="kw">table</span>(actual.hsc), <span class="dt">decreasing=</span><span class="ot">TRUE</span>))</code></pre></div>
<pre><code>## actual.hsc
##        Stem cells (SC.STSL)       Stem cells (SC.LT34F) 
##                         109                          98 
##       Stem cells (SC.ST34F) Stem cells (SC.CD150-CD48-) 
##                          37                          15 
##          Stem cells (LTHSC)            Stem cells (MLP) 
##                          12                           8</code></pre>
<p>Filtering the annotation results in the above manner is possible because <code>SingleR()</code> operates independently on each cell. The annotation is orthogonal to any decisions about the relative quality of the cells in the test dataset; the same results will be obtained regardless of whether the annotation is performed before or after quality control. This is logisitically convenient as it means that the annotation does not have to be repeated if the quality control scheme (or any other downstream step, like clustering) changes throughout the lifetime of the analysis.</p>
</div>
<div id="choices-of-assay-data" class="section level2">
<h2><span class="header-section-number">2.4</span> Choices of assay data</h2>
<p>For the reference dataset, the assay matrix <em>must</em> contain log-transformed normalized expression values. This is because the default marker detection scheme computes log-fold changes by subtracting the medians, which makes little sense unless the input expression values are already log-transformed. For alternative schemes, this requirement may be relaxed (e.g., Wilcoxon rank sum tests do not require transformation); similarly, if pre-defined markers are supplied, no transformation or normalization is necessary.</p>
<p>For the test data, the assay data need not be log-transformed or even (scale) normalized. This is because <code>SingleR()</code> computes Spearman correlations within each cell, which is unaffected by monotonic transformations like cell-specific scaling or log-transformation. It is perfectly satisfactory to provide the raw counts for the test dataset to <code>SingleR()</code>, which is the reason for setting <code>assay.type.test=1</code> in our previous <code>SingleR()</code> call for the Grun dataset.</p>
<p>The exception to this rule occurs when comparing data from full-length technologies to the built-in references. The built-in references are constructed to be comparable to unique molecular identifier (UMI) protocols, where the expression values are less sensitive to differences in gene length. Thus, when comparing Smart-seq2 test datasets to the built-in references, better performance can often be achieved by processing the test counts to transcripts-per-million values.</p>
<p>We demonstrate below using another HSC dataset that was generated using the Smart-seq2 protocol <span class="citation">(Nestorowa et al. <a href="#ref-nestorowa2016singlecell">2016</a>)</span>. Again, we see that most of the predicted labels are related to stem cells, which is comforting.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">sce.nest &lt;-<span class="st"> </span><span class="kw">NestorowaHSCData</span>()

<span class="co"># Getting the exonic gene lengths.</span>
<span class="kw">library</span>(AnnotationHub)
mm.db &lt;-<span class="st"> </span><span class="kw">AnnotationHub</span>()[[<span class="st">&quot;AH73905&quot;</span>]]
mm.exons &lt;-<span class="st"> </span><span class="kw">exonsBy</span>(mm.db, <span class="dt">by=</span><span class="st">&quot;gene&quot;</span>)
mm.exons &lt;-<span class="st"> </span><span class="kw">reduce</span>(mm.exons)
mm.len &lt;-<span class="st"> </span><span class="kw">sum</span>(<span class="kw">width</span>(mm.exons))

<span class="co"># Computing the TPMs with a simple scaling by gene length.</span>
<span class="kw">library</span>(scater)
keep &lt;-<span class="st"> </span><span class="kw">intersect</span>(<span class="kw">names</span>(mm.len), <span class="kw">rownames</span>(sce.nest))
tpm.nest &lt;-<span class="st"> </span><span class="kw">calculateTPM</span>(sce.nest[keep,], <span class="dt">lengths=</span>mm.len[keep])

<span class="co"># Performing the assignment.</span>
pred &lt;-<span class="st"> </span><span class="kw">SingleR</span>(<span class="dt">test =</span> tpm.nest, <span class="dt">ref =</span> immgen, <span class="dt">labels =</span> immgen<span class="op">$</span>label.fine)
<span class="kw">head</span>(<span class="kw">sort</span>(<span class="kw">table</span>(pred<span class="op">$</span>labels), <span class="dt">decreasing=</span><span class="ot">TRUE</span>), <span class="dv">10</span>)</code></pre></div>
<pre><code>## 
##         Stem cells (SC.MEP)       Stem cells (SC.ST34F) 
##                         409                         357 
##      Stem cells (SC.MPP34F)      Stem cells (SC.CMP.DR) 
##                         329                         298 
##            Stem cells (MLP)            Stem cells (GMP) 
##                         167                         102 
##        Stem cells (SC.STSL)         Stem cells (SC.MDP) 
##                          71                          66 
## Stem cells (SC.CD150-CD48-)       Stem cells (SC.LT34F) 
##                          55                          37</code></pre>
</div>
<div id="session-information-1" class="section level2 unnumbered">
<h2>Session information</h2>
<button class="aaron-collapse">
View session info
</button>
<div class="aaron-content">
<pre><code>R version 4.0.0 Patched (2020-05-01 r78341)
Platform: x86_64-pc-linux-gnu (64-bit)
Running under: Ubuntu 18.04.4 LTS

Matrix products: default
BLAS:   /home/luna/Software/R/R-4-0-branch-dev/lib/libRblas.so
LAPACK: /home/luna/Software/R/R-4-0-branch-dev/lib/libRlapack.so

locale:
 [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              
 [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    
 [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   
 [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 
 [9] LC_ADDRESS=C               LC_TELEPHONE=C            
[11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       

attached base packages:
[1] stats4    parallel  stats     graphics  grDevices utils     datasets 
[8] methods   base     

other attached packages:
 [1] scater_1.17.0               ggplot2_3.3.0              
 [3] AnnotationHub_2.21.0        BiocFileCache_1.13.0       
 [5] dbplyr_1.4.3                SingleR_1.3.4              
 [7] ensembldb_2.13.1            AnnotationFilter_1.13.0    
 [9] GenomicFeatures_1.41.0      AnnotationDbi_1.51.0       
[11] scRNAseq_2.3.0              SingleCellExperiment_1.11.1
[13] SummarizedExperiment_1.19.4 DelayedArray_0.15.1        
[15] matrixStats_0.56.0          Biobase_2.49.0             
[17] GenomicRanges_1.41.1        GenomeInfoDb_1.25.0        
[19] IRanges_2.23.4              S4Vectors_0.27.6           
[21] BiocGenerics_0.35.2         BiocStyle_2.17.0           
[23] rebook_0.99.0              

loaded via a namespace (and not attached):
 [1] ProtGenerics_1.21.0           bitops_1.0-6                 
 [3] bit64_0.9-7                   progress_1.2.2               
 [5] httr_1.4.1                    tools_4.0.0                  
 [7] irlba_2.3.3                   R6_2.4.1                     
 [9] vipor_0.4.5                   colorspace_1.4-1             
[11] DBI_1.1.0                     lazyeval_0.2.2               
[13] withr_2.2.0                   gridExtra_2.3                
[15] tidyselect_1.1.0              prettyunits_1.1.1            
[17] processx_3.4.2                bit_1.1-15.2                 
[19] curl_4.3                      compiler_4.0.0               
[21] graph_1.67.0                  BiocNeighbors_1.7.0          
[23] rtracklayer_1.49.1            bookdown_0.19                
[25] scales_1.1.1                  callr_3.4.3                  
[27] askpass_1.1                   rappdirs_0.3.1               
[29] stringr_1.4.0                 digest_0.6.25                
[31] Rsamtools_2.5.0               rmarkdown_2.1                
[33] XVector_0.29.0                pkgconfig_2.0.3              
[35] htmltools_0.4.0               fastmap_1.0.1                
[37] rlang_0.4.6                   RSQLite_2.2.0                
[39] DelayedMatrixStats_1.11.0     shiny_1.4.0.2                
[41] BiocParallel_1.23.0           dplyr_0.8.5                  
[43] BiocSingular_1.5.0            RCurl_1.98-1.2               
[45] magrittr_1.5                  GenomeInfoDbData_1.2.3       
[47] Matrix_1.2-18                 ggbeeswarm_0.6.0             
[49] munsell_0.5.0                 Rcpp_1.0.4.6                 
[51] viridis_0.5.1                 lifecycle_0.2.0              
[53] stringi_1.4.6                 yaml_2.2.1                   
[55] zlibbioc_1.35.0               grid_4.0.0                   
[57] blob_1.2.1                    promises_1.1.0               
[59] ExperimentHub_1.15.0          crayon_1.3.4                 
[61] lattice_0.20-41               Biostrings_2.57.0            
[63] hms_0.5.3                     CodeDepends_0.6.5            
[65] knitr_1.28                    ps_1.3.3                     
[67] pillar_1.4.4                  codetools_0.2-16             
[69] biomaRt_2.45.0                XML_3.99-0.3                 
[71] glue_1.4.1                    BiocVersion_3.12.0           
[73] evaluate_0.14                 BiocManager_1.30.10          
[75] vctrs_0.3.0                   httpuv_1.5.2                 
[77] gtable_0.3.0                  openssl_1.4.1                
[79] purrr_0.3.4                   assertthat_0.2.1             
[81] xfun_0.13                     rsvd_1.0.3                   
[83] mime_0.9                      xtable_1.8-4                 
[85] later_1.0.0                   viridisLite_0.3.0            
[87] tibble_3.0.1                  beeswarm_0.2.3               
[89] GenomicAlignments_1.25.0      memoise_1.1.0                
[91] ellipsis_0.3.1                interactiveDisplayBase_1.27.0</code></pre>
</div>

</div>
</div>
<h3> Bibliography</h3>
<div id="refs" class="references">
<div id="ref-aran2019reference">
<p>Aran, D., A. P. Looney, L. Liu, E. Wu, V. Fong, A. Hsu, S. Chak, et al. 2019. “Reference-based analysis of lung single-cell sequencing reveals a transitional profibrotic macrophage.” <em>Nat. Immunol.</em> 20 (2): 163–72.</p>
</div>
<div id="ref-grun2016denovo">
<p>Grun, D., M. J. Muraro, J. C. Boisset, K. Wiebrands, A. Lyubimova, G. Dharmadhikari, M. van den Born, et al. 2016. “De Novo Prediction of Stem Cell Identity using Single-Cell Transcriptome Data.” <em>Cell Stem Cell</em> 19 (2): 266–77.</p>
</div>
<div id="ref-ImmGenRef">
<p>Heng, Tracy S.P., Michio W. Painter, Kutlu Elpek, Veronika Lukacs-Kornek, Nora Mauermann, Shannon J. Turley, Daphne Koller, et al. 2008. “The immunological genome project: Networks of gene expression in immune cells.” <em>Nature Immunology</em> 9 (10): 1091–4. doi:<a href="https://doi.org/10.1038/ni1008-1091">10.1038/ni1008-1091</a>.</p>
</div>
<div id="ref-nestorowa2016singlecell">
<p>Nestorowa, S., F. K. Hamey, B. Pijuan Sala, E. Diamanti, M. Shepherd, E. Laurenti, N. K. Wilson, D. G. Kent, and B. Gottgens. 2016. “A single-cell resolution map of mouse hematopoietic stem and progenitor cell differentiation.” <em>Blood</em> 128 (8): 20–31.</p>
</div>
</div>
            </section>

          </div>
        </div>
      </div>
<a href="introduction.html" class="navigation navigation-prev " aria-label="Previous page"><i class="fa fa-angle-left"></i></a>
<a href="using-single-cell-references.html" class="navigation navigation-next " aria-label="Next page"><i class="fa fa-angle-right"></i></a>
    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/lunr.js"></script>
<script src="libs/gitbook-2.6.7/js/clipboard.min.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-clipboard.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": {
"github": false,
"facebook": false,
"twitter": false,
"linkedin": false,
"weibo": false,
"instapaper": false,
"vk": false,
"all": ["facebook", "twitter", "linkedin", "weibo", "instapaper"]
},
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": {
"link": null,
"text": null
},
"history": {
"link": null,
"text": null
},
"view": {
"link": null,
"text": null
},
"download": null,
"toc": {
"collapse": "subsection"
},
"bookdown::epub_book": "default",
"cover": "images/cover.png"
});
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    var src = "";
    if (src === "" || src === "true") src = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-MML-AM_CHTML";
    if (location.protocol !== "file:")
      if (/^https?:/.test(src))
        src = src.replace(/^https?:/, '');
    script.src = src;
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>
</body>

</html>
