<!DOCTYPE html>
<html lang="" xml:lang="">
<head>

  <meta charset="utf-8" />
  <meta http-equiv="X-UA-Compatible" content="IE=edge" />
  <title>Chapter 5 Using multiple references | Assigning cell types with SingleR</title>
  <meta name="description" content="The SingleR book. Because sometimes, a vignette just isn’t enough." />
  <meta name="generator" content="bookdown 0.19 and GitBook 2.6.7" />

  <meta property="og:title" content="Chapter 5 Using multiple references | Assigning cell types with SingleR" />
  <meta property="og:type" content="book" />
  
  
  <meta property="og:description" content="The SingleR book. Because sometimes, a vignette just isn’t enough." />
  

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="Chapter 5 Using multiple references | Assigning cell types with SingleR" />
  
  <meta name="twitter:description" content="The SingleR book. Because sometimes, a vignette just isn’t enough." />
  



<meta name="date" content="2020-05-20" />

  <meta name="viewport" content="width=device-width, initial-scale=1" />
  <meta name="apple-mobile-web-app-capable" content="yes" />
  <meta name="apple-mobile-web-app-status-bar-style" content="black" />
  
  
<link rel="prev" href="annotation-diagnostics.html"/>
<link rel="next" href="harmonizing-labels.html"/>
<script src="libs/jquery-2.2.3/jquery.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-table.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-clipboard.css" rel="stylesheet" />











<style type="text/css">
div.sourceCode { overflow-x: auto; }
table.sourceCode, tr.sourceCode, td.lineNumbers, td.sourceCode {
  margin: 0; padding: 0; vertical-align: baseline; border: none; }
table.sourceCode { width: 100%; line-height: 100%; background-color: #f8f8f8; }
td.lineNumbers { text-align: right; padding-right: 4px; padding-left: 4px; color: #aaaaaa; border-right: 1px solid #aaaaaa; }
td.sourceCode { padding-left: 5px; }
pre, code { background-color: #f8f8f8; }
code > span.kw { color: #204a87; font-weight: bold; } /* Keyword */
code > span.dt { color: #204a87; } /* DataType */
code > span.dv { color: #0000cf; } /* DecVal */
code > span.bn { color: #0000cf; } /* BaseN */
code > span.fl { color: #0000cf; } /* Float */
code > span.ch { color: #4e9a06; } /* Char */
code > span.st { color: #4e9a06; } /* String */
code > span.co { color: #8f5902; font-style: italic; } /* Comment */
code > span.ot { color: #8f5902; } /* Other */
code > span.al { color: #ef2929; } /* Alert */
code > span.fu { color: #000000; } /* Function */
code > span.er { color: #a40000; font-weight: bold; } /* Error */
code > span.wa { color: #8f5902; font-weight: bold; font-style: italic; } /* Warning */
code > span.cn { color: #000000; } /* Constant */
code > span.sc { color: #000000; } /* SpecialChar */
code > span.vs { color: #4e9a06; } /* VerbatimString */
code > span.ss { color: #4e9a06; } /* SpecialString */
code > span.im { } /* Import */
code > span.va { color: #000000; } /* Variable */
code > span.cf { color: #204a87; font-weight: bold; } /* ControlFlow */
code > span.op { color: #ce5c00; font-weight: bold; } /* Operator */
code > span.pp { color: #8f5902; font-style: italic; } /* Preprocessor */
code > span.ex { } /* Extension */
code > span.at { color: #c4a000; } /* Attribute */
code > span.do { color: #8f5902; font-weight: bold; font-style: italic; } /* Documentation */
code > span.an { color: #8f5902; font-weight: bold; font-style: italic; } /* Annotation */
code > span.cv { color: #8f5902; font-weight: bold; font-style: italic; } /* CommentVar */
code > span.in { color: #8f5902; font-weight: bold; font-style: italic; } /* Information */
</style>

<link rel="stylesheet" href="style.css" type="text/css" />
</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li><a href="./">The SingleR book</a></li>

<li class="divider"></li>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html"><i class="fa fa-check"></i>Preface</a></li>
<li class="part"><span><b>I Basic usage</b></span></li>
<li class="chapter" data-level="1" data-path="introduction.html"><a href="introduction.html"><i class="fa fa-check"></i><b>1</b> Introduction</a><ul>
<li class="chapter" data-level="1.1" data-path="introduction.html"><a href="introduction.html#motivation"><i class="fa fa-check"></i><b>1.1</b> Motivation</a></li>
<li class="chapter" data-level="1.2" data-path="introduction.html"><a href="introduction.html#method-description"><i class="fa fa-check"></i><b>1.2</b> Method description</a></li>
<li class="chapter" data-level="1.3" data-path="introduction.html"><a href="introduction.html#quick-start"><i class="fa fa-check"></i><b>1.3</b> Quick start</a></li>
<li class="chapter" data-level="1.4" data-path="introduction.html"><a href="introduction.html#where-to-get-help"><i class="fa fa-check"></i><b>1.4</b> Where to get help</a></li>
<li class="chapter" data-level="" data-path="introduction.html"><a href="introduction.html#session-information"><i class="fa fa-check"></i>Session information</a></li>
</ul></li>
<li class="chapter" data-level="2" data-path="using-the-built-in-references.html"><a href="using-the-built-in-references.html"><i class="fa fa-check"></i><b>2</b> Using the built-in references</a><ul>
<li class="chapter" data-level="2.1" data-path="using-the-built-in-references.html"><a href="using-the-built-in-references.html#overview"><i class="fa fa-check"></i><b>2.1</b> Overview</a></li>
<li class="chapter" data-level="2.2" data-path="using-the-built-in-references.html"><a href="using-the-built-in-references.html#annotation-with-default-marker-detection"><i class="fa fa-check"></i><b>2.2</b> Annotation with default marker detection</a></li>
<li class="chapter" data-level="2.3" data-path="using-the-built-in-references.html"><a href="using-the-built-in-references.html#interaction-with-quality-control"><i class="fa fa-check"></i><b>2.3</b> Interaction with quality control</a></li>
<li class="chapter" data-level="2.4" data-path="using-the-built-in-references.html"><a href="using-the-built-in-references.html#choices-of-assay-data"><i class="fa fa-check"></i><b>2.4</b> Choices of assay data</a></li>
<li class="chapter" data-level="" data-path="using-the-built-in-references.html"><a href="using-the-built-in-references.html#session-information-1"><i class="fa fa-check"></i>Session information</a></li>
</ul></li>
<li class="chapter" data-level="3" data-path="using-single-cell-references.html"><a href="using-single-cell-references.html"><i class="fa fa-check"></i><b>3</b> Using single-cell references</a><ul>
<li class="chapter" data-level="3.1" data-path="using-single-cell-references.html"><a href="using-single-cell-references.html#overview-1"><i class="fa fa-check"></i><b>3.1</b> Overview</a></li>
<li class="chapter" data-level="3.2" data-path="using-single-cell-references.html"><a href="using-single-cell-references.html#annotation-with-test-based-marker-detection"><i class="fa fa-check"></i><b>3.2</b> Annotation with test-based marker detection</a></li>
<li class="chapter" data-level="3.3" data-path="using-single-cell-references.html"><a href="using-single-cell-references.html#defining-custom-markers"><i class="fa fa-check"></i><b>3.3</b> Defining custom markers</a></li>
<li class="chapter" data-level="3.4" data-path="using-single-cell-references.html"><a href="using-single-cell-references.html#pseudo-bulk-aggregation"><i class="fa fa-check"></i><b>3.4</b> Pseudo-bulk aggregation</a></li>
<li class="chapter" data-level="" data-path="using-single-cell-references.html"><a href="using-single-cell-references.html#session-information-2"><i class="fa fa-check"></i>Session information</a></li>
</ul></li>
<li class="chapter" data-level="4" data-path="annotation-diagnostics.html"><a href="annotation-diagnostics.html"><i class="fa fa-check"></i><b>4</b> Annotation diagnostics</a><ul>
<li class="chapter" data-level="4.1" data-path="annotation-diagnostics.html"><a href="annotation-diagnostics.html#overview-2"><i class="fa fa-check"></i><b>4.1</b> Overview</a></li>
<li class="chapter" data-level="4.2" data-path="annotation-diagnostics.html"><a href="annotation-diagnostics.html#based-on-the-scores-within-cells"><i class="fa fa-check"></i><b>4.2</b> Based on the scores within cells</a></li>
<li class="chapter" data-level="4.3" data-path="annotation-diagnostics.html"><a href="annotation-diagnostics.html#based-on-the-deltas-across-cells"><i class="fa fa-check"></i><b>4.3</b> Based on the deltas across cells</a></li>
<li class="chapter" data-level="4.4" data-path="annotation-diagnostics.html"><a href="annotation-diagnostics.html#based-on-marker-gene-expression"><i class="fa fa-check"></i><b>4.4</b> Based on marker gene expression</a></li>
<li class="chapter" data-level="" data-path="annotation-diagnostics.html"><a href="annotation-diagnostics.html#session-information-3"><i class="fa fa-check"></i>Session information</a></li>
</ul></li>
<li class="part"><span><b>II Advanced usage</b></span></li>
<li class="chapter" data-level="5" data-path="using-multiple-references.html"><a href="using-multiple-references.html"><i class="fa fa-check"></i><b>5</b> Using multiple references</a><ul>
<li class="chapter" data-level="5.1" data-path="using-multiple-references.html"><a href="using-multiple-references.html#overview-3"><i class="fa fa-check"></i><b>5.1</b> Overview</a></li>
<li class="chapter" data-level="5.2" data-path="using-multiple-references.html"><a href="using-multiple-references.html#using-reference-specific-labels"><i class="fa fa-check"></i><b>5.2</b> Using reference-specific labels</a></li>
<li class="chapter" data-level="5.3" data-path="using-multiple-references.html"><a href="using-multiple-references.html#using-harmonized-labels"><i class="fa fa-check"></i><b>5.3</b> Using harmonized labels</a></li>
<li class="chapter" data-level="5.4" data-path="using-multiple-references.html"><a href="using-multiple-references.html#comparing-scores-across-references"><i class="fa fa-check"></i><b>5.4</b> Comparing scores across references</a></li>
<li class="chapter" data-level="" data-path="using-multiple-references.html"><a href="using-multiple-references.html#session-info"><i class="fa fa-check"></i>Session info</a></li>
</ul></li>
<li class="chapter" data-level="6" data-path="harmonizing-labels.html"><a href="harmonizing-labels.html"><i class="fa fa-check"></i><b>6</b> Harmonizing labels</a></li>
<li class="chapter" data-level="7" data-path="advanced-options.html"><a href="advanced-options.html"><i class="fa fa-check"></i><b>7</b> Advanced options</a><ul>
<li class="chapter" data-level="7.1" data-path="advanced-options.html"><a href="advanced-options.html#preconstructed-indices"><i class="fa fa-check"></i><b>7.1</b> Preconstructed indices</a></li>
<li class="chapter" data-level="7.2" data-path="advanced-options.html"><a href="advanced-options.html#parallelization"><i class="fa fa-check"></i><b>7.2</b> Parallelization</a></li>
<li class="chapter" data-level="7.3" data-path="advanced-options.html"><a href="advanced-options.html#approximate-algorithms"><i class="fa fa-check"></i><b>7.3</b> Approximate algorithms</a></li>
<li class="chapter" data-level="" data-path="advanced-options.html"><a href="advanced-options.html#session-information-4"><i class="fa fa-check"></i>Session information</a></li>
</ul></li>
<li class="part"><span><b>III Workflows</b></span></li>
<li class="chapter" data-level="8" data-path="cross-annotating-pancreas.html"><a href="cross-annotating-pancreas.html"><i class="fa fa-check"></i><b>8</b> Cross-annotating pancreas</a><ul>
<li class="chapter" data-level="8.1" data-path="cross-annotating-pancreas.html"><a href="cross-annotating-pancreas.html#loading-the-data"><i class="fa fa-check"></i><b>8.1</b> Loading the data</a></li>
<li class="chapter" data-level="8.2" data-path="cross-annotating-pancreas.html"><a href="cross-annotating-pancreas.html#applying-the-annotation"><i class="fa fa-check"></i><b>8.2</b> Applying the annotation</a></li>
<li class="chapter" data-level="8.3" data-path="cross-annotating-pancreas.html"><a href="cross-annotating-pancreas.html#diagnostics"><i class="fa fa-check"></i><b>8.3</b> Diagnostics</a></li>
<li class="chapter" data-level="" data-path="cross-annotating-pancreas.html"><a href="cross-annotating-pancreas.html#session-information-5"><i class="fa fa-check"></i>Session information</a></li>
</ul></li>
<li class="part"><span><b>IV Appendix</b></span></li>
<li class="chapter" data-level="9" data-path="contributors.html"><a href="contributors.html"><i class="fa fa-check"></i><b>9</b> Contributors</a><ul>
<li><a href="contributors.html#aaron-lun"><em>Aaron Lun</em></a></li>
</ul></li>
<li class="chapter" data-level="10" data-path="bibliography.html"><a href="bibliography.html"><i class="fa fa-check"></i><b>10</b> Bibliography</a></li>
<li class="divider"></li>
<li><a href="https://bioconductor.org" target="blank">Published by Bioconductor</a></li>

</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">Assigning cell types with SingleR</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="using-multiple-references" class="section level1">
<h1><span class="header-section-number">Chapter 5</span> Using multiple references</h1>
<script>
document.addEventListener("click", function (event) {
    if (event.target.classList.contains("aaron-collapse")) {
        event.target.classList.toggle("active");
        var content = event.target.nextElementSibling;
        if (content.style.display === "block") {
            content.style.display = "none";
        } else {
            content.style.display = "block";
        }
    }
})
</script>
<style>
.aaron-collapse {
  background-color: #eee;
  color: #444;
  cursor: pointer;
  padding: 18px;
  width: 100%;
  border: none;
  text-align: left;
  outline: none;
  font-size: 15px;
}

.aaron-content {
  padding: 0 18px;
  display: none;
  overflow: hidden;
  background-color: #f1f1f1;
}
</style>
<div id="overview-3" class="section level2">
<h2><span class="header-section-number">5.1</span> Overview</h2>
<p>In some cases, we may wish to use multiple references for annotation of a test dataset. This yields a more comprehensive set of cell types that are not covered by any individual reference, especially when differences in the resolution are considered. However, it is not trivial due to the presence of batch effects across references (from differences in technology, experimental protocol or the biological system) as well as differences in the annotation vocabulary between investigators.</p>
<p>Several strategies are available to combine inferences from multiple references:</p>
<ul>
<li>using reference-specific labels in a combined reference</li>
<li>using harmonized labels in a combined reference</li>
<li>combining scores across multiple references</li>
</ul>
<p>This chapter discusses the various strengths and weaknesses of each strategy and provides some practical demonstrations of each. Here, we will use the HPCA and BlueprintEncode datasets as our references and (yet another) PBMC dataset as the test.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(TENxPBMCData)
pbmc &lt;-<span class="st"> </span><span class="kw">TENxPBMCData</span>(<span class="st">&quot;pbmc8k&quot;</span>)

<span class="kw">library</span>(SingleR)
hpca &lt;-<span class="st"> </span><span class="kw">HumanPrimaryCellAtlasData</span>(<span class="dt">ensembl=</span><span class="ot">TRUE</span>)
bpe &lt;-<span class="st"> </span><span class="kw">BlueprintEncodeData</span>(<span class="dt">ensembl=</span><span class="ot">TRUE</span>)</code></pre></div>
</div>
<div id="using-reference-specific-labels" class="section level2">
<h2><span class="header-section-number">5.2</span> Using reference-specific labels</h2>
<p>In this strategy, each label is defined in the context of its reference dataset. This means that a label - say, “B cell” - in reference dataset X is considered to be different from a “B cell” label in reference dataset Y. Use of reference-specific labels is most appropriate if there are relevant biological differences between the references; for example, if one reference is concerned with healthy tissue while the other reference considers diseased tissue, it can be helpful to distinguish between the same cell type in different biological contexts.</p>
<p>We can easily implement this approach by combining the expression matrices together and pasting the reference name onto the corresponding character vector of labels. This modification ensures that the downstream <code>SingleR()</code> call will treat each label-reference combination as a distinct entity.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">hpca2 &lt;-<span class="st"> </span>hpca
hpca2<span class="op">$</span>label.main &lt;-<span class="st"> </span><span class="kw">paste0</span>(<span class="st">&quot;HPCA.&quot;</span>, hpca2<span class="op">$</span>label.main)

bpe2 &lt;-<span class="st"> </span>bpe
bpe2<span class="op">$</span>label.main &lt;-<span class="st"> </span><span class="kw">paste0</span>(<span class="st">&quot;BPE.&quot;</span>, bpe2<span class="op">$</span>label.main)

shared &lt;-<span class="st"> </span><span class="kw">intersect</span>(<span class="kw">rownames</span>(hpca2), <span class="kw">rownames</span>(bpe2))
combined &lt;-<span class="st"> </span><span class="kw">cbind</span>(hpca2[shared,], bpe2[shared,])</code></pre></div>
<p>It is then straightforward to perform annotation with the usual methods.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">com.res1 &lt;-<span class="st"> </span><span class="kw">SingleR</span>(pbmc, <span class="dt">ref=</span>combined, <span class="dt">labels=</span>combined<span class="op">$</span>label.main, <span class="dt">assay.type.test=</span><span class="dv">1</span>)
<span class="kw">table</span>(com.res1<span class="op">$</span>labels)</code></pre></div>
<pre><code>## 
##      BPE.B-cells BPE.CD4+ T-cells BPE.CD8+ T-cells          BPE.HSC 
##             1178             1708             2656               20 
##    BPE.Monocytes     BPE.NK cells  HPCA.HSC_-G-CSF   HPCA.Platelets 
##             2349              460                1                7 
##     HPCA.T_cells 
##                2</code></pre>
<p>However, this strategy identifies markers by directly comparing expression values across references, meaning that the marker set is likely to contain genes responsible for uninteresting batch effects. This will increase noise during the calculation of the score in each reference, possibly leading to a loss of precision and a greater risk of technical variation dominating the classification results. It also complicates interpretation as the cell type is always qualified by its reference of origin in the results.</p>
</div>
<div id="using-harmonized-labels" class="section level2">
<h2><span class="header-section-number">5.3</span> Using harmonized labels</h2>
<p>This strategy also involves combining the reference datasets into a single matrix but the labels are now harmonized so that the same cell type is given the same label across references. This allows feature selection methods to identify robust sets of label-specific markers that are more likely to generalize to other datasets. It also simplifies interpretation as there is no need to worry about the reference of origin for each label.</p>
<p>Many of the <em><a href="https://bioconductor.org/packages/3.12/SingleR">SingleR</a></em> datasets already have their labels mapped to the <a href="https://www.ebi.ac.uk/ols/ontologies/cl">Cell Ontology</a>. This provides a standard vocabulary to refer to the same cell type across multiple references. To simplify interpretation, we set <code>cell.ont=&quot;nonna&quot;</code> to remove all samples that could not be mapped to the ontology.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">hpca.ont &lt;-<span class="st"> </span><span class="kw">HumanPrimaryCellAtlasData</span>(<span class="dt">ensembl=</span><span class="ot">TRUE</span>, <span class="dt">cell.ont=</span><span class="st">&quot;nonna&quot;</span>)
bpe.ont &lt;-<span class="st"> </span><span class="kw">BlueprintEncodeData</span>(<span class="dt">ensembl=</span><span class="ot">TRUE</span>, <span class="dt">cell.ont=</span><span class="st">&quot;nonna&quot;</span>)

shared &lt;-<span class="st"> </span><span class="kw">intersect</span>(<span class="kw">rownames</span>(hpca.ont), <span class="kw">rownames</span>(bpe.ont))
combined.ont &lt;-<span class="st"> </span><span class="kw">cbind</span>(hpca.ont[shared,], bpe.ont[shared,])
combined.ont<span class="op">$</span>source &lt;-<span class="st"> </span><span class="kw">rep</span>(<span class="kw">c</span>(<span class="st">&quot;HPCA&quot;</span>, <span class="st">&quot;BPE&quot;</span>), <span class="kw">c</span>(<span class="kw">ncol</span>(hpca.ont), <span class="kw">ncol</span>(bpe.ont)))

<span class="co"># Showing the top 10 most frequent terms:</span>
tab &lt;-<span class="st"> </span><span class="kw">table</span>(combined.ont<span class="op">$</span>label.ont, combined.ont<span class="op">$</span>source)
<span class="kw">head</span>(tab[<span class="kw">order</span>(<span class="kw">rowSums</span>(tab), <span class="dt">decreasing=</span><span class="ot">TRUE</span>),])</code></pre></div>
<pre><code>##             
##              BPE HPCA
##   CL:0000235  18   83
##   CL:0000576  16   57
##   CL:0000451   1   57
##   CL:0000134   0   55
##   CL:0000775  23   21
##   CL:0002618   0   42</code></pre>
<p>We perform annotation with <code>SingleR()</code> as previously described. We set the <code>block=</code> argument in <code>de.args=</code> so as to indicate that DE genes should only be performed <em>within</em> each reference, and then the statistics merged <em>across</em> references to identify the top markers. This ensures that we do not directly compare expression values across references, which reduces the susceptibility of the marker detection to batch effects.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># </span><span class="al">TODO</span><span class="co">: add blocking mode for default marker detection.</span>
com.res2 &lt;-<span class="st"> </span><span class="kw">SingleR</span>(pbmc, <span class="dt">ref=</span>combined.ont, <span class="dt">labels=</span>combined.ont<span class="op">$</span>label.ont, 
    <span class="dt">assay.type.test=</span><span class="dv">1</span>)
tab &lt;-<span class="st"> </span><span class="kw">table</span>(com.res2<span class="op">$</span>labels)
<span class="kw">head</span>(<span class="kw">sort</span>(tab, <span class="dt">decreasing=</span><span class="ot">TRUE</span>))</code></pre></div>
<pre><code>## 
## CL:0000624 CL:0000576 CL:0000788 CL:0000815 CL:0000623 CL:0000625 
##       2571       2200        899        681        646        426</code></pre>
<p>The most obvious problem with this approach is that it assumes that harmonized labels are available. This is not always the case due to the use of different (often author-specific!) naming schemes, which requires at best semi-automated mapping of the author-derived labels to the standard vocabulary. The mapping process also runs the risk of discarding relevant information about the biological status (e.g., activation status, disease condition) if there is no obvious counterpart for that state in the ontology.</p>
</div>
<div id="comparing-scores-across-references" class="section level2">
<h2><span class="header-section-number">5.4</span> Comparing scores across references</h2>
<p>The final strategy - and the default approach implemented in <code>SingleR()</code> - involves performing classification separately within each reference, and then collating the results to choose the label with the highest score across references. This is a relatively expedient approach that avoids the need for explicit harmonization while also reducing exposure to reference-specific batch effects.</p>
<p>Use of this approach simply involves passing multiple objects to the <code>ref=</code> and <code>label=</code> argument in <code>SingleR()</code>. This instructs the function to annotate the test dataset with each reference individually; it then collects the best labels for each cell across all references and identifies the overall best-scoring label as the final prediction for that cell. The second step requires a recomputation of scores across a subset of relevant markers to ensure that these scores are comparable across references.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">com.res3 &lt;-<span class="st"> </span><span class="kw">SingleR</span>(<span class="dt">test =</span> pbmc, <span class="dt">assay.type.test=</span><span class="dv">1</span>,
    <span class="dt">ref =</span> <span class="kw">list</span>(<span class="dt">BPE=</span>bpe, <span class="dt">HPCA=</span>hpca), 
    <span class="dt">labels =</span> <span class="kw">list</span>(bpe<span class="op">$</span>label.main, hpca<span class="op">$</span>label.main))
<span class="kw">table</span>(com.res3<span class="op">$</span>labels)</code></pre></div>
<pre><code>## 
##           B_cell          B-cells     CD4+ T-cells     CD8+ T-cells 
##               14             1170             1450             2936 
##              GMP              HSC         Monocyte        Monocytes 
##                1               22              753             1560 
##         NK cells          NK_cell        Platelets Pre-B_cell_CD34- 
##              372               10                9               16 
##          T_cells 
##               68</code></pre>
<p>The main appeal of this approach lies in the fact that it is based on the results of annotation with individual references. This avoids batch effects from comparing expression values across references; it reduces the need for any coordination in the label scheme between references; and simultaneously provides the per-reference annotations in the results. The last feature is particularly useful as it allows for more detailed diagnostics, troubleshooting and further analysis.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">head</span>(com.res3<span class="op">$</span>orig.results<span class="op">$</span>BPE<span class="op">$</span>labels)</code></pre></div>
<pre><code>## [1] &quot;B-cells&quot;      &quot;Monocytes&quot;    &quot;CD8+ T-cells&quot; &quot;CD8+ T-cells&quot; &quot;Monocytes&quot;   
## [6] &quot;Monocytes&quot;</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">head</span>(com.res3<span class="op">$</span>orig.results<span class="op">$</span>HPCA<span class="op">$</span>labels)</code></pre></div>
<pre><code>## [1] &quot;B_cell&quot;   &quot;Monocyte&quot; &quot;T_cells&quot;  &quot;T_cells&quot;  &quot;Monocyte&quot; &quot;Monocyte&quot;</code></pre>
<p>The main downside is that it is somewhat suboptimal if there are many reference-specific labels, as markers are not identified with the aim of distinguishing a label in one reference from another label in another reference. The lack of harmonization in the labels also complicates interpretation of the results, though this can be addressed in the same manner as described above (i.e., replacing <code>label.main</code> with <code>label.ont</code>).</p>
</div>
<div id="session-info" class="section level2 unnumbered">
<h2>Session info</h2>
<button class="aaron-collapse">
View session info
</button>
<div class="aaron-content">
<pre><code>R version 4.0.0 Patched (2020-05-01 r78341)
Platform: x86_64-pc-linux-gnu (64-bit)
Running under: Ubuntu 18.04.4 LTS

Matrix products: default
BLAS:   /home/luna/Software/R/R-4-0-branch-dev/lib/libRblas.so
LAPACK: /home/luna/Software/R/R-4-0-branch-dev/lib/libRlapack.so

locale:
 [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              
 [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    
 [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   
 [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 
 [9] LC_ADDRESS=C               LC_TELEPHONE=C            
[11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       

attached base packages:
[1] parallel  stats4    stats     graphics  grDevices utils     datasets 
[8] methods   base     

other attached packages:
 [1] ensembldb_2.13.1            AnnotationFilter_1.13.0    
 [3] GenomicFeatures_1.41.0      AnnotationDbi_1.51.0       
 [5] SingleR_1.3.4               TENxPBMCData_1.7.0         
 [7] HDF5Array_1.17.0            rhdf5_2.33.0               
 [9] SingleCellExperiment_1.11.1 SummarizedExperiment_1.19.4
[11] DelayedArray_0.15.1         matrixStats_0.56.0         
[13] Biobase_2.49.0              GenomicRanges_1.41.1       
[15] GenomeInfoDb_1.25.0         IRanges_2.23.4             
[17] S4Vectors_0.27.6            BiocGenerics_0.35.2        
[19] BiocStyle_2.17.0            rebook_0.99.0              

loaded via a namespace (and not attached):
 [1] ProtGenerics_1.21.0           bitops_1.0-6                 
 [3] bit64_0.9-7                   progress_1.2.2               
 [5] httr_1.4.1                    tools_4.0.0                  
 [7] R6_2.4.1                      irlba_2.3.3                  
 [9] lazyeval_0.2.2                DBI_1.1.0                    
[11] prettyunits_1.1.1             tidyselect_1.1.0             
[13] processx_3.4.2                bit_1.1-15.2                 
[15] curl_4.3                      compiler_4.0.0               
[17] graph_1.67.0                  BiocNeighbors_1.7.0          
[19] rtracklayer_1.49.1            bookdown_0.19                
[21] askpass_1.1                   callr_3.4.3                  
[23] rappdirs_0.3.1                Rsamtools_2.5.0              
[25] stringr_1.4.0                 digest_0.6.25                
[27] rmarkdown_2.1                 XVector_0.29.0               
[29] pkgconfig_2.0.3               htmltools_0.4.0              
[31] dbplyr_1.4.3                  fastmap_1.0.1                
[33] rlang_0.4.6                   RSQLite_2.2.0                
[35] shiny_1.4.0.2                 DelayedMatrixStats_1.11.0    
[37] BiocParallel_1.23.0           dplyr_0.8.5                  
[39] RCurl_1.98-1.2                magrittr_1.5                 
[41] BiocSingular_1.5.0            GenomeInfoDbData_1.2.3       
[43] Matrix_1.2-18                 Rcpp_1.0.4.6                 
[45] Rhdf5lib_1.11.0               lifecycle_0.2.0              
[47] stringi_1.4.6                 yaml_2.2.1                   
[49] zlibbioc_1.35.0               BiocFileCache_1.13.0         
[51] AnnotationHub_2.21.0          grid_4.0.0                   
[53] blob_1.2.1                    promises_1.1.0               
[55] ExperimentHub_1.15.0          crayon_1.3.4                 
[57] lattice_0.20-41               beachmat_2.5.0               
[59] Biostrings_2.57.0             hms_0.5.3                    
[61] CodeDepends_0.6.5             knitr_1.28                   
[63] ps_1.3.3                      pillar_1.4.4                 
[65] codetools_0.2-16              biomaRt_2.45.0               
[67] XML_3.99-0.3                  glue_1.4.1                   
[69] BiocVersion_3.12.0            evaluate_0.14                
[71] BiocManager_1.30.10           vctrs_0.3.0                  
[73] httpuv_1.5.2                  openssl_1.4.1                
[75] purrr_0.3.4                   assertthat_0.2.1             
[77] xfun_0.13                     rsvd_1.0.3                   
[79] mime_0.9                      xtable_1.8-4                 
[81] later_1.0.0                   tibble_3.0.1                 
[83] GenomicAlignments_1.25.0      memoise_1.1.0                
[85] ellipsis_0.3.1                interactiveDisplayBase_1.27.0</code></pre>
</div>

</div>
</div>
            </section>

          </div>
        </div>
      </div>
<a href="annotation-diagnostics.html" class="navigation navigation-prev " aria-label="Previous page"><i class="fa fa-angle-left"></i></a>
<a href="harmonizing-labels.html" class="navigation navigation-next " aria-label="Next page"><i class="fa fa-angle-right"></i></a>
    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/lunr.js"></script>
<script src="libs/gitbook-2.6.7/js/clipboard.min.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-clipboard.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": {
"github": false,
"facebook": false,
"twitter": false,
"linkedin": false,
"weibo": false,
"instapaper": false,
"vk": false,
"all": ["facebook", "twitter", "linkedin", "weibo", "instapaper"]
},
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": {
"link": null,
"text": null
},
"history": {
"link": null,
"text": null
},
"view": {
"link": null,
"text": null
},
"download": null,
"toc": {
"collapse": "subsection"
},
"bookdown::epub_book": "default",
"cover": "images/cover.png"
});
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    var src = "";
    if (src === "" || src === "true") src = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-MML-AM_CHTML";
    if (location.protocol !== "file:")
      if (/^https?:/.test(src))
        src = src.replace(/^https?:/, '');
    script.src = src;
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>
</body>

</html>
